// @oracle/oraclejet-webdriver/elements/oj-tree-view/OjTreeView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjTreeView = void 0;
const OjTreeViewBase_1 = require("./OjTreeViewBase");
const __1 = require("../../");
const selenium_webdriver_1 = require("selenium-webdriver");
const oj_module_proxy_1 = require("../../lib/oj-module-proxy");
/**
 * The component WebElement for [oj-tree-view](../../jsdocs/oj.ojTreeView.html).
 * Do not instantiate this class directly, instead, use
 * [ojTreeView](../functions/elements.ojTreeView-1.html).
 */
class OjTreeView extends OjTreeViewBase_1.OjTreeViewBase {
    /**
     * Sets the value of "selected" property.
     * Specifies the current selected items in the treeview. See the Help documentation for more information.
     * @param selected The value to set for "selected"
     * @override
     * @typeparam K Type of keys
     */
    async changeSelected(selected) {
        await this.whenReady();
        return (0, oj_module_proxy_1.executeWithModules)(this.getDriver(), ['KeySet'], [this, selected], ({ KeySetImpl }, ele, selected) => {
            ele.selected = new KeySetImpl(selected);
        });
    }
    /**
     * Gets the value of "selected" property.
     * Retrieves the current selected items in the treeview. See the Help documentation for more information.
     * @override
     * @typeparam K Type of keys
     * @return The value of "selected" property.
     */
    async getSelected() {
        await this.whenReady();
        const selected = await this.getDriver().executeScript((ele) => Array.from(ele.selected.values()), this);
        return selected;
    }
    /**
     * Retrieve a SlotProxy which represents a single treeview item.
     * @param key The key within the Collection's dataset associated with the item.
     * @param id The id of the li item, used when the treeview is static.
     */
    async findItem(itemLocator) {
        let item;
        if (itemLocator.key) {
            await this.whenBusyContextReady();
            item = await (0, oj_module_proxy_1.executeWithModules)(this.getDriver(), ['KeyUtils'], [this, itemLocator.key], ({ equals }, element, key) => {
                const items = element.querySelectorAll('.oj-treeview-item');
                let item;
                for (var i = 0; i < items.length; i++) {
                    if (equals(items[i]['oj-item-metadata'].key, key)) {
                        item = items[i];
                        break;
                    }
                }
                if (item) {
                    return item.querySelector('.oj-treeview-item-content');
                }
            });
        }
        else if (itemLocator.id) {
            await this.whenBusyContextReady();
            item = await this.getDriver().executeScript((id) => {
                const item = document.getElementById(id);
                return item?.querySelector('.oj-treeview-item-content');
            }, itemLocator.id);
        }
        if (item) {
            return (0, __1.slotProxy)(item, this);
        }
        throw Error('No element found for given locator');
    }
    /**
     * Sets the value of "expanded" property.
     * Specifies the key set containing the keys of the items that should be expanded.
     * Use the ExpandedKeySet class to specify items to expand.
     * Use the ExpandAllKeySet class to expand all items.
     * @param expanded The value to set for "expanded"
     * @override
     */
    async changeExpanded(expanded) {
        await this.whenReady();
        return (0, oj_module_proxy_1.executeWithModules)(this.getDriver(), ['KeySet'], [this, expanded], ({ ExpandedKeySet }, ele, expanded) => {
            const expandedKeySet = new ExpandedKeySet(expanded);
            ele.expanded = expandedKeySet;
        });
    }
    /**
     * Expands a collapsed treeview item. If already expanded it will remain expanded.
     * @param key The key of the item that is being expanded.
     * @param id The id of the li item to be expanded, used when the treeview is static.
     */
    async doExpand(itemLocator) {
        let itemContent = await this.findItem(itemLocator);
        let isExpanded = await this._isExpanded(itemContent);
        if (!isExpanded) {
            const ins = await this._getDisclosureIcon(itemContent);
            return ins.click();
        }
    }
    /**
     * Collapses an expanded treeview item. If already collapsed it will remain collapsed.
     * @param key The key of the item that is being collapsed.
     * @param id The id of the li item to be collapsed, used when the treeview is static.
     */
    async doCollapse(itemLocator) {
        let itemContent = await this.findItem(itemLocator);
        let isExpanded = await this._isExpanded(itemContent);
        if (isExpanded) {
            const ins = await this._getDisclosureIcon(itemContent);
            return ins.click();
        }
    }
    /**
     * Returns expanded state.
     * @param itemContent The itemContent WebElement that is being checked.
     */
    async _isExpanded(itemContent) {
        const ins = await this._getDisclosureIcon(itemContent);
        const driver = await ins.getDriver();
        return driver.executeScript((el) => el.parentNode.parentNode.classList.contains('oj-expanded'), ins);
    }
    /**
     * Returns the ins element.
     * @param itemContent The itemContent WebElement that contains the disclosure icon.
     */
    async _getDisclosureIcon(itemContent) {
        return itemContent.findElement(selenium_webdriver_1.By.css('INS'));
    }
}
exports.OjTreeView = OjTreeView;
