// @oracle/oraclejet-preact/cjs/UNSAFE_BaseButton/themes/stable/BaseButtonTheme.dark.tokens.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('../../../utils-a66430fc.js');
var themeContract_css = require('../../../themeContract.css-a53fd740.js');
require('../../../colorUtils-01f3caa2.js');
require('../../../_curry1-e8f0d7ea.js');

const colorSchemeDark = {
    variants: {
        outlined: {
            base: {
                backgroundColor: 'transparent',
                borderColor: themeContract_css.colorSchemeVars.border.enabled,
                color: themeContract_css.colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: 'transparent',
                borderColor: themeContract_css.colorSchemeVars.border.disabled,
                color: themeContract_css.colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.hover
            },
            active: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.active
            }
        },
        callToAction: {
            base: {
                backgroundColor: utils.rgb(themeContract_css.colorSchemeVars.palette.neutral[200]),
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.inverse
            },
            disabled: {
                backgroundColor: themeContract_css.colorSchemeVars.surface.disabled,
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.inverseHover
            },
            active: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.inverseActive
            }
        },
        borderless: {
            base: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.hover
            },
            active: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.active
            }
        },
        ghost: {
            base: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: 'transparent'
            },
            active: {
                backgroundOverlayColor: 'transparent'
            }
        },
        solid: {
            base: {
                backgroundColor: utils.rgba(themeContract_css.colorSchemeVars.palette.neutral['190'], 0.08),
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: themeContract_css.colorSchemeVars.surface.disabled,
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.hover
            },
            active: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.active
            }
        },
        danger: {
            base: {
                backgroundColor: themeContract_css.colorSchemeVars.surface.danger.strong,
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.inverse
            },
            disabled: {
                backgroundColor: themeContract_css.colorSchemeVars.surface.disabled,
                borderColor: 'transparent',
                color: themeContract_css.colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.dangerHover
            },
            active: {
                backgroundOverlayColor: themeContract_css.colorSchemeVars.overlay.dangerActive
            }
        }
    }
};

exports.colorSchemeDark = colorSchemeDark;
