// @oracle/oraclejet-webdriver/elements/oj-select-one/OjSelectOneBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjSelectOneBase = void 0;
const __1 = require("../");
/**
 * This is the base class for oj-select-one WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjSelectOne.ts.
 */
class OjSelectOneBase extends __1.OjWebElement {
    /**
     * Gets the value of <code>describedBy</code> property.
     * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
     * @return The value of <code>describedBy</code> property.
     *
     */
    getDescribedBy() {
        return this.getProperty('describedBy');
    }
    /**
     * Gets the value of <code>disabled</code> property.
     * Specifies whether the component is disabled. The default is false.
     * @return The value of <code>disabled</code> property.
     *
     */
    getDisabled() {
        return this.getProperty('disabled');
    }
    /**
     * Gets the value of <code>displayOptions</code> property.
     * Display options for auxiliary content that determines whether or not it should be displayed.
     * @return The value of <code>displayOptions</code> property.
     *
     */
    getDisplayOptions() {
        return this.getProperty('displayOptions');
    }
    /**
     * Gets the value of <code>help</code> property.
     * Form component help information.
     * @return The value of <code>help</code> property.
     *
     */
    getHelp() {
        return this.getProperty('help');
    }
    /**
     * Gets the value of <code>helpHints</code> property.
     * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
     * @return The value of <code>helpHints</code> property.
     *
     */
    getHelpHints() {
        return this.getProperty('helpHints');
    }
    /**
     * Gets the value of <code>labelEdge</code> property.
     * Defines how the label of a component is created. See the Help documentation for more information.
     * @return The value of <code>labelEdge</code> property.
     *
     */
    getLabelEdge() {
        return this.getProperty('labelEdge');
    }
    /**
     * Gets the value of <code>labelHint</code> property.
     * Represents a hint for oj-form-layout element to render a label on the editable component.
     * @return The value of <code>labelHint</code> property.
     *
     */
    getLabelHint() {
        return this.getProperty('labelHint');
    }
    /**
     * Gets the value of <code>labelledBy</code> property.
     * The oj-label sets the labelledBy property programmatically on the form component.
     * @return The value of <code>labelledBy</code> property.
     *
     */
    getLabelledBy() {
        return this.getProperty('labelledBy');
    }
    /**
     * Gets the value of <code>maximumResultCount</code> property.
     * The maximum number of results displayed in the dropdown.
     * @return The value of <code>maximumResultCount</code> property.
     *
     */
    getMaximumResultCount() {
        return this.getProperty('maximumResultCount');
    }
    /**
     * Sets the value of <code>messagesCustom</code> property.
     * A list of messages added by an application to the component. See the Help documentation for more information.
     * @param messagesCustom The value to set for <code>messagesCustom</code>
     *
     */
    changeMessagesCustom(messagesCustom) {
        return this.setProperty('messagesCustom', messagesCustom);
    }
    /**
     * Gets the value of <code>messagesCustom</code> property.
     * A list of messages added by an application to the component. See the Help documentation for more information.
     * @return The value of <code>messagesCustom</code> property.
     *
     */
    getMessagesCustom() {
        return this.getProperty('messagesCustom');
    }
    /**
     * Gets the value of <code>minimumResultsForSearch</code> property.
     * The threshold for showing the search box in the dropdown.
     * @return The value of <code>minimumResultsForSearch</code> property.
     *
     */
    getMinimumResultsForSearch() {
        return this.getProperty('minimumResultsForSearch');
    }
    /**
     * Gets the value of <code>optionRenderer</code> property.
     * The renderer function that renders the content of each option.
     * @return The value of <code>optionRenderer</code> property.
     *
     */
    getOptionRenderer() {
        return this.getProperty('optionRenderer');
    }
    /**
     * Gets the value of <code>options</code> property.
     * The option items for the Select.
     * @return The value of <code>options</code> property.
     *
     */
    getOptions() {
        return this.getProperty('options');
    }
    /**
     * Gets the value of <code>optionsKeys</code> property.
     * Specify the key names to use in the options array. Depending on options-keys means that the signature of the data does not match what is supported by the options attribute.
     * @return The value of <code>optionsKeys</code> property.
     *
     */
    getOptionsKeys() {
        return this.getProperty('optionsKeys');
    }
    /**
     * Gets the value of <code>pickerAttributes</code> property.
     * The style attributes for the drop down.
     * @return The value of <code>pickerAttributes</code> property.
     *
     */
    getPickerAttributes() {
        return this.getProperty('pickerAttributes');
    }
    /**
     * Gets the value of <code>placeholder</code> property.
     * A short hint that can be displayed before user selects a value.
     * @return The value of <code>placeholder</code> property.
     *
     */
    getPlaceholder() {
        return this.getProperty('placeholder');
    }
    /**
     * Gets the value of <code>readonly</code> property.
     * Specifies whether a value is readonly
     * @return The value of <code>readonly</code> property.
     *
     */
    getReadonly() {
        return this.getProperty('readonly');
    }
    /**
     * Gets the value of <code>readonlyUserAssistanceShown</code> property.
     * Specifies which user assistance types should be shown when the component is readonly.
     * @return The value of <code>readonlyUserAssistanceShown</code> property.
     *
     */
    getReadonlyUserAssistanceShown() {
        return this.getProperty('readonlyUserAssistanceShown');
    }
    /**
     * Gets the value of <code>renderMode</code> property.
     * Specifies whether to render select in JET or as a HTML Select tag.
     * @return The value of <code>renderMode</code> property.
     * @deprecated Since 8.0.0. Support for "native" mode rendering is deprecated because JET promotes a consistent Oracle UX based upon the Redwood design system. As a result, the theme variable "$selectRenderModeOptionDefault" is also deprecated.
     */
    getRenderMode() {
        return this.getProperty('renderMode');
    }
    /**
     * Gets the value of <code>required</code> property.
     * Specifies whether the component is required or optional. See the Help documentation for more information.
     * @return The value of <code>required</code> property.
     *
     */
    getRequired() {
        return this.getProperty('required');
    }
    /**
     * Gets the value of <code>translations</code> property.
     * A collection of translated resources from the translation bundle, or null if this component has no resources.
     * @return The value of <code>translations</code> property.
     *
     */
    getTranslations() {
        return this.getProperty('translations');
    }
    /**
     * Gets the value of <code>userAssistanceDensity</code> property.
     * Specifies the density of the form component's user assistance presentation.
     * @return The value of <code>userAssistanceDensity</code> property.
     *
     */
    getUserAssistanceDensity() {
        return this.getProperty('userAssistanceDensity');
    }
    /**
     * Gets the value of <code>valid</code> property.
     * The validity state of the component
     * @return The value of <code>valid</code> property.
     *
     */
    getValid() {
        return this.getProperty('valid');
    }
    /**
     * Sets the value of <code>value</code> property.
     * The value of the element.
     * @param value The value to set for <code>value</code>
     *
     */
    changeValue(value) {
        return this.setProperty('value', value);
    }
    /**
     * Gets the value of <code>value</code> property.
     * The value of the element.
     * @return The value of <code>value</code> property.
     *
     */
    getValue() {
        return this.getProperty('value');
    }
    /**
     * Sets the value of <code>valueOption</code> property.
     * The current value of the element and its associated display label.
     * @param valueOption The value to set for <code>valueOption</code>
     *
     */
    changeValueOption(valueOption) {
        return this.setProperty('valueOption', valueOption);
    }
    /**
     * Gets the value of <code>valueOption</code> property.
     * The current value of the element and its associated display label.
     * @return The value of <code>valueOption</code> property.
     *
     */
    getValueOption() {
        return this.getProperty('valueOption');
    }
}
exports.OjSelectOneBase = OjSelectOneBase;
