// @oracle/oraclejet-preact/cjs/UNSAFE_TableView/__testing__/TableViewAdapter.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var UNSAFE_Driver = require('@oracle/oraclejet-testing/UNSAFE_Driver');
var UNSAFE_Locators = require('@oracle/oraclejet-testing/UNSAFE_Locators');
var UNSAFE_TestAdapter = require('@oracle/oraclejet-testing/UNSAFE_TestAdapter');

class TableViewAdapter extends UNSAFE_TestAdapter.TestAdapter {
    /**
     * Sort the table view by the column specified
     * @param columnKey The columnKey of the column
     * @returns Promise<void>
     */
    async doSort(columnKey) {
        const driver = this.getTestDriver();
        const tableView = await this.getElement();
        let header;
        try {
            header = await tableView.waitForElement(UNSAFE_Locators.byCss(`[data-oj-cell-type="header"][data-oj-column-key="${columnKey}"]`));
        }
        catch (e) {
            throw new Error(`Specified column key ${columnKey} is invalid or does not exist`);
        }
        return driver.sendKeys(header, UNSAFE_Driver.Keys['ENTER']);
    }
    /**
     * Triggers rowAction event
     * @param rowKey The rowKey
     * @returns Promise<void>
     */
    async doRowAction(rowKey) {
        const tableView = await this.getElement();
        let row;
        try {
            row = await tableView.waitForElement(UNSAFE_Locators.byCss(`[data-oj-key="${rowKey}"]`));
        }
        catch (e) {
            throw new Error(`Specified row key ${rowKey} is invalid or does not exist`);
        }
        return this._clickItem(row);
    }
    /**
     * Toggles row selection
     * @param rowKey The rowKey
     * @returns Promise<void>
     */
    async doToggleSelectRow(rowKey) {
        const tableView = await this.getElement();
        let row;
        try {
            row = await tableView.waitForElement(UNSAFE_Locators.byCss(`[data-oj-key="${rowKey}"]`));
        }
        catch (e) {
            throw new Error(`Specified row key ${rowKey} is invalid or does not exist`);
        }
        return this._clickItem(row, { modifiers: [UNSAFE_Driver.Keys['CONTROL_COMMAND']] });
    }
    /**
     * Toggles column selection
     * @param columnKey The columnKey
     * @returns Promise<void>
     */
    async doToggleSelectColumn(columnKey) {
        const tableView = await this.getElement();
        let column;
        try {
            column = await tableView.waitForElement(UNSAFE_Locators.byCss(`[data-oj-cell-type="header"][data-oj-column-key="${columnKey}"]`));
        }
        catch (e) {
            throw new Error(`Specified column key ${columnKey} is invalid or does not exist`);
        }
        return this._clickItem(column, { modifiers: [UNSAFE_Driver.Keys['CONTROL_COMMAND']] });
    }
    /**
     * Perform a TestDriver click() on the item, but first ensure that it's scrolled into
     * view. TODO: this shouldn't be needed once JET-73795 is fixed.
     * @param item
     * @param options
     * @returns
     */
    async _clickItem(item, options) {
        const driver = this.getTestDriver();
        await driver.executeScript((el) => el.scrollIntoView({ block: 'center', inline: 'center' }), driver.unwrapElement(item));
        return driver.click(item, options);
    }
    /**
     * Load more rows
     * @param locator The TestIdLocator which identifies the TableView
     * @returns Promise<void>
     */
    async doLoadMore() {
        const driver = this.getTestDriver();
        const tableView = await this.getElement();
        const rowCount = (await tableView.waitForElements(UNSAFE_Locators.byCss('tr[role="row"]'))).length;
        await driver.sendKeys(tableView, UNSAFE_Driver.Keys['CONTROL_COMMAND'], UNSAFE_Driver.Keys['END']);
        await UNSAFE_Driver.waitFor(async () => {
            return ((await tableView.waitForElements(UNSAFE_Locators.byCss('tr[role="row"]'))).length != rowCount ||
                (await tableView.waitForElements(UNSAFE_Locators.byCss('[data-oj-cell-type="loadMore"]'))).length === 0);
        });
    }
}

exports.TableViewAdapter = TableViewAdapter;
