// @oracle/oraclejet-webdriver/elements/oj-drawer-layout/OjDrawerLayoutBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjDrawerLayoutBase = void 0;
const __1 = require("../");
/**
 * This is the base class for oj-drawer-layout WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjDrawerLayout.ts.
 */
class OjDrawerLayoutBase extends __1.OjWebElement {
    /**
     * Sets the value of <code>startOpened</code> property.
     * Specifies the 'opened' state of the Start drawer.
     * @param startOpened The value to set for <code>startOpened</code>
     *
     */
    changeStartOpened(startOpened) {
        return this.setProperty('startOpened', startOpened);
    }
    /**
     * Gets the value of <code>startOpened</code> property.
     * Specifies the 'opened' state of the Start drawer.
     * @return The value of <code>startOpened</code> property.
     *
     */
    getStartOpened() {
        return this.getProperty('startOpened');
    }
    /**
     * Sets the value of <code>endOpened</code> property.
     * Specifies the 'opened' state of the End drawer.
     * @param endOpened The value to set for <code>endOpened</code>
     *
     */
    changeEndOpened(endOpened) {
        return this.setProperty('endOpened', endOpened);
    }
    /**
     * Gets the value of <code>endOpened</code> property.
     * Specifies the 'opened' state of the End drawer.
     * @return The value of <code>endOpened</code> property.
     *
     */
    getEndOpened() {
        return this.getProperty('endOpened');
    }
    /**
     * Sets the value of <code>bottomOpened</code> property.
     * Specifies the 'opened' state of the Bottom drawer.
     * @param bottomOpened The value to set for <code>bottomOpened</code>
     *
     */
    changeBottomOpened(bottomOpened) {
        return this.setProperty('bottomOpened', bottomOpened);
    }
    /**
     * Gets the value of <code>bottomOpened</code> property.
     * Specifies the 'opened' state of the Bottom drawer.
     * @return The value of <code>bottomOpened</code> property.
     *
     */
    getBottomOpened() {
        return this.getProperty('bottomOpened');
    }
    /**
     * Gets the value of <code>startDisplay</code> property.
     * Specifies the display mode of the Start drawer.
     * @return The value of <code>startDisplay</code> property.
     *
     */
    getStartDisplay() {
        return this.getProperty('startDisplay');
    }
    /**
     * Gets the value of <code>endDisplay</code> property.
     * Specifies the display mode of the End drawer.
     * @return The value of <code>endDisplay</code> property.
     *
     */
    getEndDisplay() {
        return this.getProperty('endDisplay');
    }
    /**
     * Gets the value of <code>bottomDisplay</code> property.
     * Specifies the display mode of the Bottom drawer.
     * @return The value of <code>bottomDisplay</code> property.
     *
     */
    getBottomDisplay() {
        return this.getProperty('bottomDisplay');
    }
}
exports.OjDrawerLayoutBase = OjDrawerLayoutBase;
