// @oracle/oraclejet-webdriver/elements/oj-combobox-many/OjComboboxManyBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjComboboxManyBase = void 0;
const __1 = require("../");
/**
 * This is the base class for oj-combobox-many WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjComboboxMany.ts.
 */
class OjComboboxManyBase extends __1.OjWebElement {
    /**
     * Gets the value of <code>asyncValidators</code> property.
     * Specifies a list of asynchronous validators used by the component when performing validation. Use async-validators when you need to perform some validation work on the server. See the Help documentation for more information.
     * @return The value of <code>asyncValidators</code> property.
     * @deprecated Since 8.0.0. Use the validators property instead for either regular Validators or AsyncValidators.
     */
    getAsyncValidators() {
        return this.getProperty('asyncValidators');
    }
    /**
     * Gets the value of <code>converter</code> property.
     * An object that converts the value. See the Help documentation for more information.
     * @return The value of <code>converter</code> property.
     *
     */
    getConverter() {
        return this.getProperty('converter');
    }
    /**
     * Gets the value of <code>describedBy</code> property.
     * The form component's oj-label automatically sets described-by to make it accessible. It is not meant to be set by application developer.
     * @return The value of <code>describedBy</code> property.
     *
     */
    getDescribedBy() {
        return this.getProperty('describedBy');
    }
    /**
     * Gets the value of <code>disabled</code> property.
     * Specifies whether the component is disabled. The default is false.
     * @return The value of <code>disabled</code> property.
     *
     */
    getDisabled() {
        return this.getProperty('disabled');
    }
    /**
     * Gets the value of <code>displayOptions</code> property.
     * Display options for auxiliary content that determines whether or not it should be displayed.
     * @return The value of <code>displayOptions</code> property.
     *
     */
    getDisplayOptions() {
        return this.getProperty('displayOptions');
    }
    /**
     * Gets the value of <code>help</code> property.
     * Form component help information.
     * @return The value of <code>help</code> property.
     *
     */
    getHelp() {
        return this.getProperty('help');
    }
    /**
     * Gets the value of <code>helpHints</code> property.
     * Represents hints for an oj-form-layout element to render help information on the label of the editable component.
     * @return The value of <code>helpHints</code> property.
     *
     */
    getHelpHints() {
        return this.getProperty('helpHints');
    }
    /**
     * Gets the value of <code>labelEdge</code> property.
     * Defines how the label of a component is created. See the Help documentation for more information.
     * @return The value of <code>labelEdge</code> property.
     *
     */
    getLabelEdge() {
        return this.getProperty('labelEdge');
    }
    /**
     * Gets the value of <code>labelHint</code> property.
     * Represents a hint for oj-form-layout element to render a label on the editable component.
     * @return The value of <code>labelHint</code> property.
     *
     */
    getLabelHint() {
        return this.getProperty('labelHint');
    }
    /**
     * Gets the value of <code>labelledBy</code> property.
     * The oj-label sets the labelledBy property programmatically on the form component.
     * @return The value of <code>labelledBy</code> property.
     *
     */
    getLabelledBy() {
        return this.getProperty('labelledBy');
    }
    /**
     * Gets the value of <code>maximumResultCount</code> property.
     * The maximum number of results displayed in the dropdown.
     * @return The value of <code>maximumResultCount</code> property.
     *
     */
    getMaximumResultCount() {
        return this.getProperty('maximumResultCount');
    }
    /**
     * Sets the value of <code>messagesCustom</code> property.
     * A list of messages added by an application to the component. See the Help documentation for more information.
     * @param messagesCustom The value to set for <code>messagesCustom</code>
     *
     */
    changeMessagesCustom(messagesCustom) {
        return this.setProperty('messagesCustom', messagesCustom);
    }
    /**
     * Gets the value of <code>messagesCustom</code> property.
     * A list of messages added by an application to the component. See the Help documentation for more information.
     * @return The value of <code>messagesCustom</code> property.
     *
     */
    getMessagesCustom() {
        return this.getProperty('messagesCustom');
    }
    /**
     * Gets the value of <code>minLength</code> property.
     * The minimum number of characters a user must type before search filtering is performed.
     * @return The value of <code>minLength</code> property.
     *
     */
    getMinLength() {
        return this.getProperty('minLength');
    }
    /**
     * Gets the value of <code>optionRenderer</code> property.
     * The renderer function that renders the content of each option.
     * @return The value of <code>optionRenderer</code> property.
     *
     */
    getOptionRenderer() {
        return this.getProperty('optionRenderer');
    }
    /**
     * Gets the value of <code>options</code> property.
     * The option items for the Combobox.
     * @return The value of <code>options</code> property.
     *
     */
    getOptions() {
        return this.getProperty('options');
    }
    /**
     * Gets the value of <code>optionsKeys</code> property.
     * Specify the key names to use in the options array. Depending on options-keys means that the signature of the data does not match what is supported by the options attribute.
     * @return The value of <code>optionsKeys</code> property.
     *
     */
    getOptionsKeys() {
        return this.getProperty('optionsKeys');
    }
    /**
     * Gets the value of <code>pickerAttributes</code> property.
     * The style attributes for the drop down.
     * @return The value of <code>pickerAttributes</code> property.
     *
     */
    getPickerAttributes() {
        return this.getProperty('pickerAttributes');
    }
    /**
     * Gets the value of <code>placeholder</code> property.
     * A short hint that can be displayed before user selects or enters a value.
     * @return The value of <code>placeholder</code> property.
     *
     */
    getPlaceholder() {
        return this.getProperty('placeholder');
    }
    /**
     * Gets the value of <code>rawValue</code> property.
     * The currently displayed text retrieved from the input field.
     * @return The value of <code>rawValue</code> property.
     *
     */
    getRawValue() {
        return this.getProperty('rawValue');
    }
    /**
     * Gets the value of <code>readonly</code> property.
     * Specifies whether the component is read-only. A read-only element cannot be modified, but user interaction is allowed. See the Help documentation for more information.
     * @return The value of <code>readonly</code> property.
     *
     */
    getReadonly() {
        return this.getProperty('readonly');
    }
    /**
     * Gets the value of <code>readonlyUserAssistanceShown</code> property.
     * Specifies which user assistance types should be shown when the component is readonly.
     * @return The value of <code>readonlyUserAssistanceShown</code> property.
     *
     */
    getReadonlyUserAssistanceShown() {
        return this.getProperty('readonlyUserAssistanceShown');
    }
    /**
     * Gets the value of <code>required</code> property.
     * Specifies whether the component is required or optional. See the Help documentation for more information.
     * @return The value of <code>required</code> property.
     *
     */
    getRequired() {
        return this.getProperty('required');
    }
    /**
     * Gets the value of <code>translations</code> property.
     * A collection of translated resources from the translation bundle, or null if this component has no resources.
     * @return The value of <code>translations</code> property.
     *
     */
    getTranslations() {
        return this.getProperty('translations');
    }
    /**
     * Gets the value of <code>userAssistanceDensity</code> property.
     * Specifies the density of the form component's user assistance presentation.
     * @return The value of <code>userAssistanceDensity</code> property.
     *
     */
    getUserAssistanceDensity() {
        return this.getProperty('userAssistanceDensity');
    }
    /**
     * Gets the value of <code>valid</code> property.
     * The validity state of the component
     * @return The value of <code>valid</code> property.
     *
     */
    getValid() {
        return this.getProperty('valid');
    }
    /**
     * Gets the value of <code>validators</code> property.
     * Specifies a list of synchronous validators for performing validation by the element. See the Help documentation for more information.
     * @return The value of <code>validators</code> property.
     *
     */
    getValidators() {
        return this.getProperty('validators');
    }
    /**
     * Sets the value of <code>value</code> property.
     * The value of the element.
     * @param value The value to set for <code>value</code>
     *
     */
    changeValue(value) {
        return this.setProperty('value', value);
    }
    /**
     * Gets the value of <code>value</code> property.
     * The value of the element.
     * @return The value of <code>value</code> property.
     *
     */
    getValue() {
        return this.getProperty('value');
    }
    /**
     * Sets the value of <code>valueOptions</code> property.
     * The current values of the element and their associated display labels.
     * @param valueOptions The value to set for <code>valueOptions</code>
     *
     */
    changeValueOptions(valueOptions) {
        return this.setProperty('valueOptions', valueOptions);
    }
    /**
     * Gets the value of <code>valueOptions</code> property.
     * The current values of the element and their associated display labels.
     * @return The value of <code>valueOptions</code> property.
     *
     */
    getValueOptions() {
        return this.getProperty('valueOptions');
    }
}
exports.OjComboboxManyBase = OjComboboxManyBase;
