// @oracle/oraclejet-webdriver/src/elements/oj-refresher/OjRefresherBase.ts
import { OjWebElement } from '../';

/**
 * This is the base class for oj-refresher WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjRefresher.ts.
 */
export class OjRefresherBase extends OjWebElement {
  /**
   * Gets the value of <code>refreshContent</code> property.
   * Specifies the function to invoke when pull-to-refresh operation is triggered. See the Help documentation for more information.
   * @return The value of <code>refreshContent</code> property.
   *
   */
  public getRefreshContent(): Promise<null> {
    return this.getProperty<null>('refreshContent');
  }

  /**
   * Gets the value of <code>text</code> property.
   * Specifies a text message to display in the pull-to-refresh panel after a pull down gesture.
   * @return The value of <code>text</code> property.
   *
   */
  public getTextProperty(): Promise<string> {
    return this.getProperty<string>('text');
  }

  /**
   * Gets the value of <code>threshold</code> property.
   * The minimum distance in pixels that the user needs to pull down in order to trigger a refresh. See the Help documentation for more information.
   * @return The value of <code>threshold</code> property.
   *
   */
  public getThreshold(): Promise<number> {
    return this.getProperty<number>('threshold');
  }

  /**
   * Gets the value of <code>translations</code> property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of <code>translations</code> property.
   *
   */
  public getTranslations(): Promise<Translations | null> {
    return this.getProperty<Translations | null>('translations');
  }
}
export interface Translations {
  /**
   * Label assigned to the refresh message to denote refresh complete that appears after triggering refresh content.
   */
  ariaRefreshCompleteLink: string;
  /**
   * Label assigned to the refresh link used for triggering refresh content when using keyboard or accessibility agents such as VoiceOver.
   */
  ariaRefreshLink: string;
  /**
   * Label assigned to the refresh message to denote that the refresh is occurring that appears after triggering refresh content.
   */
  ariaRefreshingLink: string;
}
