// @oracle/oraclejet-preact/es/BaseButton-c2fbaa6d.js
/* @oracle/oraclejet-preact: undefined */
import { jsx } from 'preact/jsx-runtime';
import { forwardRef } from 'preact/compat';
import { u as useHover } from './useHover-85077adb.js';
import { u as useActive } from './useActive-fb6d4d52.js';
import { u as useColorScheme } from './useColorScheme-96cce6a1.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { d as dimensionInterpolations } from './dimensions-ddde0eee.js';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import { g as getClientHints } from './clientHints-a42b510c.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { BaseButtonRedwoodTheme } from './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { B as BareButton } from './BareButton-0954f33d.js';
import { p as processGlobals } from './buttonUtils-55fb1df7.js';

const interpolations = [...Object.values(dimensionInterpolations)];
const styleInterpolations = mergeInterpolations(interpolations);
const clientHints = getClientHints();
// :active only works on IOS devices if a touch handler exists
const iosProps = clientHints.platform === 'ios' ? { ontouchstart: function () { } } : {};
const isHybrid = getClientHints().hoverSupport === 'events';
const BaseButton = forwardRef(({ variant, isReadOnly = false, isDisabled = false, isRepeat = false, isClick, size, onAction, elementDetails = { type: 'button', buttonType: 'button' }, styling = ['default'], edge = 'none', testId, ...props }, ref) => {
    const isStyled = (name) => {
        return styling.indexOf(name) > -1;
    };
    const { globalEventProps, ariaProps } = processGlobals(props);
    const { hoverProps, isHover } = useHover({ isDisabled: !isHybrid });
    const { activeProps, isActive } = useActive({ isDisabled: isReadOnly });
    const { class: cls, ...styles } = styleInterpolations({
        ...props,
        ...globalEventProps
    });
    const ElementType = elementDetails.type;
    const isSpan = ElementType == 'span';
    const colorScheme = useColorScheme();
    const { variantClasses } = useComponentTheme(BaseButtonRedwoodTheme, {
        size,
        variant,
        unstyled: isStyled('unstyled') ? 'isUnstyled' : 'notUnstyled',
        embedded: isStyled('embedded') ? 'isEmbedded' : 'notEmbedded',
        container: isStyled('container') ? 'isContainer' : 'notContainer',
        noBorderRadiusStart: isStyled('noBorderRadiusStart')
            ? 'isNoBorderRadiusStart'
            : 'notNoBorderRadiusStart',
        noBorderRadiusEnd: isStyled('noBorderRadiusEnd')
            ? 'isNoBorderRadiusEnd'
            : 'notNoBorderRadiusEnd',
        noBorderStart: isStyled('noBorderStart') ? 'isNoBorderStart' : 'notNoBorderStart',
        noBorderEnd: isStyled('noBorderEnd') ? 'isNoBorderEnd' : 'notNoBorderEnd',
        activeStyle: isStyled('active') ? 'isActiveStyle' : 'notActiveStyle',
        selectedStyle: isStyled('selected') ? 'isSelectedStyle' : 'notSelectedStyle',
        buttonSetItem: isStyled('buttonSetItem') ? 'isButtonSetItem' : 'notButtonSetItem',
        trailingItem: isStyled('trailingItem') ? 'isTrailingItem' : 'notTrailingItem',
        fill: isStyled('fill') ? 'isFill' : 'notFill',
        min: isStyled('min') ? 'isMin' : 'notMin',
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        edge,
        hybrid: isHybrid ? 'isHybrid' : 'notHybrid',
        active: isActive ? 'isActive' : 'notActive',
        hover: isHover ? 'isHover' : 'notHover',
        span: isSpan ? 'isSpan' : 'notSpan',
        widthEqual: isStyled('widthEqual') ? 'isWidthEqual' : 'notWidthEqual',
        widthAuto: isStyled('widthAuto') ? 'isWidthAuto' : 'notWidthAuto',
        hiddenLabel: isStyled('hiddenLabel') ? 'isHidden' : 'notHidden',
        noHoverActive: isStyled('noHoverActive') ? 'isNoHoverActive' : 'notNoHoverActive',
        newStackingContext: isStyled('newStackingContext')
            ? 'isNewStackingContext'
            : 'notNewStackingContext',
        colorScheme
    });
    const legacyGlobals = {
        autofocus: props.autofocus,
        title: props.title
    };
    const mergedProps = mergeProps(iosProps, ariaProps, hoverProps, isStyled('container') ? {} : activeProps, globalEventProps, legacyGlobals);
    const wrapperClasses = classNames([cls, variantClasses]);
    return (jsx(BareButton, { ref: ref, isDisabled: isDisabled || isReadOnly, isClick: isClick, classNames: wrapperClasses, style: styles, testId: testId, onAction: onAction, autofocus: props['autofocus'], elementDetails: elementDetails, isRepeat: isRepeat, ...ariaProps, ...mergedProps, children: props.children }));
});
BaseButton.displayName = 'BaseButton';

export { BaseButton as B };
