// @oracle/oraclejet-preact/es/buttonUtils-55fb1df7.js
/* @oracle/oraclejet-preact: undefined */
import { g as generateId } from './id-83adac50.js';

/**
 *
 * @param buttonProps
 * @returns globalEventProps that can be spread onto target element
 */
function processGlobals(buttonProps) {
    const globalEventProps = {
        onBlur: buttonProps.onBlur,
        onFocus: buttonProps.onFocus,
        onMouseEnter: buttonProps.onMouseEnter,
        onMouseLeave: buttonProps.onMouseLeave,
        onMouseDown: buttonProps.onMouseDown,
        onMouseUp: buttonProps.onMouseUp,
        onTouchEnd: buttonProps.onTouchEnd,
        onTouchStart: buttonProps.onTouchStart,
        onTouchCancel: buttonProps.onTouchCancel,
        onKeyDown: buttonProps.onKeyDown,
        onKeyUp: buttonProps.onKeyUp,
        autofocus: buttonProps.autofocus,
        title: buttonProps.title
    };
    const ariaProps = {
        'aria-label': buttonProps['aria-label'],
        'aria-describedby': buttonProps['aria-describedby'],
        'aria-hidden': buttonProps['aria-hidden'],
        'aria-roledescription': buttonProps['aria-roledescription'],
        'aria-expanded': buttonProps['aria-expanded'],
        'aria-pressed': buttonProps['aria-pressed'],
        'aria-checked': buttonProps['aria-checked'],
        'aria-live': buttonProps['aria-live'],
        'aria-disabled': buttonProps['aria-disabled'],
        'aria-haspopup': buttonProps['aria-haspopup'],
        'aria-selected': buttonProps['aria-selected']
    };
    return {
        globalEventProps,
        ariaProps
    };
}
/**
 * Merge buttonset props with toggle props for spreadable properties and flags
 * @param toggleProps
 * @param buttonSetProps
 * @returns
 */
function processSetButtonProps(toggleProps, buttonSetProps) {
    const isDisabled = toggleProps.isDisabled ?? buttonSetProps.isDisabled ? true : undefined;
    const commonProps = {
        size: toggleProps.size ?? buttonSetProps.size,
        isDisabled: isDisabled
    };
    const buttonProps = {
        variant: toggleProps.variant ?? buttonSetProps.variant,
        ...commonProps
    };
    const buttonSetButtonProps = {
        layoutWidth: toggleProps.layoutWidth ?? buttonSetProps.layoutWidth,
        ...buttonProps
    };
    const inputProps = {
        inputType: buttonSetProps.inputType ?? toggleProps.inputType ?? 'checkbox',
        inputName: buttonSetProps.inputName ?? toggleProps.inputName ?? generateId(),
        ...commonProps
    };
    const { globalEventProps } = processGlobals(toggleProps);
    return {
        isDisabled,
        buttonSetButtonProps,
        inputProps,
        globalEventProps
    };
}

export { processSetButtonProps as a, processGlobals as p };
