// @oracle/oraclejet-preact/es/ColorSwatch-52f47316.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { Fragment } from 'preact';
import { useCallback, useState, useEffect } from 'preact/hooks';
import { forwardRef } from 'preact/compat';
import { B as BareButton } from './BareButton-0954f33d.js';
import { I as IconButton } from './IconButton-c91faa06.js';
import { u as useToggleAction } from './useToggleAction-fc0f5399.js';
import { u as useTooltip } from './useTooltip-f442fedb.js';
import { u as useHover } from './useHover-85077adb.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import './LayerHost-12489cef.js';
import { u as useColorScheme } from './useColorScheme-96cce6a1.js';
import { c as classNames } from './classNames-4e12b00d.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-466c63c4.js';
import { u as useActive } from './useActive-fb6d4d52.js';
import { g as getClientHints } from './clientHints-a42b510c.js';
import { ColorSwatchRedwoodTheme } from './UNSAFE_ColorSwatch/themes/redwood/ColorSwatchTheme.js';
import { p as processGlobals } from './buttonUtils-55fb1df7.js';
import { i as isEightDigitHex, p as parseColorValue, g as getSixDigitHex } from './swatchData-64959ad2.js';
import bundle from './resources/nls/en/bundle.js';
import { S as SvgCheckboxCrossed } from './CheckboxCrossed-d1f25f09.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * The NoFill image
 */
const NoFill = () => (jsx("svg", { xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 48 48", preserveAspectRatio: "none", width: "100%", height: "100%", children: jsx("path", { "stroke-width": "2", "vector-effect": "non-scaling-stroke", d: "M0 0L48 48" }) }));

/**
 * ColorSwatches render a color, showing focus, hover, and selection variations.
 */
const ColorSwatch = forwardRef(({ size = 'sm', isSelected = false, isButton = false, onToggle, isDisabled = false, isTooltipDisabled = false, testId, value, ...props }, ref) => {
    const clientHints = getClientHints();
    const isHybrid = clientHints.hoverSupport === 'events';
    const valueText = value;
    const hasValue = valueText && valueText.length;
    const lowerValueText = hasValue ? valueText?.toLocaleLowerCase() : value;
    const isNoFill = hasValue ? lowerValueText === 'transparent' : false;
    const hasAlpha = value && hasValue ? isEightDigitHex(value) : false;
    // const isSmallButton = isButton && size === 'xs'; // if used as xs button, the icon is 20x20, while the target size is 24x24
    const selectedTranslated = bundle.vis_stateSelected();
    const noneTranslated = bundle.labelNone();
    const colorSwatchTranslated = bundle.accColorSwatch();
    const { globalEventProps, ariaProps: { 'aria-label': ariaLabel, 'aria-hidden': ariaHidden, 'aria-roledescription': ariaRoleDescription, ...swatchAriaProps } } = processGlobals(props);
    const { hoverProps, isHover } = useHover({ isDisabled: !isHybrid });
    const { activeProps, isActive } = useActive({ isDisabled: isDisabled });
    // Cope with rgb or palette colors or illegal colors
    const getHexValue = useCallback((raw) => {
        if (raw?.toLowerCase() === 'transparent') {
            return noneTranslated;
        }
        return parseColorValue(raw);
    }, [noneTranslated]);
    const [accessibleLabel, setAccessibleLabel] = useState(ariaLabel
        ? ariaLabel
        : getHexValue(value) + isSelected
            ? ' ' + selectedTranslated
            : '');
    const colorScheme = useColorScheme();
    const { classes, styles: { alphaHalfStyles, swatchStyles, buttonStyles, noFillStyles, noFillDisabledStyles, noFillHoverStyles, smallButtonContainerStyles } } = useComponentTheme(ColorSwatchRedwoodTheme, {
        size: size,
        selected: isSelected ? 'isSelected' : 'notSelected',
        active: isActive ? 'isActive' : 'notActive',
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        hover: isHover ? 'isHover' : 'notHover',
        noFill: isNoFill ? 'isNoFill' : 'notNoFill',
        hasValue: hasValue ? 'hasValue' : 'notHasValue',
        button: isButton ? 'isButton' : 'notButton',
        colorScheme
    });
    useEffect(() => {
        if (!ariaLabel) {
            setAccessibleLabel((isSelected ? selectedTranslated + ' ' : '') + getHexValue(value));
        }
    }, [isSelected, value, getHexValue, selectedTranslated, ariaLabel]);
    const { triggerProps } = useToggleAction({
        onToggle: onToggle
    });
    const { tooltipContent, tooltipProps } = useTooltip({
        text: accessibleLabel,
        isDisabled: isDisabled || isTooltipDisabled
    });
    const elementDetails = {
        type: 'div'
    };
    const splitSwatch = () => {
        return (jsxs(Fragment, { children: [jsx("div", { className: alphaHalfStyles, style: isDisabled ? {} : { backgroundColor: getSixDigitHex(value) } }), jsx("div", { className: alphaHalfStyles, style: isDisabled ? {} : { backgroundColor: value } })] }));
    };
    const noFill = () => {
        return (jsx("div", { className: isDisabled ? noFillDisabledStyles : noFillStyles, children: jsx(NoFill, {}) }));
    };
    const valueSwatch = () => {
        return (jsx("div", { className: classNames([
                smallButtonContainerStyles,
                isButton && buttonStyles,
                !isButton && swatchStyles
            ]), style: isNoFill || isDisabled || hasAlpha ? {} : { backgroundColor: value }, children: isNoFill ? noFill() : hasAlpha ? splitSwatch() : undefined }));
    };
    const noValueSwatch = () => {
        return (jsx("div", { className: classNames([
                !isButton && swatchStyles,
                isDisabled ? noFillDisabledStyles : noFillHoverStyles
            ]), children: jsx(SvgCheckboxCrossed, {}) }));
    };
    const swatchContents = () => {
        return hasValue ? valueSwatch() : noValueSwatch();
    };
    const buttonContents = () => {
        return (jsx(IconButton, { "aria-label": accessibleLabel, "aria-roledescription": ariaRoleDescription ?? colorSwatchTranslated, ...swatchAriaProps, isDisabled: isDisabled, ref: ref, size: "sm", testId: testId, variant: "ghost", onAction: () => onToggle?.({ value: isSelected }), children: swatchContents() }));
    };
    return isButton ? (buttonContents()) : (jsxs(Fragment, { children: [jsx(BareButton, { ref: ref, elementDetails: elementDetails, ...mergeProps(triggerProps, globalEventProps, hoverProps, activeProps, globalEventProps, tooltipProps), testId: testId, isDisabled: isDisabled, "aria-label": accessibleLabel, "aria-roledescription": ariaRoleDescription ?? colorSwatchTranslated, ...swatchAriaProps, classNames: classes, onAction: () => onToggle?.({ value: isSelected }), children: swatchContents() }), tooltipContent] }));
});
ColorSwatch.displayName = 'ColorSwatch';

export { ColorSwatch as C };
