// @oracle/oraclejet-preact/es/NavigationListLinkItem-a5dfc1cc.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs, Fragment } from 'preact/jsx-runtime';
import { S as SvgClose } from './Close-315568df.js';
import { useContext, useCallback, useRef } from 'preact/hooks';
import { u as useHover } from './useHover-85077adb.js';
import { u as useTranslationBundle } from './useTranslationBundle-25469f1c.js';
import { E as EmbeddedIconButton } from './EmbeddedIconButton-5b4a9aea.js';
import { T as Text } from './Text-983c749e.js';
import { u as usePress } from './usePress-051f5adb.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { u as useInteractionStyle } from './useInteractionStyle-99e32c23.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { createContext } from 'preact';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import { f as flexitemInterpolations } from './flexitem-7b7f7920.js';
import { NavigationListItemRedwoodTheme } from './UNSAFE_NavigationListCommon/themes/redwood/NavigationListItemTheme.js';
import { u as useUser } from './useUser-df8ae011.js';
import { F as Flex } from './Flex-5befe411.js';
import { S as SvgWarningS } from './WarningS-7c6d2d0e.js';
import { S as SvgErrorS } from './ErrorS-60a975db.js';
import { S as SvgInformationS } from './InformationS-daba658e.js';
import { S as SvgSuccessS } from './SuccessS-49782304.js';
import { B as Badge } from './Badge-a458c469.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Context used to pass navlist information without having to pass it to navlist children props.
 * We want to communicate information down to any interested navlist item children.
 */
const NavigationListContext = createContext(null);

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Utility hook for consuming the NavigationListContext
 *
 * @returns The value of NavigationListContext provider
 */
function useNavigationListContext() {
    return useContext(NavigationListContext);
}

const flexItemStyleInterpolations = mergeInterpolations([
    ...Object.values(flexitemInterpolations)
]);
const { class: labelContainerFlexItemClasses, ...labelContainerFlexItemStyle } = flexItemStyleInterpolations({
    alignSelf: 'center'
});
function useNavigationListItem({ itemKey }) {
    const { onSelectionChange, selection, currentKey, showFocusRing, navigationListItemPrefix } = useNavigationListContext();
    //Selection and focus state
    const isCurrent = currentKey === itemKey;
    const isSelected = selection === itemKey;
    //Styling
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle, applyActiveStyle } = useInteractionStyle();
    const { direction } = useUser();
    const isRtl = direction === 'rtl';
    const { classes: listItemClasses, styles: { labelContainerClasses } } = useComponentTheme(NavigationListItemRedwoodTheme, {
        focusRing: showFocusRing ? 'isFocusRing' : 'notFocusRing',
        current: isCurrent ? 'isCurrent' : 'notCurrtent',
        selected: isSelected ? 'isSelected' : 'notSelected',
        direction: isRtl ? 'isRtl' : 'notRtl',
        hybridHover: applyHoverStyle ? 'isHybridHover' : 'notHybridHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        active: applyActiveStyle ? 'isActive' : 'notActive'
    });
    //Item handlers
    const handlePress = useCallback(() => {
        onSelectionChange?.({ value: itemKey, reason: 'pointer' });
    }, [itemKey, onSelectionChange]);
    //Legacy navlist makes selection on keyDown, but since preact tabbar made selection after keyUp, will do the same
    //for preact navlist
    const { pressProps } = usePress(handlePress);
    return {
        itemId: navigationListItemPrefix + itemKey,
        itemHandlers: mergeProps(interactionProps, pressProps /*,onFocusProps*/),
        itemClasses: classNames([listItemClasses]),
        isSelected,
        labelContainerClasses: classNames([labelContainerClasses, labelContainerFlexItemClasses]),
        labelContainerStyle: labelContainerFlexItemStyle
    };
}

const onClickLink = (event) => {
    event.preventDefault();
};
/*TODO: JET-58535. Add ripple effect when touching navigationListItem. */
function BaseNavigationListItem({ itemKey, label, badge, metadata, severity = 'none', removeIcon, href, 'aria-controls': ariaControls }) {
    const itemRef = useRef(null);
    //Logic of the component handled here
    const { itemId, itemHandlers, itemClasses, isSelected, labelContainerClasses, labelContainerStyle } = useNavigationListItem({
        itemKey
    });
    const severityRequirement = !(severity === 'none' || severity === undefined);
    const IconComponent = (severityValue) => {
        switch (severityValue) {
            case 'error':
                return jsx(SvgErrorS, { color: "danger" });
            case 'warning':
                return jsx(SvgWarningS, { color: "warning" });
            case 'info':
                return jsx(SvgInformationS, { color: "info" });
            case 'confirmation':
                return jsx(SvgSuccessS, { color: "success" });
            default:
                return null;
        }
    };
    const { styles: { iconLabelContainer, iconContainer, navigationItemMetadata, navigationListItemLink } } = useComponentTheme(NavigationListItemRedwoodTheme);
    const navigationItemMatadata = classNames([navigationItemMetadata]);
    const navigationItemLinkClass = classNames([navigationListItemLink]);
    const endIconContainerClasses = classNames([iconLabelContainer, iconContainer]);
    const { styles: { navigationlistRemoveIcon } } = useComponentTheme(NavigationListItemRedwoodTheme);
    const removeClass = classNames([navigationlistRemoveIcon]);
    const isComplexItem = metadata || severityRequirement || badge;
    const isLinkItem = href != null;
    const labelRenderer = () => {
        return (jsxs(Fragment, { children: [jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsx(Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) }), jsx(Flex, { gap: "2x", align: "center", justify: "end", flex: "1 1 auto", children: removeIcon && !isLinkItem && jsx("span", { class: removeClass, children: removeIcon }) })] }));
    };
    const complexItemRenderer = () => {
        return (jsxs(Fragment, { children: [label && (jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsx(Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) })), (metadata || severityRequirement || badge || removeIcon) && (jsxs(Flex, { gap: "2x", align: "center", justify: "end", flex: "1 1 auto", children: [' ', badge && (jsx("span", { class: endIconContainerClasses, children: jsx(Badge, { size: "md", variant: "neutralSubtle", children: badge }) })), metadata && (jsx("span", { class: navigationItemMatadata, children: jsx(Text, { children: '( ' + metadata + ' )' }) })), severityRequirement && (jsx("span", { class: endIconContainerClasses, children: IconComponent(severity) })), removeIcon && !isLinkItem && jsx("span", { class: removeClass, children: removeIcon })] }))] }));
    };
    const navlistItemRenderer = () => {
        return (jsx("div", { role: "tab", ...itemHandlers, "data-oj-key": itemKey, id: itemId, class: itemClasses, "aria-selected": isSelected, ref: itemRef, "aria-controls": ariaControls, children: isComplexItem ? complexItemRenderer() : labelRenderer() }));
    };
    const navlistLinkItemRenderer = () => {
        return (jsx("div", { ref: itemRef, ...itemHandlers, class: itemClasses, children: jsx("a", { tabindex: -1, role: "tab", "data-oj-key": itemKey, id: itemId, class: navigationItemLinkClass, "aria-selected": isSelected, "aria-controls": ariaControls, href: href, onClick: onClickLink, children: isComplexItem ? complexItemRenderer() : labelRenderer() }) }));
    };
    return isLinkItem ? navlistLinkItemRenderer() : navlistItemRenderer();
}

function RemovableNavigationListItem({ itemKey, label, badge, metadata, severity }) {
    const { selection, currentKey, onRemove } = useNavigationListContext();
    const isSelected = selection === itemKey;
    const isCurrent = currentKey === itemKey;
    const handleRemove = useCallback(() => {
        onRemove?.({ value: itemKey });
    }, [onRemove, itemKey]);
    const { hoverProps, isHover } = useHover({});
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    //instead of specifying flex: '1 0 auto' in cssProps var
    return (jsx("div", { ...hoverProps, children: jsx(BaseNavigationListItem, { itemKey: itemKey, label: label, badge: badge, metadata: metadata, severity: severity, removeIcon: jsx("span", { "data-oj-navigationlist-item-remove-icon": true, children: jsx(EmbeddedIconButton, { "aria-label": translations.tabBarNavigationList_removeCueText(), size: "xs", onAction: handleRemove, children: (isHover || isSelected || isCurrent) && jsx(SvgClose, {}) }) }) }) }));
}

/**
 * NavigationListLinkItem is used to render links in a NavigationList. It allows to view the browser's built in link context menu to open the link in new browser tab.
 */
function NavigationListLinkItem({ itemKey, badge, metadata, severity, href, label, 'aria-controls': ariaControls }) {
    return (jsx(BaseNavigationListItem, { "aria-controls": ariaControls, itemKey: itemKey, label: label, badge: badge, metadata: metadata, severity: severity, href: href }));
}

export { BaseNavigationListItem as B, NavigationListLinkItem as N, RemovableNavigationListItem as R, NavigationListContext as a };
