// @oracle/oraclejet-preact/es/UNSAFE_ComboChart.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, Fragment, jsxs } from 'preact/jsx-runtime';
import { n as getSeriesData, o as getLineAreaSegments, h as getBarGapRatio, i as findNearestDataPoint, m as getMarkers, j as getStartAndEndAxesOffset } from './utils-adfaaca0.js';
import { A as AreaSeries, g as getLineAreaChartItemPosition } from './chartUtils-474e9229.js';
import { L as LineSeries, M as MarkersWithLabel } from './LineSeries-5ee0f790.js';
import { e as getLineAreaChartNavUtil, d as getItemColor, f as getMarkerType } from './dataStyleUtils-a529150d.js';
import { u as useUser } from './useUser-df8ae011.js';
import { g as getColorRamp } from './colorUtils-45d1cac0.js';
import './accUtils-f6657e8e.js';
import { c as colorSchemeVars } from './themeContract.css-2b5a42e2.js';
import { B as BarGroup, g as getBarPosition } from './BarGroup-2c5ca597.js';
import { C as Chart } from './Chart-a834358f.js';
import './utils-cfef1929.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './getLocale-1c41e86a.js';
import './stringUtils-0fe25200.js';
import './labelUtils-aee6ab65.js';
import './textUtils-3ed25dd1.js';
import './utils-49903d04.js';
import './datatipUtils-0a36125e.js';
import './clientUtils-cccff45f.js';
import './clientHints-a42b510c.js';
import './dimensionUtils-6d5dac30.js';
import './scale-9a834124.js';
import './lineUtils-586f3e8a.js';
import './Marker-81aae9f0.js';
import './SvgShapesStyles.styles.css';
import './PRIVATE_Chart/themes/ChartStyles.css.js';
import './ChartStyles.styles.css';
import './classNames-4e12b00d.js';
import 'preact/hooks';
import './LayerHost-12489cef.js';
import './index-3e2e69c4.js';
import 'preact';
import 'preact/compat';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './TrackResizeContainer-4411f745.js';
import './useSize-3b5c56f5.js';
import './useResizeObserver-13641643.js';
import './useAddBusyState-7a6af61f.js';
import './BusyStateContext-9ee9751d.js';
import './useBusyStateContext-4d3c27c0.js';
import './unsafeDomAccess-453eb74e.js';
import './size-a3d3aeaf.js';
import './useTextDimensions-45f79df4.js';
import './useCssVars-143870e2.js';
import './useTheme-d30637db.js';
import './useColorScheme-96cce6a1.js';
import './useScale-0ebe15ed.js';
import './Gridlines-3cc69518.js';
import './PRIVATE_PlotArea/themes/PlotAreaStyles.css.js';
import './PlotAreaStyles.styles.css';
import './hooks/PRIVATE_useVisDnd/themes/VisDndStyles.css.js';
import './VisDndStyles.styles.css';
import './useId-03dbfdf0.js';
import './Axis-5578f8bd.js';
import './PRIVATE_Axis/themes/AxisStyles.css.js';
import './AxisStyles.styles.css';
import './MultilineText-c7d80332.js';
import './PRIVATE_Axis/themes/redwood/AxisTheme.js';
import './PRIVATE_Axis/themes/redwood/AxisBaseTheme.css.js';
import './PRIVATE_Axis/themes/redwood/AxisBaseTheme.styles.css';
import './PRIVATE_Axis/themes/redwood/AxisVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './useComponentTheme-d2f9e47f.js';
import './logger-c92f309c.js';
import './useChartViewport-923b7ae0.js';
import './eventsUtils-e96290df.js';
import './usePinch-301da5f6.js';
import './usePointerGesture-ddf231cc.js';
import './mergeProps-88ea8306.js';
import './useChartNav-ed40e12b.js';
import './layoutUtils-895c6de8.js';
import './useVisHover-d7a0b2f3.js';
import './ScrollBar-c7a5d361.js';
import './PRIVATE_ChartScrollBar/themes/ScrollBarStyles.css.js';
import './ScrollBarStyles.styles.css';
import './useSelectDrill-9027a3a5.js';
import './useChartDataCursor-82b836ed.js';
import './hooks/PRIVATE_useChartDataCursor/themes/DataCursorStyles.css.js';
import './DataCursorStyles.styles.css';
import './useChartDatatip-a3ce4e23.js';
import './useDatatip-c7727336.js';
import './Layer-7d3dc8c7.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useThemeInterpolations-4b48d878.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './mergeInterpolations-30cd9a69.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './useDensity-d06ae84a.js';
import './Floating-6023b41d.js';
import './useFloating-e1167cb9.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-73a51e1d.js';
import './arrayUtils-35a58161.js';
import './useModal-953082f2.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './hooks/PRIVATE_useDatatip/themes/useDatatip.css.js';
import './useDatatip.styles.css';
import './VisTabularDatatip-4d0f026a.js';
import './PRIVATE_VisTabularDatatip/themes/VisTabularDatatipStyles.css.js';
import './VisTabularDatatipStyles.styles.css';
import './useChartDrill-d8a91907.js';
import './useVisDrill-48b6609b.js';
import './useDoubleTap-4e290707.js';
import './clientUtils-71f2dbaf.js';
import './useSelection-8a26bc9e.js';
import './BaseButton-c2fbaa6d.js';
import './useHover-85077adb.js';
import './useToggle-8b7fcefe.js';
import './useActive-fb6d4d52.js';
import './dimensions-ddde0eee.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-0954f33d.js';
import './useBareButton-07a3eef9.js';
import './usePress-051f5adb.js';
import './usePressClick-2ec8e098.js';
import './TabbableModeContext-7d8ad946.js';
import './useTabbableMode-a2fdbb06.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './useTestId-e5d44ab1.js';
import './Text-983c749e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useTranslationBundle-25469f1c.js';
import './hooks/PRIVATE_useVisSelection/themes/ClearSelection.css.js';
import './ClearSelection.styles.css';
import './cssUtils-20795e7b.js';
import './useChartContextMenu-92d9028b.js';
import './useContextMenu-6aa8193e.js';
import './SelectMenuGroupContext-0c1a7603.js';
import './Sheet-24c3e407.js';
import './Modal-678575d5.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './tabbableUtils-ce19b388.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './popupUtils-4edf62b5.js';
import './WindowOverlay-b380ce51.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './UNSAFE_Sheet/themes/SheetStyles.css.js';
import './SheetStyles.styles.css';
import './useAnimationStatus-072d6a98.js';
import './useAnimation-a5eb7604.js';
import './animationUtils-5f409a6c.js';
import './Dropdown-ead9e679.js';
import './keyboardUtils-986fe923.js';
import './usePopupAnimation-66b186ad.js';
import './popupAnimationUtils-175151d9.js';
import './CollectionInteractionContext-7cf19abc.js';
import './UNSAFE_Dropdown/themes/dropdownStyles.css.js';
import './dropdownStyles.styles.css';
import './useOutsideMousedown-39d7c9a6.js';
import './UNSAFE_Menu/themes/MenuStyles.css.js';
import './MenuStyles.styles.css';
import './UNSAFE_Menu/themes/DropdownMenuStyles.css.js';
import './DropdownMenuStyles.styles.css';
import './Flex-5befe411.js';
import './boxalignment-51b42a0f.js';
import './boxalignment.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './flexitem-7b7f7920.js';
import './flexitem.styles.css';
import './FlexStyles.styles.css';
import './Skeleton-a06f4f74.js';
import './borders-98a63040.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './SkeletonStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './UNSAFE_Menu/themes/MenuSkeletonStyles.css.js';
import './MenuSkeletonStyles.styles.css';
import './LiveRegion-79ffc814.js';
import './LiveRegionStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemTheme.js';
import './UNSAFE_Menu/themes/MenuItemStyles.css.js';
import './MenuItemStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js';
import './MenuItemVariants.styles.css';
import './useInteractionStyle-99e32c23.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './tooltipUtils-466c63c4.js';
import './EnvironmentProvider-546b7ebd.js';
import './LayerManager-a62e619b.js';
import './IconStyle.styles.css';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './SeparatorStyles.styles.css';
import './MenuSeparatorStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js';
import './useContextMenuGesture-906cf78d.js';
import './textSelectionUtils-a6554a42.js';
import './useChartDnd-5937eef0.js';
import './utils-43235750.js';
import './useDnd-c5ae190c.js';
import './useTooltip-f442fedb.js';
import './useTooltipControlled-d7724623.js';
import './useFocus-38c95977.js';
import './useTouch-4828df25.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './useChartMarquee-2d60324d.js';
import './useVisDragModeControls-92ddac92.js';
import './Button-c374b855.js';
import './Icon-ef2ac069.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './Move-195dc0bd.js';
import './hooks/PRIVATE_useVisDragModeControls/themes/useVisDragModeControls.css.js';
import './useVisDragModeControls.styles.css';
import './ZoomAlt-7725545a.js';
import './ButtonSetSingle-659479b0.js';
import './useButtonSetContext-6a483d7f.js';
import './ButtonSet-220507f5.js';
import './UNSAFE_ButtonSet/themes/ButtonSetStyles.css.js';
import './ButtonSetStyles.styles.css';
import './ButtonSetPositionContext-5730d367.js';
import './toggleButtonUtils-75e3d781.js';
import './useRovingTabIndexContainer-3903d005.js';
import './ButtonSetItem-6ef8402d.js';
import './ButtonSetButton-a34aeeb0.js';
import './ToggleButtonLabel-802339c7.js';
import './ButtonLabelLayout-d449261a.js';
import './useButtonLabelLayoutTheme-a1b95a52.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './HiddenAccessible-61197e7c.js';
import './HiddenAccessibleStyles.styles.css';
import './ButtonSetIconButton-1beb1197.js';
import './useButtonSetPositionContext-1dc5156a.js';
import './IconToggleButton-2ac4cd72.js';
import './useToggleAction-fc0f5399.js';
import './hooks/PRIVATE_useChartMarquee/themes/useChartMarquee.css.js';
import './useChartMarquee.styles.css';
import './useVisTouchResponse-a59ab166.js';
import './useOverviewContext-79377674.js';
import './OverviewSlidingWindow-36d20536.js';
import './PRIVATE_VisOverview/themes/VisOverview.css.js';
import './VisOverview.styles.css';
import './DragH-bbf55c85.js';

const LINE_WITH_AREA_OPACITY = 0.2;
/**
 * The ComboChart displays information graphically using bars, lines and filled areas, making relationships among the data easier to understand.
 */
function ComboChart({ width, height, selectedIds = [], hiddenIds = [], highlightedIds, hideAndShowBehavior = 'none', orientation = 'vertical', yAxis, series, groups, isStacked = false, getDataItem, selectionMode = 'none', drilling = 'off', dataItemGaps = 0.5, dataLabelEffect = 'outline', ...props }) {
    const { direction, forcedColors } = useUser();
    const isHighContrast = forcedColors === 'active';
    const filteredBarSeries = series.filter((currentSeries) => currentSeries.type === 'bar');
    const isRtl = direction === 'rtl';
    const isLog = yAxis?.scale === 'log';
    const timeAxisType = props.xAxis?.timeAxisType;
    const hiddenSet = new Set(hiddenIds);
    const selectedSet = new Set(selectedIds);
    const highlightedSet = new Set(highlightedIds);
    const isHoriz = orientation === 'horizontal';
    const colors = getColorRamp();
    // TODO: JET-71332 ; should only be (number | undefined)[][]
    // remove the type inference from scatter chart item.
    let renderedSeries = getSeriesData(series.length, groups.length, getDataItem, hiddenSet, isLog);
    const renderedGetDataItem = (seriesIndex, groupIndex) => {
        if (!isStacked)
            return getDataItem(seriesIndex, groupIndex);
        return {
            ...getDataItem(seriesIndex, groupIndex),
            value: renderedSeries[seriesIndex][groupIndex]
        };
    };
    const getGapRatio = () => {
        return 1;
    };
    const getNextChartItemFunc = (startIndex, endIndex) => {
        // TODO: JET-71332
        //@ts-ignore
        return getLineAreaChartNavUtil(getDataItem, series.length, startIndex, endIndex, isStacked, isRtl, isHoriz);
    };
    const isSplitDualY = typeof props.y2Axis?.split === 'number';
    const getDataItemPos = (xScale, yScale, y2Scale, axisStepWidth, averageGroupZ, gapRatio) => {
        const _getBarPosition = hasBar
            ? getBarPosition(filteredBarSeries, groups, getDataItem, xScale, yScale, y2Scale, hiddenSet, dataItemGaps, axisStepWidth, averageGroupZ, gapRatio, yAxis?.scale === 'log', isStacked, isRtl, isHoriz, isSplitDualY, timeAxisType)
            : undefined;
        return (seriesIndex, groupIndex) => {
            const seriesItem = series[seriesIndex];
            if (seriesItem.type === 'bar') {
                return _getBarPosition ? _getBarPosition(seriesIndex, groupIndex) : undefined;
            }
            else {
                return getLineAreaChartItemPosition(seriesIndex, groupIndex, renderedSeries, getDataItem, groups, isHoriz, xScale, yScale, timeAxisType);
            }
        };
    };
    const getMarkerColorAndType = (seriesIndex, groupIndex) => {
        const color = getItemColor(seriesIndex, groupIndex, series, getDataItem);
        const markerType = getMarkerType(seriesIndex, groupIndex, getDataItem);
        return { color, markerType };
    };
    const findNearest = (xScale, yScale, y2Scale) => {
        const getSeriesDataInfo = () => {
            return {
                offset: 0,
                dataWidth: 0
            };
        };
        const getYCoord = (value, yScale) => {
            return yScale.transform(value);
        };
        return findNearestDataPoint(series, groups, xScale, yScale, renderedGetDataItem, isStacked, getSeriesDataInfo, getYCoord, isLog, 0, hiddenSet, timeAxisType, y2Scale);
    };
    const markers = getMarkers();
    const allLineSegmentCoords = [];
    const { hasBar, hasCenteredSeries, hasUncenteredSeries, hasArea } = series.reduce((acc, s) => {
        return {
            hasArea: acc.hasArea || s.type === 'area' || s.type === 'lineWithArea',
            hasBar: acc.hasBar || s.type === 'bar',
            hasCenteredSeries: acc.hasCenteredSeries ||
                s.lineType === 'centeredSegmented' ||
                s.lineType === 'centeredStepped',
            hasUncenteredSeries: acc.hasUncenteredSeries || s.lineType === 'segmented' || s.lineType === 'stepped'
        };
    }, { hasBar: false, hasCenteredSeries: false, hasUncenteredSeries: false, hasArea: false });
    const getOffset = (groupCount) => {
        return getStartAndEndAxesOffset(hasCenteredSeries, hasUncenteredSeries, isHoriz, groupCount, hasBar);
    };
    return (jsx(Chart, { width: width, height: height, selectedIds: selectedSet, hiddenIds: hiddenIds, hideAndShowBehavior: hideAndShowBehavior, orientation: orientation, yAxis: yAxis, series: series, defaultOverviewContentRenderer: () => {
            return jsx(OverviewComboChart, { groups: groups, series: series, getDataItem: getDataItem });
        }, groups: groups, selectionMode: selectionMode, drilling: drilling, isStacked: isStacked, getDataItem: getDataItem, findNearest: findNearest, getDataItemPos: getDataItemPos, getMarkerInfo: getMarkerColorAndType, renderGridLinesInFront: hasArea, getNavUtil: getNextChartItemFunc, getGapRatio: getGapRatio, getStartAndEndAxesOffset: getOffset, isRtl: isRtl, ...props, children: ({ xStartIndex, xEndIndex, xScale, groupsInfo, yScale, y2Scale, activeId, focusedItemInfo, hoveredItemInfo, isPointInsideMarquee, getTextDimensions, defaultFontSize, axisStepWidth, averageGroupZ, getItemAriaLabel, plotAreaClipPathId }) => {
            let barSeriesIndex = -1;
            const leafGroups = groupsInfo.map(({ item }) => item);
            renderedSeries = getSeriesData(series.length, leafGroups.length, getDataItem, hiddenSet, isLog);
            return (jsx(Fragment, { children: renderedSeries.map((valueSeries, index) => {
                    const currentSeries = series[index];
                    const isAssociatedToY2 = currentSeries.associatedYAxis === 'y2';
                    const areaOpacity = currentSeries.type === 'lineWithArea' ? LINE_WITH_AREA_OPACITY : 1;
                    const isAreaRendered = currentSeries.type === 'area' || currentSeries.type === 'lineWithArea';
                    const isContrastLineNeeded = currentSeries.type === 'area';
                    if (valueSeries.length === 0) {
                        return;
                    }
                    const lineSegmentCoords = getLineAreaSegments(isAssociatedToY2 ? y2Scale : yScale, xScale, valueSeries, xStartIndex, xEndIndex, isLog, leafGroups, index, renderedGetDataItem, timeAxisType);
                    allLineSegmentCoords[index] = lineSegmentCoords[0];
                    const isBottomSegmentCurved = currentSeries.type !== 'bar'
                        ? currentSeries.lineType === 'curved' &&
                            isStacked &&
                            renderedSeries.length > 1 &&
                            index > 0
                        : isStacked && renderedSeries.length > 1 && index > 0;
                    const isHighlighted = highlightedIds === undefined ||
                        highlightedSet.size === 0 ||
                        highlightedSet.has(currentSeries.id);
                    const prevCoords = isStacked && index != 0
                        ? allLineSegmentCoords.reduce((prevValue, newValue, j) => {
                            // find the index i such that i < index, and series i  has at least
                            // one or more data item unhidden. Only such series will can be
                            // used as bottomCoord for current series. Otherwise we have to use
                            // baseline coord
                            if (j >= index) {
                                return prevValue;
                            }
                            if (newValue.length > 0) {
                                return j;
                            }
                            return prevValue;
                        }, 0)
                        : undefined;
                    if (currentSeries.type === 'bar')
                        barSeriesIndex = barSeriesIndex + 1;
                    return (jsxs(Fragment, { children: [currentSeries.type === 'bar' &&
                                leafGroups.slice(xStartIndex, xEndIndex + 1).map((_, index) => {
                                    const barGroup = (jsx(BarGroup, { plotAreaClipPathId: plotAreaClipPathId, yScale: yScale, y2Scale: y2Scale, xScale: xScale, splitDualY: isSplitDualY, groupIndex: index + xStartIndex, selectedIds: selectedSet, hiddenIds: hiddenSet, highlightedIds: highlightedIds ? highlightedSet : highlightedIds, isDrillEnabled: drilling === 'on', series: filteredBarSeries, barGapRatio: getBarGapRatio(isStacked, xStartIndex, xEndIndex), dataItemGaps: dataItemGaps, isPointInsideMarquee: isPointInsideMarquee, isStacked: isStacked, axisStepWidth: axisStepWidth, averageGroupZ: averageGroupZ, group: leafGroups[index], timeAxisType: props.xAxis?.timeAxisType, colors: colors, activeId: activeId, isLog: yAxis?.scale === 'log', orientation: orientation, getDataItem: getDataItem, isSelectionEnabled: selectionMode != 'none', isRtl: isRtl, focusedItemInfo: focusedItemInfo, hoveredItemInfo: hoveredItemInfo, getTextDimensions: getTextDimensions, defaultFontSize: defaultFontSize, seriesIndex: barSeriesIndex, getItemAriaLabel: getItemAriaLabel }));
                                    return barGroup;
                                }), isAreaRendered && (jsx(AreaSeries, { yScale: isAssociatedToY2 ? y2Scale : yScale, plotAreaClipPathId: plotAreaClipPathId, lineType: currentSeries.lineType, color: currentSeries.areaColor || colors[index % colors.length], isHorizontal: isHoriz, areaColorOpacity: areaOpacity, isBottomSegmentCurved: isBottomSegmentCurved, bottomCoords: prevCoords !== undefined &&
                                    allLineSegmentCoords[prevCoords].some((value) => value != null)
                                    ? allLineSegmentCoords[prevCoords]
                                    : undefined, lineSegments: lineSegmentCoords, isLog: isLog, isHighlighted: isHighlighted, seriesIndex: index, groupWidth: axisStepWidth, isRtl: isRtl, timeAxisType: timeAxisType })), currentSeries.type === 'line' && (jsx(LineSeries, { plotAreaClipPathId: plotAreaClipPathId, lineType: currentSeries.lineType, lineColor: isContrastLineNeeded
                                    ? colorSchemeVars.dvt.contrastLine
                                    : currentSeries.lineColor, lineWidth: isContrastLineNeeded ? 1.25 : currentSeries.lineWidth, lineStyle: currentSeries.lineStyle || 'solid', color: currentSeries.lineColor || colors[index % colors.length], lineSegmentCoords: lineSegmentCoords, isHorizontal: isHoriz, isHighlighted: isHighlighted, seriesIndex: index, groupWidth: axisStepWidth, isRtl: isRtl, timeAxisType: timeAxisType })), 
                            /*for non stacked charts, markers should render in the order of series.
                             some markers might get hidden behind area but this preserves the legacy behavior. */
                            !isStacked && (currentSeries.type === 'line' || isAreaRendered) && (jsx(MarkersWithLabel, { yScale: isAssociatedToY2 ? y2Scale : yScale, xScale: xScale, groups: leafGroups, series: currentSeries, timeAxisType: timeAxisType, color: colors[index % colors.length] /* TODO : make this use getItemColor function from dataStyleUtils. */, seriesIndex: index, startIndex: xStartIndex, endIndex: xEndIndex, markerType: markers[index % markers.length], orientation: orientation, isLog: isLog, activeId: activeId, getDataItem: getDataItem, isPointInsideMarquee: isPointInsideMarquee, selectedIds: selectedSet, focusedItemIndex: focusedItemInfo.seriesIndex === index &&
                                    focusedItemInfo.isFocusVisible &&
                                    !(props.isDataCursorEnabled && props.dataCursorStyle?.isMarkerDisplayed)
                                    ? focusedItemInfo.groupIndex
                                    : undefined, hoveredItemIndex: hoveredItemInfo?.isCurrent && hoveredItemInfo.seriesIndex === index
                                    ? hoveredItemInfo.groupIndex
                                    : undefined, isSelectionEnabled: selectionMode === 'single' || selectionMode === 'multiple', isDrillEnabled: drilling === 'on', defaultFontSize: defaultFontSize, dataLabelEffect: dataLabelEffect, isHighContrast: isHighContrast, getItemAriaLabel: getItemAriaLabel }))] }));
                }) })
            //TODO - handle isStack part
            );
        } }));
}
function OverviewComboChart(props) {
    return (jsx(ComboChart, { datatip: () => ({ content: '' }), zoomAndScroll: "off", plotArea: {
            yMajorTick: {
                isRendered: false
            },
            xMajorTick: {}
        }, xAxis: {
            tickLabel: {
                autoRotate: false,
                // @ts-ignore
                isRendered: props.xAxis?.timeAxisType ? true : false
            }
        }, series: props.series, groups: props.groups, getDataItem: props.getDataItem, yAxis: {
            tickLabel: {
                isRendered: false
            }
        }, overview: "off" }));
}

export { ComboChart };
