// @oracle/oraclejet-preact/es/UNSAFE_FlatTreeView.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs, Fragment } from 'preact/jsx-runtime';
import { useState, useEffect, useRef, useCallback, useMemo } from 'preact/hooks';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { k as keyExtractor, I as ITEM_SELECTOR, f as findElementByKey, m as getPrevNextKey, j as getViewportConfig } from './collectionUtils-893f2808.js';
import { L as LOADMORE_STYLE_CLASS, a as LoadMoreCollection } from './LoadMoreCollection-02671d0d.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { T as TabbableModeContext } from './TabbableModeContext-7d8ad946.js';
import { u as useId } from './useId-03dbfdf0.js';
import { V as View } from './View-ef109be6.js';
import { styles } from './UNSAFE_FlatTreeView/themes/TreeViewStyles.css.js';
import { treeViewVars } from './UNSAFE_FlatTreeView/themes/TreeViewContract.css.js';
import { F as Flex } from './Flex-5befe411.js';
import { S as Skeleton } from './Skeleton-a06f4f74.js';
import { u as useHover } from './useHover-85077adb.js';
import { g as getClientHints } from './clientHints-a42b510c.js';
import { E as Expander } from './Expander-cf44951b.js';
import { TreeViewRedwoodTheme } from './UNSAFE_FlatTreeView/themes/redwood/TreeViewTheme.css.js';
import { u as useSelection } from './useSelection-c405739d.js';
import { u as useAdditiveSelection } from './useAdditiveSelection-dc54508b.js';
import './ReorderableContext-0daeea89.js';
import { i as isKeyDefined, c as containsKey, b as addKeys } from './keys-49c92dbb.js';
import './LayerHost-12489cef.js';
import 'preact/compat';
import { u as useExpandCollapse } from './useExpandCollapse-e68ba8a1.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useUser } from './useUser-df8ae011.js';
import { a as useNavigation } from './useNavigation-cd59ca08.js';
import { S as Selector } from './Selector-1ceeaf0a.js';
import { L as LiveRegion } from './LiveRegion-79ffc814.js';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { u as useTranslationBundle } from './useTranslationBundle-25469f1c.js';
import 'preact';
import './Collection-c4f2f3ea.js';
import './useViewportIntersect-8a2b324b.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './borders-98a63040.js';
import './dimensions-ddde0eee.js';
import './size-a3d3aeaf.js';
import './utils-cfef1929.js';
import './themeContract.css-2b5a42e2.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './flexitem-7b7f7920.js';
import './flexitem.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './colors-b6530fec.js';
import './padding-9b3afaa2.js';
import './arrayUtils-35a58161.js';
import './aria-efa0582c.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './mergeInterpolations-30cd9a69.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './EnvironmentProvider-546b7ebd.js';
import './LayerManager-a62e619b.js';
import './useDensity-d06ae84a.js';
import './useTestId-e5d44ab1.js';
import './equals-625352e8.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './TreeViewStyles.styles.css';
import './boxalignment-51b42a0f.js';
import './boxalignment.styles.css';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './FlexStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './SkeletonStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './useToggle-8b7fcefe.js';
import './ChevronDown-42f09951.js';
import './Icon-ef2ac069.js';
import './useTooltip-f442fedb.js';
import './useTooltipControlled-d7724623.js';
import './useFocus-38c95977.js';
import './useTouch-4828df25.js';
import './useAnimation-a5eb7604.js';
import './useThemeInterpolations-4b48d878.js';
import './useColorScheme-96cce6a1.js';
import './useScale-0ebe15ed.js';
import './unsafeDomAccess-453eb74e.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './tooltipUtils-466c63c4.js';
import './Floating-6023b41d.js';
import './useFloating-e1167cb9.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-73a51e1d.js';
import './Layer-7d3dc8c7.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useModal-953082f2.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './logger-c92f309c.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './IconStyle.styles.css';
import './ChevronRight-e0152ded.js';
import './EmbeddedIconButton-5b4a9aea.js';
import './BaseButton-c2fbaa6d.js';
import './useActive-fb6d4d52.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-0954f33d.js';
import './useBareButton-07a3eef9.js';
import './usePress-051f5adb.js';
import './usePressClick-2ec8e098.js';
import './useTabbableMode-a2fdbb06.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-d449261a.js';
import './Text-983c749e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useButtonLabelLayoutTheme-a1b95a52.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './UNSAFE_FlatTreeView/themes/redwood/TreeViewBaseTheme.styles.css';
import './textSelectionUtils-a6554a42.js';
import './index-3e2e69c4.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './useCurrentKey-7c652655.js';
import './useReorderableContext-79b1c941.js';
import './useCollectionFocusRing-71cdd52c.js';
import './useTabbableModeSet-274065a0.js';
import './UNSAFE_Selector/themes/SelectorStyles.css.js';
import './SelectorStyles.styles.css';
import './StyledCheckbox-bf4ac68c.js';
import './Check-67a66daf.js';
import './CheckboxOff-79645ca6.js';
import './CheckboxOn-cf142463.js';
import './CheckboxMixed-56054e95.js';
import './UNSAFE_Checkbox/themes/CheckboxIconContract.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js';
import './UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js';
import './CheckboxIconStyles.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js';
import './CheckboxIconVariants.styles.css';
import './StyledCheckbox.styles.css';
import './CollectionInteractionContext-7cf19abc.js';
import './useCollectionInteractionContext-3895c632.js';
import './LiveRegionStyles.styles.css';
import './HiddenAccessibleStyles.styles.css';

/**
 * Allows to specify the time delay for rendering the component
 **/
const timerValue$1 = 50;
/**
 * Represents the skeleton for a tree item in expanding state
 **/
function NodeSkeletonContainer({ children, indentation = 0 }) {
    const [isVisible, setIsVisible] = useState(false);
    useEffect(() => {
        setTimeout(() => {
            setIsVisible(true);
        }, timerValue$1);
    }, []);
    return isVisible && children ? (jsx("div", { role: "presentation", children: jsx(View, { paddingInlineStart: indentation, children: [...Array(1)].map((_element, index) => children(index)) }) })) : null;
}

function TreeItemSkeleton({ indentation = 0, isLeaf = false }) {
    return (jsx(View, { paddingInlineStart: calcIndentation(indentation, isLeaf ? undefined : true), children: jsx(View, { padding: [0, '4px'], children: jsxs(Flex, { height: "9x", align: "center", gap: "2x", children: [!isLeaf && jsx(Skeleton, { width: "6x", height: "6x", borderRadius: "4px" }), jsx(Skeleton, { height: "5x", borderRadius: "4px" })] }) }) }));
}
function TreeItemSkeletons() {
    return (jsxs(Fragment, { children: [jsx(TreeItemSkeleton, {}), jsx(TreeItemSkeleton, { indentation: 1 }), jsx(TreeItemSkeleton, { indentation: 2, isLeaf: true })] }));
}

function TreeItem({ itemKey, depth, context, children, isExpanded, onToggle, selected, isSelectable, isCheckboxless, isTabbable, isLoading, isFocusRingVisible }) {
    const rootRef = useRef(null);
    const indentation = calcIndentation(depth, isExpanded);
    const skeletonIndentation = calcIndentation(depth + 1, isExpanded);
    const { isHover, hoverProps } = useHover();
    const classes = classNames([
        styles.treeItemStyles.base,
        isFocusRingVisible && styles.treeItemStyles.focused,
        selected === true && styles.treeItemStyles.selected,
        selected === true && isCheckboxless && styles.treeItemStyles.checkboxlessSelected,
        isSelectable && !selected && isHover && styles.treeItemStyles.hovered
    ]);
    return (jsxs(Fragment, { children: [jsx("div", { ref: rootRef, id: useId(), tabIndex: isMobile() ? -1 : undefined, "data-oj-key": itemKey, class: classes, role: "treeitem", "aria-level": depth + 1, "aria-selected": selected ? 'true' : 'false', ...(isExpanded !== undefined && { 'aria-expanded': isExpanded }), ...(typeof itemKey === 'number' && { 'data-oj-key-type': 'number' }), ...hoverProps, children: jsxs(Fragment, { children: [selected === true && isHover && jsx("div", { class: styles.selectedHoveredOverlay }), jsx(TabbableModeContext.Provider, { value: { isTabbable }, children: jsx(View, { paddingInlineStart: indentation, children: jsxs(Flex, { align: "center", children: [isExpanded !== undefined && (jsx(Expander, { itemKey: itemKey, expanded: isExpanded, onToggle: onToggle })), children(context)] }) }) })] }) }), isLoading ? (jsx(NodeSkeletonContainer, { indentation: skeletonIndentation, children: () => {
                    return jsx(TreeItemSkeletons, {});
                } })) : null] }));
}
const isMobile = () => {
    const { platform } = getClientHints();
    return platform === 'ios' || platform === 'android';
};
const calcIndentation = (depth, isExpanded) => {
    const expanderPlaceholder = isExpanded === undefined ? '40px' : '0px';
    return `calc(${Math.max(0, depth)} * ${treeViewVars.indentWidth} + ${expanderPlaceholder})`;
};

/**
 * Allows to specify the time delay for rendering the component
 **/
const timerValue = 50;
/**
 * SkeletonContainer renders 'minimumCount' number of skeletons
 * of the variant specified from its child element - Skeleton's prop after
 * 'timerValue' ms delay
 **/
function SkeletonContainer({ children, minimumCount = 1, onSkeletonVisible }) {
    const [isVisible, setIsVisible] = useState(false);
    useEffect(() => {
        setTimeout(() => {
            setIsVisible(true);
            onSkeletonVisible?.();
        }, timerValue);
    }, [onSkeletonVisible]);
    const containerClasses = classNames([styles.skeletonStyles.container]);
    return isVisible && children ? (jsx("div", { class: containerClasses, role: "presentation", children: [...Array(minimumCount)].map((_element, index) => children(index)) })) : null;
}

/**
 * Component that receives the flattened data and renders the hierarchy betweens items based on the shaped metadata.
 * In order to maximize performance, only items that are visible in the viewport are rendered.
 * This is an internal component that is used by FlatTreeView.
 */
function LoadMoreTreeCollection({ data, children, selectionMode, onToggle = () => { }, viewportConfig, hasMore = false, onLoadMore = () => { }, loadMoreIndicator = defaultLoadMoreIndicator, updateContext }) {
    const createTreeItem = (context) => {
        const key = context.data.metadata.key;
        const treeItemContext = {
            index: context.index,
            data: context.data.data,
            metadata: { key }
        };
        // allow container to inject additional things to context
        if (updateContext) {
            updateContext(treeItemContext);
        }
        const expanded = context.data.metadata.expanded;
        return (jsx(TreeItem, { itemKey: treeItemContext.metadata.key, context: treeItemContext, depth: context.data.metadata.treeDepth, isExpanded: expanded === 'ongoing' ? true : expanded, onToggle: onToggle, isSelectable: selectionMode !== 'none', isCheckboxless: selectionMode === 'single', selected: context.data.metadata.selected, isLoading: expanded === 'ongoing', isTabbable: false, isFocusRingVisible: context.data.metadata.hasFocus, children: children }, treeItemContext.metadata.key));
    };
    return (jsx(Fragment, { children: data === null ? null : (jsx(LoadMoreCollection, { data: data, viewportConfig: viewportConfig, hasMore: hasMore, onLoadMore: onLoadMore, loadMoreIndicator: loadMoreIndicator, children: createTreeItem })) }));
}
const defaultLoadMoreIndicator = (jsx("div", { class: LOADMORE_STYLE_CLASS, children: jsx(SkeletonContainer, { minimumCount: 1, children: () => {
            return jsx(TreeItemSkeletons, {});
        } }) }));

function useTreeNavigation(data, rootRef, isExpanded, idExtractor) {
    const { direction } = useUser();
    const arrowKeyAndHandlers = {
        ArrowUp: (key) => getPrevNextKey(rootRef.current, key, true, ITEM_SELECTOR),
        ArrowDown: (key) => getPrevNextKey(rootRef.current, key, false, ITEM_SELECTOR)
    };
    const { navigationProps, currentKey, onCurrentKeyChange, showFocusRing } = useNavigation(data, rootRef, idExtractor, arrowKeyAndHandlers, undefined, // currentItemOverride,
    undefined, //onPersistCurrentItem,
    undefined, undefined, true, // should always allow shift key
    false, // TODO: update allowTabbableMode when F2 mode is added
    true, false);
    const itemKeyExtractor = useCallback((element) => {
        if (element.getAttribute('role') === 'tree') {
            return currentKey === undefined ? null : currentKey;
        }
        return keyExtractor(element, ITEM_SELECTOR);
    }, [currentKey, keyExtractor]);
    const onKeyUp = useCallback((event) => {
        if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
            const isRtl = direction === 'rtl';
            const isLeftArrow = event.key === 'ArrowLeft' && !isRtl;
            const itemKey = itemKeyExtractor(event.target);
            if (itemKey) {
                const contains = isExpanded(itemKey);
                // arrow left on a collapsed node, go to its parent
                if (isLeftArrow && !contains && rootRef.current) {
                    const item = findElementByKey(rootRef.current, itemKey, ITEM_SELECTOR);
                    if (item) {
                        const treeDepth = item.getAttribute('aria-level');
                        const parent = treeDepth !== null &&
                            findParentElementByKey(rootRef.current, itemKey, Number(treeDepth), ITEM_SELECTOR);
                        const newKey = parent && itemKeyExtractor(parent);
                        if (isKeyDefined(newKey)) {
                            onCurrentKeyChange?.({ value: newKey });
                        }
                    }
                }
                else if (!isLeftArrow && contains) {
                    // arrow right on an expanded node, go to its first child
                    const newKey = getPrevNextKey(rootRef.current, itemKey, false, ITEM_SELECTOR)();
                    if (isKeyDefined(newKey)) {
                        onCurrentKeyChange?.({ value: newKey });
                    }
                }
            }
        }
    }, [direction, isExpanded, onCurrentKeyChange, itemKeyExtractor]);
    const treeNavigationProps = mergeProps(navigationProps, { onKeyUp });
    return { treeNavigationProps, currentKey, showFocusRing };
}
const findParentElementByKey = (root, key, treeDepth, itemSelector) => {
    if (root && key !== null) {
        const items = Array.from(root.querySelectorAll(itemSelector));
        const index = items.findIndex((i) => i.dataset['ojKey'] == key);
        for (let i = index; i >= 0; i--) {
            const level = items[i].getAttribute('aria-level');
            if (level !== null && Number(level) == treeDepth - 1) {
                return items[i];
            }
        }
    }
    return null;
};

/**
 * A FlatTreeView displays flattened data as a tree with highly interactive features.
 */
function FlatTreeView({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, data, selectionMode = 'none', selectedKeys = { all: false, keys: new Set() }, onSelectionChange, onToggle, viewportConfig, ...rest }) {
    const rootRef = useRef(null);
    const [liveRegion, setLiveRegion] = useState();
    const { collection_loading, collection_loaded } = useTranslationBundle('@oracle/oraclejet-preact');
    const loadingDescId = useId();
    const isExpanded = useCallback((key) => {
        if (rootRef.current) {
            const elem = findElementByKey(rootRef.current, key, ITEM_SELECTOR);
            return elem?.getAttribute('aria-expanded') === 'true';
        }
        return false;
    }, []);
    const { treeNavigationProps, currentKey, showFocusRing } = useTreeNavigation(data, rootRef, isExpanded, idExtractor);
    const items = useMemo(() => data?.map((value) => {
        const itemKey = value.metadata.key;
        const newValue = { ...value };
        newValue.metadata.selected = containsKey(selectedKeys, itemKey);
        newValue.metadata.hasFocus = currentKey === itemKey && showFocusRing;
        return newValue;
    }) || [], [data, selectedKeys, currentKey, showFocusRing]);
    const expandCollapseProps = useExpandCollapse((element) => {
        if (element.getAttribute('role') === 'tree') {
            return currentKey === undefined ? null : currentKey;
        }
        return keyExtractor(element, ITEM_SELECTOR);
    }, isExpanded, onToggle, true);
    // tracking the anchor key which is used for shift+click selection
    const anchorKey = useRef();
    const handleSelectionChange = useCallback((detail) => {
        if (onSelectionChange) {
            if (detail.value.all === false && detail.value.keys.size > 0) {
                // during regular selection, the last key selected by a user gesture is the anchor key
                anchorKey.current = Array.from(detail.value.keys.values()).pop();
            }
            onSelectionChange(detail);
        }
    }, [anchorKey, onSelectionChange]);
    const handleSelectionRangeChange = useCallback((detail) => {
        if (data && onSelectionChange) {
            const value = handleSelectionRange(detail, data);
            if (Array.isArray(value)) {
                if (detail.eventType === 'shiftSpace') {
                    // during range selection by shift+space, the first key selected by a user gesture is the anchor key
                    anchorKey.current = detail.value.start;
                }
                else {
                    // during range selection by shift+click/arrow keys, the last key selected by a user gesture is the anchor key
                    anchorKey.current = detail.value.end;
                }
                if (selectionMode === 'multipleToggle') {
                    const keys = addKeys(selectedKeys, value);
                    onSelectionChange({ value: keys, target: null });
                }
                else {
                    onSelectionChange({ value: { all: false, keys: new Set(value) }, target: null });
                }
            }
        }
    }, [data, selectedKeys, onSelectionChange]);
    // single or multiple selection mode
    const { selectionProps } = useSelection((element) => {
        if (element === rootRef.current) {
            return currentKey === undefined ? null : currentKey;
        }
        return keyExtractor(element, ITEM_SELECTOR);
    }, selectedKeys, selectionMode === 'multipleToggle' ? 'none' : selectionMode, false, 'replace', handleSelectionChange, anchorKey.current, currentKey, (currentKey, isPrev) => getPrevNextKey(rootRef.current, currentKey, isPrev, ITEM_SELECTOR), undefined, (key) => {
        if (rootRef.current) {
            const elem = findElementByKey(rootRef.current, key, ITEM_SELECTOR);
            if (elem) {
                elem.scrollIntoView({ block: 'nearest' });
            }
        }
    }, handleSelectionRangeChange);
    // multipleToggle selection mode
    const { eventProps: multipleToggleSelectionProps } = useAdditiveSelection((element) => {
        if (element === rootRef.current) {
            return currentKey === undefined ? null : currentKey;
        }
        return keyExtractor(element, ITEM_SELECTOR);
    }, selectedKeys, selectionMode === 'multipleToggle' ? 'multiple' : 'none', false, (key) => {
        if (data) {
            return data.findIndex((value) => value.metadata.key === key);
        }
        return -1;
    }, (startIndex, endIndex) => {
        if (data) {
            const keys = data.map((value) => value.metadata.key);
            return keys.slice(startIndex, endIndex);
        }
        return [];
    }, (currentKey, isPrev) => getPrevNextKey(rootRef.current, currentKey, isPrev, ITEM_SELECTOR), currentKey, handleSelectionChange, handleSelectionRangeChange);
    //Note: onKeyUp doesn't work with the cmd key
    const selectAllWithKeyboard = useCallback((event) => {
        if (selectionMode === 'multiple' || selectionMode === 'multipleToggle') {
            if (onSelectionChange && (event.metaKey || event.ctrlKey) && event.key === 'a') {
                event.preventDefault();
                onSelectionChange({
                    value: { all: true, deletedKeys: new Set() },
                    target: event.target
                });
            }
        }
    }, [onSelectionChange, selectionMode]);
    const { classes } = useComponentTheme(TreeViewRedwoodTheme);
    viewportConfig = getViewportConfig(rootRef, viewportConfig);
    const updateContextCallback = useCallback((context) => {
        const key = context.metadata.key;
        context.selector =
            selectionMode === 'none' || selectionMode === 'single'
                ? undefined
                : (ariaLabel) => (jsx("div", { "aria-hidden": isMobile() ? undefined : true, children: jsx(Selector, { "aria-label": isMobile() ? ariaLabel : undefined, rowKey: key, selectedKeys: selectedKeys, onChange: onSelectionChange }) }));
    }, [selectionMode, selectedKeys, onSelectionChange]);
    const updateLiveRegion = useCallback(() => {
        setLiveRegion(collection_loading());
    }, [collection_loading]);
    return (jsxs(Fragment, { children: [jsx("div", { "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": data == null
                    ? ariaDescribedBy
                        ? [...new Set(ariaDescribedBy.split(/\s+/).concat(loadingDescId))].join(' ')
                        : loadingDescId
                    : ariaDescribedBy, ref: rootRef, role: "tree", "aria-multiselectable": selectionMode === 'multiple' ? true : false, tabIndex: 0, class: classes, ...mergeProps(treeNavigationProps, selectionProps, multipleToggleSelectionProps, expandCollapseProps, { onKeyDown: selectAllWithKeyboard }), children: data == null ? (defaultLoadingIndicator(updateLiveRegion)) : (jsx(LoadMoreTreeCollection, { data: items, selectionMode: selectionMode, onToggle: onToggle, viewportConfig: viewportConfig, updateContext: updateContextCallback, ...rest })) }), jsx(HiddenAccessible, { id: loadingDescId, children: collection_loading() }), jsx(LiveRegion, { children: data == null ? liveRegion : liveRegion === collection_loading() ? collection_loaded() : '' })] }));
}
const defaultLoadingIndicator = (callback) => {
    return (jsx(SkeletonContainer, { minimumCount: 8, onSkeletonVisible: callback, children: () => {
            return jsx(TreeItemSkeletons, {});
        } }));
};
const idExtractor = (elem) => elem.id;
/**
 * A helper function to handle range selection
 */
const handleSelectionRange = (detail, data) => {
    const keys = data.map((value) => {
        return value.metadata.key;
    });
    const startIndex = keys.indexOf(detail.value.start);
    const endIndex = keys.indexOf(detail.value.end);
    const minIndex = Math.min(startIndex, endIndex);
    const maxIndex = Math.max(startIndex, endIndex);
    if (minIndex === -1) {
        // return a range to fetch so we can find all the keys, maxIndex should not be -1
        return { offset: 0, count: -1 };
    }
    else {
        // range of keys are in the current viewport, return them
        return keys.slice(minIndex, maxIndex + 1);
    }
};

export { FlatTreeView };
