// @oracle/oraclejet-preact/es/UNSAFE_ReorderableTabBar.js
/* @oracle/oraclejet-preact: undefined */
import { jsxs, Fragment, jsx } from 'preact/jsx-runtime';
import { toChildArray } from 'preact';
import { forwardRef } from 'preact/compat';
import { useRef, useCallback, useState } from 'preact/hooks';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { L as LiveRegion } from './LiveRegion-79ffc814.js';
import { T as TabBar } from './TabBar-0de29233.js';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import { f as flexitemInterpolations } from './flexitem-7b7f7920.js';
import { a as getKey } from './collectionUtils-893f2808.js';
import './LayerHost-12489cef.js';
import { a as useReorderableItem, u as useReorderable } from './useReorderableItem-26d8f85b.js';
import { R as ReorderableContext } from './ReorderableContext-0daeea89.js';
import { styles } from './UNSAFE_ReorderableTabBar/themes/ReorderableTabBarItemStyles.css.js';
import { i as isKeyDefined } from './keys-49c92dbb.js';
import './DragAndDropStyles.styles.css';
import './ReorderableItemStyles.styles.css';
import './HiddenAccessibleStyles.styles.css';
import './LiveRegionStyles.styles.css';
import './useTestId-e5d44ab1.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './Flex-5befe411.js';
import './classNames-4e12b00d.js';
import './boxalignment-51b42a0f.js';
import './size-a3d3aeaf.js';
import './utils-cfef1929.js';
import './themeContract.css-2b5a42e2.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './arrayUtils-35a58161.js';
import './boxalignment.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './dimensions-ddde0eee.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './FlexStyles.styles.css';
import './TabBarLayout-e6b564fc.js';
import './logger-c92f309c.js';
import './TabBarItemStyles.styles.css';
import './UNSAFE_TabBarCommon/themes/redwood/TabBarItemBaseTheme.styles.css';
import './UNSAFE_TabBarCommon/themes/redwood/TabBarItemVariants.css.js';
import './TabBarItemVariants.styles.css';
import './ConveyorBeltContext-76a29c59.js';
import './BusyStateContext-9ee9751d.js';
import './TabbableModeContext-7d8ad946.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './tooltipUtils-466c63c4.js';
import './EnvironmentProvider-546b7ebd.js';
import './LayerManager-a62e619b.js';
import './Floating-6023b41d.js';
import './useFloating-e1167cb9.js';
import './useUser-df8ae011.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-73a51e1d.js';
import './Layer-7d3dc8c7.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useThemeInterpolations-4b48d878.js';
import './useColorScheme-96cce6a1.js';
import './useScale-0ebe15ed.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './useDensity-d06ae84a.js';
import './unsafeDomAccess-453eb74e.js';
import './useModal-953082f2.js';
import './useComponentTheme-d2f9e47f.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './IconStyle.styles.css';
import './keyboardUtils-986fe923.js';
import './clientHints-a42b510c.js';
import './tabbableUtils-ce19b388.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './CollectionInteractionContext-7cf19abc.js';
import './dropdownStyles.styles.css';
import './Sheet-24c3e407.js';
import './Modal-678575d5.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-4edf62b5.js';
import './WindowOverlay-b380ce51.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './UNSAFE_Sheet/themes/SheetStyles.css.js';
import './SheetStyles.styles.css';
import './useAnimationStatus-072d6a98.js';
import './useAnimation-a5eb7604.js';
import './animationUtils-5f409a6c.js';
import './useTranslationBundle-25469f1c.js';
import './OverflowTabBarItemStyles.styles.css';
import './UNSAFE_TabBar/themes/TabBarStyles.css.js';
import './TabBarStyles.styles.css';
import './useTabBar-816d84de.js';
import './mergeProps-88ea8306.js';
import './useCollectionFocusRing-71cdd52c.js';
import './useCurrentKey-7c652655.js';
import './useId-03dbfdf0.js';
import './useReorderableContext-79b1c941.js';
import './useContextMenuGesture-906cf78d.js';
import './textSelectionUtils-a6554a42.js';
import './TabBarContextMenu-47631108.js';
import './SelectMenuGroupContext-0c1a7603.js';
import './Dropdown-ead9e679.js';
import './usePopupAnimation-66b186ad.js';
import './popupAnimationUtils-175151d9.js';
import './UNSAFE_Dropdown/themes/dropdownStyles.css.js';
import './useOutsideMousedown-39d7c9a6.js';
import './UNSAFE_Menu/themes/MenuStyles.css.js';
import './MenuStyles.styles.css';
import './UNSAFE_Menu/themes/DropdownMenuStyles.css.js';
import './DropdownMenuStyles.styles.css';
import './Skeleton-a06f4f74.js';
import './borders-98a63040.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './SkeletonStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './UNSAFE_Menu/themes/MenuSkeletonStyles.css.js';
import './MenuSkeletonStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemTheme.js';
import './UNSAFE_Menu/themes/MenuItemStyles.css.js';
import './MenuItemStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js';
import './MenuItemVariants.styles.css';
import './Text-983c749e.js';
import './usePress-051f5adb.js';
import './useInteractionStyle-99e32c23.js';
import './useHover-85077adb.js';
import './useToggle-8b7fcefe.js';
import './useActive-fb6d4d52.js';
import './Separator-93c70737.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './SeparatorStyles.styles.css';
import './MenuSeparatorStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js';
import './tabBarUtils-9483d091.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_isObject-1fab0f5b.js';
import './flexitem.styles.css';
import './index-3e2e69c4.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './dndUtils-deb68797.js';
import './hooks/PRIVATE_useReorderable/themes/ReorderableItemStyles.css.js';
import './ReorderableTabBarItemStyles.styles.css';

const ITEM_SELECTOR = '[role="tab"]';
/**
 * A private item that will be used by ReorderableTabBar, location TBD (could be moved to PRIVATE_TabBar)
 */
function ReorderableTabBarItem({ children, dragKey, setDragKey, onReorder, layout, rootRef }) {
    const tabBarItemRef = useRef(null);
    const getItem = useCallback(() => tabBarItemRef?.current?.querySelector(ITEM_SELECTOR), []);
    const setTransferData = useCallback((event) => {
        event.dataTransfer?.setData('text/tabbar-item-key', getKey(getItem()) + '');
    }, [getItem]);
    const { rootProps, isDragged, overlayZoneBeforeProps, overlayZoneAfterProps } = useReorderableItem({
        isDisabled: onReorder ? false : true,
        dragKey,
        setDragKey,
        onReorder,
        rootRef,
        ref: tabBarItemRef,
        getItem,
        setTransferData,
        itemStyles: styles,
        itemSelector: ITEM_SELECTOR
    });
    //instead of specifying flex: '1 0 auto' in style
    const styleInterpolations = mergeInterpolations([
        ...Object.values(flexitemInterpolations)
    ]);
    const flexDimensions = styleInterpolations({
        flex: layout === 'stretch' ? '1 0 auto' : '0 0 auto'
    });
    return (jsxs("div", { ...rootProps, draggable: onReorder ? true : false, style: flexDimensions, children: [children, !isDragged && isKeyDefined(dragKey) && (jsxs(Fragment, { children: [jsx("div", { ...overlayZoneBeforeProps }), jsx("div", { ...overlayZoneAfterProps })] }))] }));
}

/**
 * A component that allows reordering of tabs within a TabBar.
 */
const ReorderableTabBar = forwardRef(({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledby, children, layout = 'stretch', display = 'standard', size = 'md', edge = 'top', selection, onSelect, onRemove, onReorder, contextMenuConfig, testId }, ref) => {
    const rootRef = useRef(null);
    const [dragKey, setDragKey] = useState(undefined);
    const displayStackProps = {
        display: display
    };
    const displayNonStackProps = {
        display: display,
        size: size
    };
    const displayProps = display === 'stacked' ? displayStackProps : displayNonStackProps;
    const { reorderProps, reorderContext, reorderInstructionsId, reorderStatus, reorderInstructions } = useReorderable({
        onReorder,
        rootRef,
        itemSelector: ITEM_SELECTOR
    });
    const tabBarReorderContext = {
        currentItemKey: reorderContext.currentItemKey,
        onReorder: onReorder
    };
    return (jsxs("div", { ref: rootRef, ...reorderProps, children: [jsx(ReorderableContext.Provider, { value: tabBarReorderContext, children: jsx(TabBar, { ref: ref, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledby, layout: layout, edge: edge, selection: selection, onSelect: onSelect, onRemove: onRemove, testId: testId, ...displayProps, "aria-describedby": reorderInstructionsId, contextMenuConfig: contextMenuConfig, children: toChildArray(children).map((child) => {
                        return (jsx(ReorderableTabBarItem, { dragKey: dragKey, setDragKey: setDragKey, onReorder: onReorder, layout: layout, rootRef: rootRef, children: child }));
                    }) }) }), jsx("span", { id: reorderInstructionsId, children: jsx(HiddenAccessible, { children: reorderInstructions }) }), jsx(LiveRegion, { children: reorderStatus })] }));
});

export { ReorderableTabBar };
