// @oracle/oraclejet-preact/es/UNSAFE_FormLayout.js
/* @oracle/oraclejet-preact: undefined */
import { jsx } from 'preact/jsx-runtime';
import { a as assignInlineVars } from './vanilla-extract-dynamic.esm-c909758c.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { F as FormContext } from './FormContext-55fb8000.js';
import { u as useFormContext } from './useFormContext-bb9dc478.js';
import { styles } from './UNSAFE_FormLayout/themes/FormLayoutStyles.css.js';
export { compoundVariants, styles, variants } from './UNSAFE_FormLayout/themes/FormLayoutStyles.css.js';
import { l as layoutSpanStyles, a as layoutLocalVars } from './LayoutStyles.css-6b02421a.js';
import { FormLayoutRedwoodTheme } from './UNSAFE_FormLayout/themes/redwood/FormLayoutTheme.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { u as useTestId } from './useTestId-e5d44ab1.js';
import './logger-c92f309c.js';
import 'preact';
import 'preact/hooks';
import './LayoutStyles.styles.css';
import './UNSAFE_FormLayout/themes/redwood/FormLayoutBaseTheme.styles.css';
import './FormLayoutStyles.styles.css';
import './UNSAFE_FormLayout/themes/redwood/FormLayoutBaseTheme.css.js';
import './UNSAFE_FormLayout/themes/redwood/FormLayoutVariants.css.js';
import './FormLayoutVariants.styles.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './LayerHost-12489cef.js';
import './index-3e2e69c4.js';
import 'preact/compat';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';

const FormLayout = ({ columns = 1, columnBehavior = 'responsive', columnSpan = 1, direction = 'row', isFullWidth, isReadonly: propIsReadonly, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, labelWrapping: propLabelWrapping, userAssistanceDensity: propUserAssistanceDensity, children, testId }) => {
    const { isDisabled, isFormLayout: formIsFormLayout, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, labelWrapping: formLabelWrapping, 
    // Components in a FormLayout should default to textAlign="start" in all configurations
    textAlign = 'start', userAssistanceDensity: formUserAssistanceDensity } = useFormContext();
    const testIdProps = useTestId(testId);
    const layoutElementTestIdProps = useTestId(testId ? `${testId}_layout` : undefined);
    const isReadonly = propIsReadonly ?? isFormReadonly;
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    const labelWrapping = propLabelWrapping ?? formLabelWrapping;
    // If we have the prop, pass it on, or if we are in a form layout, use the context value, otherwise default to 'efficient'
    const userAssistanceDensity = propUserAssistanceDensity ?? (formIsFormLayout ? formUserAssistanceDensity : 'efficient');
    // We need to set the localVars.minColumnCount and localVars.maxColumnCount so that we can correctly set the max width to the current column count
    const assignVarStyles = assignInlineVars({
        [layoutLocalVars.maxColumnCount]: `${columns}`,
        [layoutLocalVars.minColumnCount]: `${columnBehavior === 'responsive' ? 1 : columns}`
    });
    const { baseTheme, variantClasses } = useComponentTheme(FormLayoutRedwoodTheme, {
        columnBehavior,
        direction
    });
    const { variantClasses: rootMaxWidthVariant } = useComponentTheme(FormLayoutRedwoodTheme, {
        labelEdge,
        fullWidth: isFullWidth ? 'isFullWidth' : 'notFullWidth'
    });
    return (jsx(FormContext.Provider, { value: {
            isDisabled,
            isFormLayout: true,
            isReadonly,
            labelEdge,
            labelStartWidth,
            labelWrapping,
            textAlign,
            userAssistanceDensity
        }, children: jsx("div", { class: classNames([
                baseTheme,
                rootMaxWidthVariant,
                styles.rootWrapperStyle,
                layoutSpanStyles.layoutSpanColumn[columnSpan]
            ]), ...testIdProps, style: assignVarStyles, children: jsx("div", { class: variantClasses, ...layoutElementTestIdProps, children: children }) }) }));
};

export { FormLayout };
