// @oracle/oraclejet-preact/es/UNSAFE_TabBarMixed.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, Fragment, jsxs } from 'preact/jsx-runtime';
import { useRef, useMemo } from 'preact/hooks';
import { forwardRef, useImperativeHandle } from 'preact/compat';
import './LayerHost-12489cef.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './Flex-5befe411.js';
import { T as TabBarContext } from './TabBarLayout-e6b564fc.js';
export { a as TabBarLayout } from './TabBarLayout-e6b564fc.js';
import './flexitem-7b7f7920.js';
import './dimensions-ddde0eee.js';
import './boxalignment-51b42a0f.js';
import './logger-c92f309c.js';
import './TabBarItemStyles.styles.css';
import './UNSAFE_TabBarCommon/themes/redwood/TabBarItemBaseTheme.styles.css';
import './UNSAFE_TabBarCommon/themes/redwood/TabBarItemVariants.css.js';
import './ConveyorBeltContext-76a29c59.js';
import './BusyStateContext-9ee9751d.js';
import './TabbableModeContext-7d8ad946.js';
import 'preact';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-466c63c4.js';
import './IconStyle.styles.css';
import './Floating-6023b41d.js';
import './LayerStyles.styles.css';
import './keyboardUtils-986fe923.js';
import './tabbableUtils-ce19b388.js';
import './CollectionInteractionContext-7cf19abc.js';
import './dropdownStyles.styles.css';
import './Sheet-24c3e407.js';
import './OverflowTabBarItemStyles.styles.css';
import { i as isMobile } from './clientUtils-71f2dbaf.js';
import { styles } from './UNSAFE_TabBarMixed/themes/TabBarMixedStyles.css.js';
import { u as useTabBar, g as getItemAttribute, R as REMOVABLE_ITEM_ATTRIBUTE } from './useTabBar-816d84de.js';
import { L as LiveRegion } from './LiveRegion-79ffc814.js';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { T as TabBarContextMenu } from './TabBarContextMenu-47631108.js';
import { V as View } from './View-ef109be6.js';
import { styles as styles$1 } from './UNSAFE_TabBarMixed/themes/TabBarMixedSeparator.css.js';
export { O as OverflowTabBarItem, R as RemovableTabBarItem, T as TabBarItem } from './OverflowTabBarItem-34746a16.js';
import './index-3e2e69c4.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './TextStyles.styles.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './useTestId-e5d44ab1.js';
import './classNames-4e12b00d.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './mergeInterpolations-30cd9a69.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry1-63949a9b.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './FlexStyles.styles.css';
import './flexitem.styles.css';
import './size-a3d3aeaf.js';
import './utils-cfef1929.js';
import './themeContract.css-2b5a42e2.js';
import './colorUtils-b3804ee8.js';
import './arrayUtils-35a58161.js';
import './boxalignment.styles.css';
import './TabBarItemVariants.styles.css';
import './TooltipContentVariants.styles.css';
import './EnvironmentProvider-546b7ebd.js';
import './LayerManager-a62e619b.js';
import './Layer-7d3dc8c7.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './useThemeInterpolations-4b48d878.js';
import './useColorScheme-96cce6a1.js';
import './useScale-0ebe15ed.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './useDensity-d06ae84a.js';
import './unsafeDomAccess-453eb74e.js';
import './useFloating-e1167cb9.js';
import './useUser-df8ae011.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-73a51e1d.js';
import './useModal-953082f2.js';
import './useComponentTheme-d2f9e47f.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './clientHints-a42b510c.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './Modal-678575d5.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-4edf62b5.js';
import './WindowOverlay-b380ce51.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './UNSAFE_Sheet/themes/SheetStyles.css.js';
import './SheetStyles.styles.css';
import './useAnimationStatus-072d6a98.js';
import './useAnimation-a5eb7604.js';
import './animationUtils-5f409a6c.js';
import './useTranslationBundle-25469f1c.js';
import './TabBarMixedStyles.styles.css';
import './mergeProps-88ea8306.js';
import './collectionUtils-893f2808.js';
import './keys-49c92dbb.js';
import './useCollectionFocusRing-71cdd52c.js';
import './useCurrentKey-7c652655.js';
import './useId-03dbfdf0.js';
import './useReorderableContext-79b1c941.js';
import './ReorderableContext-0daeea89.js';
import './useContextMenuGesture-906cf78d.js';
import './textSelectionUtils-a6554a42.js';
import './LiveRegionStyles.styles.css';
import './HiddenAccessibleStyles.styles.css';
import './SelectMenuGroupContext-0c1a7603.js';
import './Dropdown-ead9e679.js';
import './usePopupAnimation-66b186ad.js';
import './popupAnimationUtils-175151d9.js';
import './UNSAFE_Dropdown/themes/dropdownStyles.css.js';
import './useOutsideMousedown-39d7c9a6.js';
import './UNSAFE_Menu/themes/MenuStyles.css.js';
import './MenuStyles.styles.css';
import './UNSAFE_Menu/themes/DropdownMenuStyles.css.js';
import './DropdownMenuStyles.styles.css';
import './Skeleton-a06f4f74.js';
import './borders-98a63040.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './SkeletonStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './UNSAFE_Menu/themes/MenuSkeletonStyles.css.js';
import './MenuSkeletonStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemTheme.js';
import './UNSAFE_Menu/themes/MenuItemStyles.css.js';
import './MenuItemStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js';
import './MenuItemVariants.styles.css';
import './Text-983c749e.js';
import './usePress-051f5adb.js';
import './useInteractionStyle-99e32c23.js';
import './useHover-85077adb.js';
import './useToggle-8b7fcefe.js';
import './useActive-fb6d4d52.js';
import './Separator-93c70737.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './SeparatorStyles.styles.css';
import './MenuSeparatorStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js';
import './tabBarUtils-9483d091.js';
import './colors-b6530fec.js';
import './padding-9b3afaa2.js';
import './aria-efa0582c.js';
import './equals-625352e8.js';
import './TabBarMixedSeparator.styles.css';
import './Close-315568df.js';
import './Icon-ef2ac069.js';
import './useTooltip-f442fedb.js';
import './useTooltipControlled-d7724623.js';
import './useFocus-38c95977.js';
import './useTouch-4828df25.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './EmbeddedIconButton-5b4a9aea.js';
import './BaseButton-c2fbaa6d.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-0954f33d.js';
import './useBareButton-07a3eef9.js';
import './usePressClick-2ec8e098.js';
import './useTabbableMode-a2fdbb06.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-d449261a.js';
import './useButtonLabelLayoutTheme-a1b95a52.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './UNSAFE_TabBarCommon/themes/redwood/TabBarItemTheme.js';
import './UNSAFE_TabBarCommon/themes/TabBarItemStyles.css.js';
import './UNSAFE_TabBarCommon/themes/redwood/TabBarItemBaseTheme.css.js';
import './useConveyorBeltItem-ce7d69d6.js';
import './useDetectTruncation-eb704b0c.js';
import './truncationUtils-e37f95ee.js';
import './useResizeObserver-13641643.js';
import './useAddBusyState-7a6af61f.js';
import './useBusyStateContext-4d3c27c0.js';
import './WarningS-7c6d2d0e.js';
import './ErrorS-60a975db.js';
import './InformationS-daba658e.js';
import './SuccessS-49782304.js';
import './UNSAFE_TabBarCommon/themes/TabBarItemContract.css.js';
import './Badge-a458c469.js';
import './UNSAFE_Badge/themes/redwood/BadgeTheme.js';
import './UNSAFE_Badge/themes/BadgeStyles.css.js';
import './BadgeStyles.styles.css';
import './UNSAFE_Badge/themes/redwood/BadgeThemeBaseTheme.css.js';
import './UNSAFE_Badge/themes/redwood/BadgeThemeBaseTheme.styles.css';
import './UNSAFE_Badge/themes/redwood/BadgeVariants.css.js';
import './BadgeVariants.styles.css';
import './UNSAFE_TabBarCommon/themes/OverflowTabBarItemStyles.css.js';
import './NavigationListItem-78b1f538.js';
import './UNSAFE_NavigationList/themes/NavigationListStyles.css.js';
import './NavigationListStyles.styles.css';
import './NavigationListLinkItem-a5dfc1cc.js';
import './UNSAFE_NavigationListCommon/themes/redwood/NavigationListItemTheme.js';
import './UNSAFE_NavigationList/themes/NavigationListItemStyles.css.js';
import './NavigationListItemStyles.styles.css';
import './UNSAFE_NavigationList/themes/redwood/NavigationListItemBaseTheme.css.js';
import './UNSAFE_NavigationList/themes/redwood/NavigationListItemBaseTheme.styles.css';
import './UNSAFE_NavigationList/themes/redwood/NavigationListItemVariants.css.js';
import './NavigationListItemVariants.styles.css';

/**
 * A thin wrapper around useTabBar hook
 */
function useTabBarMixed(options) {
    return useTabBar(options);
}

/**
 * A navigation component that enables horizontal navigation between distinct content with a mixture of static and dynamic tabs.
 *
 * @param {TabBarMixedProps} props TabBarMixed component props.
 * @returns {JSX.Element} TabBarMixed component element.
 */
const TabBarMixed = forwardRef((props, focusHandleRef) => {
    const { children, onRemove, onSelect, selection, size, 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledby } = props;
    const rootRef = useRef(null);
    const isMobileDevice = isMobile();
    const defaultContextMenuConfig = useMemo(() => {
        return {
            itemsRenderer: (context) => {
                //Only if context is available, it is a mobile device and it is removable item, we will return remove context menu item
                if (context) {
                    const isRemovableItem = getItemAttribute(context.itemKey, rootRef.current, REMOVABLE_ITEM_ATTRIBUTE) != null;
                    if (isRemovableItem && isMobileDevice) {
                        const defaultItems = context.defaultMenuItems;
                        return jsx(Fragment, { children: defaultItems });
                    }
                }
                return null;
            }
        };
    }, []);
    const { rootProps, tabBarContext, menuProps, contextMenuContext, contextMenuDescription, contextMenuDescriptionId, accStatusInfo } = useTabBarMixed({
        class: styles.tabBarBase,
        edge: 'bottom',
        onRemove,
        onSelect,
        ref: rootRef,
        selection,
        size,
        children,
        contextMenuConfig: defaultContextMenuConfig
    });
    const isContextMenuAvailable = defaultContextMenuConfig != null && isMobileDevice;
    //Allows to call focus on rootRef without having to expose it
    useImperativeHandle(focusHandleRef, () => ({
        focus: () => {
            if (rootRef.current) {
                rootRef.current.focus();
            }
        },
        blur: () => {
            if (rootRef.current) {
                rootRef.current.blur();
            }
        }
    }));
    return (jsxs(Fragment, { children: [jsx(TabBarContext.Provider, { value: tabBarContext, children: jsx("div", { "aria-label": ariaLabel, "aria-labelledby": ariaLabelledby, ...rootProps, children: children }) }), isContextMenuAvailable && contextMenuDescriptionId && contextMenuDescription && (jsx("span", { id: contextMenuDescriptionId, children: jsx(HiddenAccessible, { children: contextMenuDescription }) })), isContextMenuAvailable && accStatusInfo && jsx(LiveRegion, { children: accStatusInfo }), isContextMenuAvailable && (jsx(TabBarContextMenu, { ...menuProps, contextMenuConfig: defaultContextMenuConfig, contextMenuContext: contextMenuContext }))] }));
});

/**
 * TabBarMixedSeparator is a component that renders a visual separator (displayed
 * as a vertical line) between collections of tabs within TabBarMixed.
 *
 * @param {TabBarMixedSeparatorProps} props Component props for TabBarMixedSeparator.
 * @returns {JSX.Element} Component element for TabBarMixedSeparator.
 */
function TabBarMixedSeparator(props) {
    const { padding = 0 } = props;
    return (jsx(View, { padding: [0, padding, 0, padding], children: jsx("div", { class: styles$1.separator, role: "presentation" }) }));
}

export { TabBarMixed, TabBarMixedSeparator };
