// @oracle/oraclejet-preact/es/UNSAFE_InlineSelectSingle.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs, Fragment } from 'preact/jsx-runtime';
import { forwardRef } from 'preact/compat';
import { useRef, useState, useImperativeHandle, useCallback, useEffect } from 'preact/hooks';
import { F as Flex } from './Flex-5befe411.js';
import { L as Label } from './Label-5c64c52d.js';
import { u as useDisplayValue, S as SelectSingleField, a as SelectSingleFieldInput } from './useDisplayValue-13d53b0b.js';
import { S as Separator } from './Separator-93c70737.js';
import './logger-c92f309c.js';
import { T as TabbableModeContext } from './TabbableModeContext-7d8ad946.js';
import './LayerHost-12489cef.js';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import { a as InlineUserAssistance } from './InlineUserAssistance-f57a9043.js';
import './MessageBannerStyles.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js';
import 'preact';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-466c63c4.js';
import './IconStyle.styles.css';
import './IconButton-c91faa06.js';
import './MessageStyles.styles.css';
import './MessageFormattingUtils-cdb25115.js';
import './TransitionGroup-5fd80dc9.js';
import './MessagesContext-76544845.js';
import './HiddenAccessibleStyles.styles.css';
import './ComponentMessageStyles.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js';
import { F as FormFieldContext } from './FormFieldContext-47ca0c20.js';
import './Popup-0f026f03.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './InputGroupContext-20bdbd6e.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupVariants.css.js';
import { V as View } from './View-ef109be6.js';
import './FormContext-55fb8000.js';
import { u as useFormContext } from './useFormContext-bb9dc478.js';
import { u as useLoadingIndicatorTimer } from './useLoadingIndicatorTimer-898d4e3e.js';
import { u as useTextField } from './useTextField-eb9db737.js';
import { u as useTranslationBundle } from './useTranslationBundle-25469f1c.js';
import { i as isControlOrFunctionKey, K as KEYS } from './keyboardUtils-986fe923.js';
import { a as allTabbableElements } from './tabbableUtils-ce19b388.js';
import { U as UNSAFE_DOM_ACCESS } from './unsafeDomAccess-453eb74e.js';
import { h as isBeforeDataFetch, d as useSelectCommon, j as isSearchTextEmptyOrUndefined, f as isPhone, u as useCollectionRenderer, D as DefaultList, b as DropdownList, E as EmptyResults } from './useSelectCommon-83a6eaeb.js';
import './useTestId-e5d44ab1.js';
import './classNames-4e12b00d.js';
import './boxalignment-51b42a0f.js';
import './size-a3d3aeaf.js';
import './utils-cfef1929.js';
import './themeContract.css-2b5a42e2.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './arrayUtils-35a58161.js';
import './boxalignment.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './dimensions-ddde0eee.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './flexitem-7b7f7920.js';
import './flexitem.styles.css';
import './mergeInterpolations-30cd9a69.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './FlexStyles.styles.css';
import './useFormFieldContext-daf7a676.js';
import './useComponentTheme-d2f9e47f.js';
import './useTooltip-f442fedb.js';
import './useTooltipControlled-d7724623.js';
import './useId-03dbfdf0.js';
import './useHover-85077adb.js';
import './useToggle-8b7fcefe.js';
import './useFocus-38c95977.js';
import './useTouch-4828df25.js';
import './mergeProps-88ea8306.js';
import './useAnimation-a5eb7604.js';
import './useThemeInterpolations-4b48d878.js';
import './useColorScheme-96cce6a1.js';
import './useScale-0ebe15ed.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './useDensity-d06ae84a.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './TooltipContentVariants.styles.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './clientHints-a42b510c.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './SeparatorStyles.styles.css';
import './CompactLabelAssistance-1082f67f.js';
import './useTabbableMode-a2fdbb06.js';
import './Help-2b3a3ccf.js';
import './Icon-ef2ac069.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './useFocusWithin-30b1e2d8.js';
import './UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js';
import './UNSAFE_Label/themes/redwood/LabelTheme.js';
import './UNSAFE_Label/themes/LabelStyles.css.js';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js';
import './truncationUtils-e37f95ee.js';
import './UNSAFE_TextField/themes/redwood/TextFieldTheme.js';
import './UNSAFE_TextField/themes/TextFieldStyles.css.js';
import './SkeletonStyles.styles.css';
import './TextFieldLoadingStyles.styles.css';
import './TextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js';
import './TextFieldVariants.styles.css';
import './NavLeft-65df34b6.js';
import './StyledTextField-9c60efac.js';
import './useAccessibleContext-5744de8b.js';
import './useTextFieldInputHandlers-8122048f.js';
import './formControlUtils-84e2af0e.js';
import './textAlign-ad252afa.js';
import './text.styles.css';
import './TextFieldUtils-8232bca7.js';
import './UNSAFE_TextField/themes/FormControlUtilsStyles.css.js';
import './FormControlUtilsStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldInputTheme.js';
import './UNSAFE_TextField/themes/TextFieldInputStyles.css.js';
import './TextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldInputVariants.css.js';
import './TextFieldInputVariants.styles.css';
import './LabelValueLayout-d7a71947.js';
import './StyledLabelValueLayout-32bdf9d1.js';
import './LabelValueLayoutStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/FormLayoutTheme.js';
import './UNSAFE_TextField/themes/FormLayoutStyles.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.styles.css';
import './formControlUtils-4ec34117.js';
import './LayoutStyles.css-6b02421a.js';
import './LayoutStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './useInputGroupContext-1d19411e.js';
import './useUser-df8ae011.js';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldTheme.js';
import './UNSAFE_TextField/themes/ReadonlyTextFieldStyles.css.js';
import './ReadonlyTextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldVariants.css.js';
import './ReadonlyTextFieldVariants.styles.css';
import './ObfuscatedTextFieldInputStyles.styles.css';
import './ReadonlyTextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputVariants.css.js';
import './TextField-2643ae3c.js';
import './LiveRegionStyles.styles.css';
import './LiveRegion-79ffc814.js';
import './index-3e2e69c4.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './ComponentMessageContainer-6149c829.js';
import './MessagesManager-c46786cb.js';
import './SuccessS-49782304.js';
import './ErrorS-60a975db.js';
import './InformationS-daba658e.js';
import './WarningS-7c6d2d0e.js';
import './PRIVATE_Message/themes/MessageStyles.css.js';
import './soundUtils-88264cb5.js';
import './useMessagesContext-9e1dbe91.js';
import './Transition-f9501682.js';
import './HiddenAccessible-61197e7c.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js';
import './UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js';
import './InlineHelp-8293a502.js';
import './MessageBannerVariants.styles.css';
import './EnvironmentProvider-546b7ebd.js';
import './LayerManager-a62e619b.js';
import './Floating-6023b41d.js';
import './useFloating-e1167cb9.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-73a51e1d.js';
import './Layer-7d3dc8c7.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useModal-953082f2.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './BaseButton-c2fbaa6d.js';
import './useActive-fb6d4d52.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-0954f33d.js';
import './useBareButton-07a3eef9.js';
import './usePress-051f5adb.js';
import './usePressClick-2ec8e098.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-d449261a.js';
import './Text-983c749e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useButtonLabelLayoutTheme-a1b95a52.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './getLocale-1c41e86a.js';
import './stringUtils-0fe25200.js';
import './FocusTrap-9f37aa41.js';
import './FocusTracker-02ad6723.js';
import './PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js';
import './FocusTrackerStyles.styles.css';
import './Modal-678575d5.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-4edf62b5.js';
import './WindowOverlay-b380ce51.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './usePopupAnimation-66b186ad.js';
import './useAnimationStatus-072d6a98.js';
import './popupAnimationUtils-175151d9.js';
import './UNSAFE_Popup/themes/PopupContract.css.js';
import './UNSAFE_Popup/themes/redwood/PopupTheme.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js';
import './PopupStyles.styles.css';
import './colors-b6530fec.js';
import './borders-98a63040.js';
import './padding-9b3afaa2.js';
import './aria-efa0582c.js';
import './equals-625352e8.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './useTimer-571e41e7.js';
import './useMessageSeverity-65295e8f.js';
import './useFormFieldContextProps-31bbc12f.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './HighlightText-d8012ba6.js';
import './HighlightTextStyles.styles.css';
import './Grid-5d56c8aa.js';
import './grid-65aa641c.js';
import './GridStyles.styles.css';
import './ListItemLayoutStyles.styles.css';
import './ListItemText-df37628b.js';
import './Skeleton-a06f4f74.js';
import './PRIVATE_SelectCommon/themes/dropdownStyles.css.js';
import './dropdownStyles.styles2.css';
import './List-ec8a9e41.js';
import './keys-49c92dbb.js';
import './collectionUtils-893f2808.js';
import './LoadMoreCollection-02671d0d.js';
import './Collection-c4f2f3ea.js';
import './useViewportIntersect-8a2b324b.js';
import './VirtualizedCollection-54b5d12a.js';
import './useNavigation-cd59ca08.js';
import './useCurrentKey-7c652655.js';
import './useReorderableContext-79b1c941.js';
import './ReorderableContext-0daeea89.js';
import './useCollectionFocusRing-71cdd52c.js';
import './useTabbableModeSet-274065a0.js';
import './useInteractionStyle-99e32c23.js';
import './PRIVATE_List/themes/ListStyles.css.js';
import './ImageVars.styles.css';
import './ListStyles.styles.css';
import './Selector-1ceeaf0a.js';
import './UNSAFE_Selector/themes/SelectorStyles.css.js';
import './SelectorStyles.styles.css';
import './StyledCheckbox-bf4ac68c.js';
import './Check-67a66daf.js';
import './CheckboxOff-79645ca6.js';
import './CheckboxOn-cf142463.js';
import './CheckboxMixed-56054e95.js';
import './UNSAFE_Checkbox/themes/CheckboxIconContract.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js';
import './UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js';
import './CheckboxIconStyles.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js';
import './CheckboxIconVariants.styles.css';
import './StyledCheckbox.styles.css';
import './CollectionInteractionContext-7cf19abc.js';
import './useCollectionInteractionContext-3895c632.js';
import './ImageVars.css-7b3180c8.js';
import './useTheme-d30637db.js';
import './useSelection-c405739d.js';
import './textSelectionUtils-a6554a42.js';
import './useAdditiveSelection-dc54508b.js';
import './useItemAction-f256a7af.js';
import './Inset-f310ac0b.js';
import './UNSAFE_Inset/themes/InsetStyles.css.js';
import './InsetStyles.styles.css';
import './PRIVATE_List/themes/redwood/ListBaseTheme.styles.css';
import './CaretDown-1777ecfc.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * A custom hook for determining prop values for the InlineSelectSingle's child
 * components from its own props.
 * @param param0 options
 * @param ref The ref handle of the component.
 */
function useInlineSelectSingle({ 'aria-describedby': ariaDescribedBy, assistiveText, data, hasBackIcon = 'never', hasClearIcon = 'never', helpSourceLink, helpSourceText, isLoading: propIsLoading, itemText, label, onBackIconAction, onCommit: propOnCommit, onFilter, placeholder, textAlign: propTextAlign, valueItem, virtualKeyboard }, ref) {
    const inputRef = useRef(null);
    const { textAlign: formTextAlign } = useFormContext();
    const textAlign = propTextAlign ?? formTextAlign ?? 'start';
    const isDataFetched = data != null && !isBeforeDataFetch(data);
    const isEmptyResults = isDataFetched && data?.totalSize === 0;
    // Refs to keep track of prop changes and sync internal state
    const prevDataRef = useRef(data);
    const prevValueItemRef = useRef(valueItem);
    const isInitialRender = useRef(true);
    // State for holding the selected keys in the list
    const [selectedKeys, setSelectedKeys] = useState(valueItem === undefined ? undefined : new Set([valueItem.key]));
    // The incoming propIsLoading tells us that we are in a loading state, but we don't want to
    // show the loading indicator until after a delay, because showing it immediately could result
    // in unwanted flashing.  The isLoading var below will be set to true by useLoadingIndicatorTimer
    // after the delay, when the timer expires, at which point we should show the loading indicator.
    // If propIsLoading is set to false before the timer expires, then the timer will be cancelled
    // and no loading indicator will be shown.
    // (ListView detects its loading state by checking data === null.  While it would be nice for
    // us to be consistent and check valueItems === null, we need to use a separate isLoading prop.
    // If valueItems === null triggers our loading state, then the selections in the dropdown
    // are cleared for that period of time, which is undesirable.)
    const isLoading = useLoadingIndicatorTimer(propIsLoading ?? false);
    const displayValue = useDisplayValue(itemText, valueItem);
    // add methods to the ref
    useImperativeHandle(ref, () => ({
        focus: () => inputRef.current?.focus(),
        blur: () => inputRef.current?.blur(),
        [UNSAFE_DOM_ACCESS]: inputRef.current
    }), []);
    const { currentRowKeyRef, currentRowOverride, dropdownRef, isFocused, isUserFiltering, onCurrentKeyChange, searchText, setCurrentRowOverride, stopFiltering, updateFiltering, userInput } = useSelectCommon({
        data,
        inputRef,
        // The component can only be in a focused visual state
        isFocused: true,
        onFilter
    });
    const { baseId, formFieldContext, inputProps, labelProps, userAssistanceProps } = useTextField({
        ariaDescribedBy,
        helpSourceLink,
        helpSourceText,
        isFocused,
        isLoading,
        value: valueItem !== undefined ? true : undefined
    });
    const listId = `${baseId}-list`;
    // If we're waiting for the selected values to load, the component should look like it
    // has a value in the meantime.
    // Use the original propIsLoading instead of isLoading because we don't want to delay
    // the inside label position by a timer like we do the progressive loading indicator.
    formFieldContext.hasValue = propIsLoading || formFieldContext.hasValue;
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const liveRegionText = isDataFetched
        ? data?.totalSize === 0
            ? translations.select_noMatchesFound()
            : data?.totalSize === 1
                ? translations.select_oneMatchFound()
                : data?.sizePrecision === 'exact'
                    ? translations.select_sizeMatchesFound({ TOTAL_SIZE: `${data?.totalSize}` })
                    : translations.select_sizeOrMoreMatchesFound({ TOTAL_SIZE: `${data?.totalSize}` })
        : '';
    /**
     * Lifecycle updates
     */
    // When the valueItem changes, save the new selectedKeys
    if (prevValueItemRef.current !== valueItem) {
        prevValueItemRef.current = valueItem;
        setSelectedKeys(valueItem === undefined ? undefined : new Set([valueItem.key]));
    }
    // When data changes, highlight the first result
    if (prevDataRef.current !== data) {
        prevDataRef.current = data;
        // when filtering, set the first result of the filtered data set to be the current item
        // (check the data offset so we don't try to highlight again if the user scrolls through
        // the list and more data is fetched)
        if (isUserFiltering && data != null && data.offset === 0 && data.data.length > 0) {
            // if there are results and search text, highlight the first row, otherwise clear
            // an existing highlight
            setCurrentRowOverride({
                rowKey: !isSearchTextEmptyOrUndefined(searchText) ? data.data[0].metadata.key : undefined
            });
        }
    }
    // During the initial render, set the current row override if we have selected values.
    // This way pressing DOWN arrow will start the navigation from the selected row.
    if (isInitialRender.current) {
        isInitialRender.current = false;
        setCurrentRowOverride({ rowKey: [...(selectedKeys ?? [undefined])][0] });
    }
    /**
     * A helper onCommit wrapper that consumes the metadata and calls the
     * propOnCommit with only the necessary information.
     * (Similar to onCommit in useSelectSingle.ts)
     */
    const onCommit = useCallback((value, metadata) => {
        propOnCommit?.({
            value: value,
            previousValue: valueItem !== undefined ? valueItem.key : undefined,
            reason: metadata.reason
        });
    }, [propOnCommit, valueItem]);
    /**
     * A helper function to clear the selectedKeys and commits empty value.
     * (Similar to clearValue in useSelectSingle.ts but without dropdown related logic)
     */
    const clearValue = useCallback((metadata) => {
        setSelectedKeys(new Set());
        // call the onCommit callback so that the consuming code can update the valueItem
        onCommit(undefined, metadata);
    }, [onCommit]);
    /**
     * Determine if we need to clear the value because the user has deleted all the text.
     * (Similar to isClearValueNeeded in useSelectSingle.ts)
     */
    const isClearValueNeeded = useCallback(() => {
        return isUserFiltering && userInput === '' && valueItem !== undefined;
    }, [isUserFiltering, userInput, valueItem]);
    /**
     * Handler for List.onItemAction
     * (Similar to useSelectSingle.ts's onItemAction but without dropdown related logic)
     * This is used for handling:
     * - Commits value when one performs an action on the list
     */
    const handleItemAction = useCallback((detail) => {
        // after selecting a value, the user is no longer actively filtering, so update related
        // state accordingly so that the newly selected value appears in the field
        stopFiltering();
        const newKey = detail.context.key;
        // update the selectedKeys so that the selection visual state of the collection
        // updates immediately while waiting for a new valueItem to be set
        setSelectedKeys(new Set([newKey]));
        // call the onCommit callback so that the consuming code can update the valueItem
        onCommit(newKey, { reason: 'itemAction' });
    }, [onCommit, stopFiltering]);
    /**
     * Handler for SelectSingleFieldInput.onInput
     * This is used for handling:
     * - Typing in text in the field should filter the results (i.e. invoke onFilter with the text)
     */
    const handleFieldInput = useCallback(({ value }) => {
        updateFiltering(value);
        onFilter?.({ searchText: value });
    }, [onFilter, updateFiltering]);
    /**
     * Handler for SelectSingleFieldInput.onKeyDown
     * This is used for handling:
     * - Prevents default browser behavior on UP/DOWN to avoid scrolling
     * - Clears value or selects current row on ENTER
     * - Clears value or selects current row on TAB
     */
    const handleFieldKeyDown = useCallback((event) => {
        // ignore ctrl and fn key events
        if (isControlOrFunctionKey(event))
            return;
        switch (event.code) {
            case KEYS.UP:
            case KEYS.DOWN:
                // prevent the default action, since we do not want the page to scroll
                event.preventDefault();
                break;
            case KEYS.ENTER:
            case KEYS.TAB: {
                // if this is on a mobile device (only phone and not in tablet devices), do not do anything
                if (isPhone())
                    return;
                // Otherwise, proceed
                const commitMetadata = {
                    reason: event.code === KEYS.TAB ? 'tabKey' : 'itemAction'
                };
                // check to see if we need to clear the value
                if (isClearValueNeeded()) {
                    // in this case, clear the value and return.
                    clearValue(commitMetadata);
                    return;
                }
                // if there is no data to select, no need to handle this event
                if (data == null || data.data.length === 0)
                    return;
                // If we will be making a selection, the user is no longer actively filtering,
                // so update related state accordingly so that the newly selected value appears in the field.
                stopFiltering();
                // Current design have this component in a dropdown/modal. So, pressing Tab from
                // the main field without doing any other interaction is still considered a selection
                // unlike SelectSingle (see JET-68215). So, we handle both Enter and Tab similarly.
                // If there is a currentRow in the dropdown (meaning a highlighted item that would
                // have keyboard focus if the dropdown list had physical focus, which may be different
                // from the currently selected item), select it and commit it.
                if (currentRowKeyRef.current !== undefined && data != null && data.data.length > 0) {
                    setSelectedKeys(new Set([currentRowKeyRef.current]));
                    onCommit(currentRowKeyRef.current, commitMetadata);
                    return;
                }
                // Otherwise commit the currently selected value again so that a valueAction event
                // can be fired from the core-pack side.
                onCommit(valueItem?.key ?? undefined, commitMetadata);
                break;
            }
        }
    }, [clearValue, currentRowKeyRef, data, isClearValueNeeded, onCommit, stopFiltering, valueItem]);
    /**
     * Handler for SelectSingleFieldInput.onKeyUp
     * This is used for handling:
     * - Pressing UP/DOWN from the main field should move the focus to the list
     * - If Ctrl or Fn key is pressed, then ignore the event
     */
    const handleFieldKeyUp = useCallback((event) => {
        // ignore ctrl and fn key events
        if (isControlOrFunctionKey(event))
            return;
        switch (event.code) {
            case KEYS.UP:
            case KEYS.DOWN:
                // Transfer physical focus into the dropdown only if the data is present and the dropdown
                // has rendered the data. Otherwise, keep the focus in the field.
                if (data != null && dropdownRef.current) {
                    // get the focusable element from the dropdown.
                    // consider element with tabindex=-1 as well.
                    const firstFocusableElem = allTabbableElements(dropdownRef.current, undefined, true)[0];
                    // focus the first element if it exists
                    firstFocusableElem?.focus();
                    // set the first row as current row if there is no current row at the moment
                    if (currentRowOverride.rowKey === undefined) {
                        setCurrentRowOverride({ rowKey: data.data[0]?.metadata?.key });
                    }
                }
                // prevent the default action, since we are handling it
                event.preventDefault();
                break;
        }
    }, [currentRowOverride, data, dropdownRef, setCurrentRowOverride]);
    /**
     * Handler for DropdownList.onKeyDown
     * This is used for handling:
     * - Pressing Tab when highlighted on a list item should select & commit that item as value
     */
    const handleListKeyDown = useCallback((event) => {
        switch (event.code) {
            case KEYS.TAB:
                // if there is a currentRow in the dropdown, select it and commit it
                if (currentRowKeyRef.current !== undefined && data != null && data.data.length > 0) {
                    setSelectedKeys(new Set([currentRowKeyRef.current]));
                    onCommit(currentRowKeyRef.current, { reason: 'tabKey' });
                }
                break;
        }
    }, [currentRowKeyRef, data, onCommit]);
    /**
     * Handler for ClearButton.onClick
     * This is for handling:
     * - on clicking the clear icon, the value & selection should be cleared.
     */
    const handleClearButtonClick = useCallback(() => {
        // reset the user filter and the value of the component
        stopFiltering();
        // clear the value
        clearValue({ reason: 'itemAction' });
        // set focus to the input
        inputRef.current?.focus();
        // call onFilter with undefined
        onFilter?.({ searchText: undefined });
    }, [clearValue, onFilter, stopFiltering]);
    /**
     * Create collection Renderer context. This will be used by the default
     * collectionRenderer.
     */
    const collectionRendererContext = useCollectionRenderer({
        currentKey: currentRowOverride.rowKey,
        searchText,
        selectedKeys,
        onCurrentKeyChange,
        onItemAction: handleItemAction
    });
    // TODO: JET-73973 - ListView is not respecting TabbableMode
    // Currently List does not respect TabbableMode, so for now explicitly set tabindex="-1".
    // When this bug is fixed, this useEffect can be removed.
    useEffect(() => {
        const list = dropdownRef.current?.querySelector('div[role="listbox"][tabindex="0"]');
        list?.setAttribute('tabindex', '-1');
    });
    return {
        collectionRendererContext,
        formFieldContext,
        hasUserAssistance: Boolean(assistiveText || helpSourceLink),
        isEmptyResults,
        isUserFiltering,
        labelProps: {
            ...labelProps,
            children: label
        },
        listContainerProps: {
            id: listId,
            dropdownListRef: dropdownRef,
            onKeyDown: handleListKeyDown
        },
        selectSingleFieldProps: {
            contentVariant: 'input',
            // InlineSelectSingle currently only supports inside label.
            hasInsideLabel: true,
            isBackButtonShown: hasBackIcon === 'always',
            isClearButtonShown: hasClearIcon === 'conditionally' && valueItem !== undefined,
            onBackButtonClick: onBackIconAction,
            onClearButtonClick: handleClearButtonClick,
            styleVariant: 'embedded'
        },
        selectFieldInputProps: {
            ...inputProps,
            'aria-controls': listId,
            // InlineSelectSingle always have the list open, so it will always be in an expanded state.
            'aria-expanded': true,
            displayValue,
            // InlineSelectSingle currently only supports inside label,
            // so it is expected to always have a label.
            hasEmptyLabel: false,
            // InlineSelectSingle currently only supports inside label.
            hasInsideLabel: true,
            inputRef,
            // InlineSelectSingle currently does not support Add to List.
            isAddToListShown: false,
            // InlineSelectSingle currently does not support Advanced Search.
            isAdvancedSearchShown: false,
            // InlineSelectSingle currently does not support Required.
            isRequired: false,
            isUserFiltering,
            liveRegionText,
            onInput: handleFieldInput,
            onKeyDown: handleFieldKeyDown,
            onKeyUp: handleFieldKeyUp,
            placeholder,
            textAlign,
            userInput,
            variant: 'embedded',
            virtualKeyboard
        },
        userAssistanceProps: {
            ...userAssistanceProps,
            assistiveText,
            fieldLabel: label,
            helpSourceLink,
            helpSourceText
        }
    };
}

const PREVENT_TABBING = { isTabbable: false };
/**
 * A InlineSelectSingle is a text field with an inline list that supports single selection and search filtering.
 * Note: This is only meant to be used by the Spectra Smart Search component. As such, it only has limited features
 * needed for the Smart Search component.
 */
function InlineSelectSingleImpl(props, ref) {
    const { data, itemRenderer, itemText, label, onLoadRange, testId, userAssistancePosition = 'belowList' } = props;
    const { collectionRendererContext, formFieldContext, hasUserAssistance, isEmptyResults, isUserFiltering, labelProps, listContainerProps, selectSingleFieldProps, selectFieldInputProps, userAssistanceProps } = useInlineSelectSingle(props, ref);
    const defaultCollectionRenderer = useCallback(({ onPersistCurrentRowKey, onRowAction, selected, searchText, currentRowKeyOverride }) => (jsx(DefaultList, { "aria-label": label, currentItemVariant: "highlight", currentKey: currentRowKeyOverride, data: data, itemRenderer: itemRenderer, itemText: itemText, onCurrentKeyChange: onPersistCurrentRowKey, onItemAction: onRowAction, onLoadRange: onLoadRange, searchText: isUserFiltering ? searchText : undefined, selectedKeys: selected, selectionMode: "single" })), [data, isUserFiltering, itemRenderer, itemText, label, onLoadRange]);
    const userAssistance = (jsx(View, { minWidth: "100%", padding: "2x", children: jsx(InlineUserAssistance, { testId: testId && `${testId}_userAssistance`, ...userAssistanceProps }) }));
    return (jsx(Flex, { direction: "column", gap: "2x", maxHeight: "100%", width: "100%", children: jsxs(FormFieldContext.Provider, { value: formFieldContext, children: [jsxs(View, { flex: 0, children: [jsx(SelectSingleField, { ...selectSingleFieldProps, label: jsx(Label, { ...labelProps }), testId: testId && `${testId}_selectSingleField`, children: jsx(SelectSingleFieldInput, { ...selectFieldInputProps }) }), userAssistancePosition === 'aboveList' && jsx(Separator, {}), userAssistancePosition === 'aboveList' && hasUserAssistance && (jsxs(Fragment, { children: [userAssistance, jsx(Separator, {})] }))] }), userAssistancePosition !== 'aboveList' && jsx(Separator, {}), jsx(DropdownList, { ...listContainerProps, 
                    // Add bottom gap only when we are configured to show UA above the list
                    hasBottomGap: userAssistancePosition === 'aboveList', hasTopGap: false, children: jsx(TabbableModeContext.Provider, { value: PREVENT_TABBING, children: isEmptyResults ? (jsx(EmptyResults, {})) : (defaultCollectionRenderer(collectionRendererContext)) }) }), userAssistancePosition === 'belowList' && hasUserAssistance && (jsxs("div", { children: [jsx(Separator, {}), userAssistance] }))] }) }));
}
const InlineSelectSingle = forwardRef(InlineSelectSingleImpl); // forwardRef will remove generics, so we cast it explicitly to preserve generics.;

export { InlineSelectSingle };
