// @oracle/oraclejet-preact/es/useProgressButtonLabelLayoutTheme-7ce1cf07.js
/* @oracle/oraclejet-preact: undefined */
import { ButtonLabelLayoutRedwoodTheme } from './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useButtonLabelLayoutTheme } from './useButtonLabelLayoutTheme-a1b95a52.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useProgressButtonLabelLayoutTheme({ display, styling, size, children, startIcon, isLoading = false }) {
    const { startIconClasses, containerClasses, isShowIcons, isShowText, hasTextStartIcon } = useButtonLabelLayoutTheme({ display, size, styling, children, startIcon });
    const { classes: labelClasses, styles: labelStyles } = useComponentTheme(ButtonLabelLayoutRedwoodTheme, {
        target: 'label',
        size: size,
        fill: styling == 'fill' ? 'isFill' : 'notFill',
        textEndIcon: 'notTextEndIcon',
        textStartIcon: hasTextStartIcon ? 'hasTextStartIcon' : 'notTextStartIcon',
        suffix: 'notSuffix',
        twoIcons: 'notTwoIcons',
        loading: isLoading && !hasTextStartIcon ? 'isLoading' : 'notLoading'
    });
    const { classes: progressIconClasses } = useComponentTheme(ButtonLabelLayoutRedwoodTheme, {
        target: 'progressIcon',
        size: size
    });
    return {
        startIconClasses,
        labelClasses,
        labelStyles,
        progressIconClasses,
        containerClasses,
        isShowIcons,
        isShowText,
        hasTextStartIcon
    };
}

export { useProgressButtonLabelLayoutTheme as u };
