// @oracle/oraclejet-preact/es/useSelectDrill-9027a3a5.js
/* @oracle/oraclejet-preact: undefined */
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { u as useChartDataCursor } from './useChartDataCursor-82b836ed.js';
import { u as useChartDatatip } from './useChartDatatip-a3ce4e23.js';
import { g as getInfo } from './eventsUtils-e96290df.js';
import './accUtils-f6657e8e.js';
import { u as useChartDrill } from './useChartDrill-d8a91907.js';
import { u as useSelection } from './useSelection-8a26bc9e.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useChartDatatipAndCursor({ series, leafGroupsInfo, width, height, findNearest, getMarkerInfo, onDataCursorPositionChange, isDataCursorEnabled, dataCursorPosition, dataCursorStyle, hoveredItemInfo, focusedItemInfo, getDataItem, getDataItemPos, datatip, valueFormats, isHoriz, xScale, yScale, y2Scale, plotAreaSpace, timeAxisType, touchResponse, isDisabled, defaultXFormatter, defaultYFormatter, defaultY2Formatter, hasXYValues, getGroupFromPath, translations }) {
    const { datatipContent, datatipProps } = useChartDatatip({
        series,
        leafGroupsInfo,
        width,
        height,
        isDataCursorEnabled,
        dataCursorPosition,
        hoveredItemInfo,
        getDataItem,
        getDataItemPos,
        focusedItemInfo,
        datatip,
        valueFormats,
        isHoriz,
        xScale,
        findNearest,
        timeAxisType,
        touchResponse,
        isDisabled,
        defaultXFormatter,
        defaultYFormatter,
        defaultY2Formatter,
        hasXYValues,
        translations,
        getGroupFromPath
    });
    const { dataCursorContent, dataCursorProps } = useChartDataCursor({
        touchResponse,
        findNearest,
        dataCursorPosition,
        onDataCursorPositionChange,
        isDataCursorEnabled,
        dataCursorStyle,
        isHoriz,
        getMarkerInfo,
        focusedItemInfo,
        plotAreaSpace,
        yScale,
        y2Scale,
        xScale,
        getDataItemPos
    });
    return {
        datatipContent,
        dataCursorContent,
        datatipProps: mergeProps(dataCursorProps, datatipProps)
    };
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useSelectDrill({ selectionMode, selection, onChange, rootRef, drilling, onItemDrill, onGroupDrill, focusedGroupInfo, focusedItemInfo, levelsArray, getDataItem, getDataById, ignoreAsTrigger }) {
    const idExtracter = (event) => {
        const isItemFocused = focusedItemInfo.isCurrent;
        const info = isItemFocused && event.type === 'keyup'
            ? focusedItemInfo
            : getInfo(rootRef, event.target);
        return info && info.seriesIndex != null && info.groupIndex != null
            ? getDataItem(info.seriesIndex, info.groupIndex, levelsArray[levelsArray.length - 1][info.groupIndex].path)?.id
            : undefined;
    };
    const { selectionProps, selectionContent } = useSelection({
        idExtracter,
        isDrillEnabled: drilling === 'on',
        selection,
        selectionMode,
        onChange,
        getDataById,
        ignoreAsTrigger
    });
    const drillProps = useChartDrill({
        rootRef,
        selectionMode,
        drilling,
        focusedItemInfo,
        focusedGroupInfo,
        levelsArray,
        getDataItem,
        onItemDrill,
        onGroupDrill
    });
    return { selectDrillProps: mergeProps(selectionProps, drillProps), selectionContent };
}

export { useSelectDrill as a, useChartDatatipAndCursor as u };
