// @oracle/oraclejet-preact/es/useTextField-eb9db737.js
/* @oracle/oraclejet-preact: undefined */
import { useRef, useState, useCallback } from 'preact/hooks';
import { l } from './stringUtils-0fe25200.js';
import { d as detectTruncation } from './truncationUtils-e37f95ee.js';
import { u as useMessageSeverity } from './useMessageSeverity-65295e8f.js';
import { u as useFocusWithin } from './useFocusWithin-30b1e2d8.js';
import { u as useFormFieldContextProps } from './useFormFieldContextProps-31bbc12f.js';
import { u as useId } from './useId-03dbfdf0.js';

/**
 * A custom hook to determine the props for a component that renders
 * a text field
 */
function useTextField({ ariaDescribedBy, contentVariant, helpSourceLink, helpSourceText, isDisabled, isFocused, isLoading, isReadonly, isRequiredShown, labelEdge, messages, styleVariant, userAssistanceDensity, value }) {
    const baseId = useId();
    const labelId = labelEdge !== 'none' ? `${baseId}-label` : undefined;
    const inputId = !isReadonly ? `${baseId}-input` : undefined;
    const uaId = !isDisabled ? `${baseId}-ua` : undefined;
    const helpIconId = userAssistanceDensity === 'compact' ? `${baseId}-helpIcon` : undefined;
    // JET-73197: Truncated Text Enhancements for Core Pack
    // Create a ref to the Label element. When we see focusWithin the input, measure the label
    // to determine if the text is truncated and if so, set showTruncatedTooltip to true.
    // If we see a blurWithin event and the input is no longer focused, set the flag to false.
    const labelRef = useRef(null);
    const [showTruncatedTooltip, setShowTruncatedTooltip] = useState(false);
    const onBlurWithin = useCallback(() => {
        setShowTruncatedTooltip(false);
    }, []);
    const onFocusWithin = useCallback(() => {
        const isTruncated = labelEdge === 'inside' && !isDisabled ? detectTruncation(labelRef) : false;
        setShowTruncatedTooltip(isTruncated);
    }, [isDisabled, labelEdge, labelRef]);
    const { focusProps, isFocused: isInputFocused } = useFocusWithin({
        isDisabled,
        onBlurWithin,
        onFocusWithin
    });
    const hasValue = value !== undefined && (typeof value !== 'string' || value !== '');
    // Create the form field context
    const formFieldContext = useFormFieldContextProps({
        hasTruncatedLabel: showTruncatedTooltip,
        hasValue,
        isDisabled,
        isFocused,
        isInputFocused,
        isLoading,
        isReadonly
    });
    // Determine the highest severity from the messages
    const messageSeverity = useMessageSeverity(messages);
    // Determine the label variant
    const labelVariant = labelEdge !== 'none' ? labelEdge : undefined;
    return {
        baseId,
        formFieldContext,
        inputProps: {
            id: inputId,
            'aria-describedby': l([helpIconId, uaId, ariaDescribedBy]),
            'aria-invalid': messageSeverity === 'error'
                ? 'true'
                : undefined,
            variant: styleVariant,
            onBlur: focusProps?.onFocusOut,
            onFocus: focusProps?.onFocusIn
        },
        labelProps: {
            forId: inputId,
            helpIconId,
            helpSourceLink,
            helpSourceText,
            id: labelId,
            isRequiredShown,
            labelRef,
            userAssistanceDensity,
            variant: labelVariant
        },
        textFieldProps: {
            contentVariant,
            statusVariant: /error|warning/.test(messageSeverity ?? '')
                ? messageSeverity
                : undefined,
            styleVariant
        },
        userAssistanceProps: {
            id: uaId
        }
    };
}

export { useTextField as u };
