// @oracle/oraclejet-webdriver/elements/oj-refresher/OjRefresherBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjRefresherBase = void 0;
const __1 = require("../");
/**
 * This is the base class for oj-refresher WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjRefresher.ts.
 */
class OjRefresherBase extends __1.OjWebElement {
    /**
     * Gets the value of <code>refreshContent</code> property.
     * Specifies the function to invoke when pull-to-refresh operation is triggered. See the Help documentation for more information.
     * @return The value of <code>refreshContent</code> property.
     *
     */
    getRefreshContent() {
        return this.getProperty('refreshContent');
    }
    /**
     * Gets the value of <code>text</code> property.
     * Specifies a text message to display in the pull-to-refresh panel after a pull down gesture.
     * @return The value of <code>text</code> property.
     *
     */
    getTextProperty() {
        return this.getProperty('text');
    }
    /**
     * Gets the value of <code>threshold</code> property.
     * The minimum distance in pixels that the user needs to pull down in order to trigger a refresh. See the Help documentation for more information.
     * @return The value of <code>threshold</code> property.
     *
     */
    getThreshold() {
        return this.getProperty('threshold');
    }
    /**
     * Gets the value of <code>translations</code> property.
     * A collection of translated resources from the translation bundle, or null if this component has no resources.
     * @return The value of <code>translations</code> property.
     *
     */
    getTranslations() {
        return this.getProperty('translations');
    }
}
exports.OjRefresherBase = OjRefresherBase;
