// @oracle/oraclejet-preact/cjs/BaseButton-f6eb6de1.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var useHover = require('./useHover-49b0430d.js');
var useActive = require('./useActive-6770f917.js');
var useColorScheme = require('./useColorScheme-23e4aab4.js');
var classNames = require('./classNames-c14c6ef3.js');
var dimensions = require('./dimensions-2fcc0acf.js');
var mergeInterpolations = require('./mergeInterpolations-6727b536.js');
var clientHints$1 = require('./clientHints-c40c5250.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var UNSAFE_BaseButton_themes_redwood_BaseButtonTheme = require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var BareButton = require('./BareButton-814ae4b7.js');
var buttonUtils = require('./buttonUtils-dff2ea3e.js');

const interpolations = [...Object.values(dimensions.dimensionInterpolations)];
const styleInterpolations = mergeInterpolations.mergeInterpolations(interpolations);
const clientHints = clientHints$1.getClientHints();
// :active only works on IOS devices if a touch handler exists
const iosProps = clientHints.platform === 'ios' ? { ontouchstart: function () { } } : {};
const isHybrid = clientHints$1.getClientHints().hoverSupport === 'events';
const BaseButton = compat.forwardRef(({ variant, isReadOnly = false, isDisabled = false, isRepeat = false, isClick, size, onAction, elementDetails = { type: 'button', buttonType: 'button' }, styling = ['default'], edge = 'none', testId, ...props }, ref) => {
    const isStyled = (name) => {
        return styling.indexOf(name) > -1;
    };
    const { globalEventProps, ariaProps } = buttonUtils.processGlobals(props);
    const { hoverProps, isHover } = useHover.useHover({ isDisabled: !isHybrid });
    const { activeProps, isActive } = useActive.useActive({ isDisabled: isReadOnly });
    const { class: cls, ...styles } = styleInterpolations({
        ...props,
        ...globalEventProps
    });
    const ElementType = elementDetails.type;
    const isSpan = ElementType == 'span';
    const colorScheme = useColorScheme.useColorScheme();
    const { variantClasses } = useComponentTheme.useComponentTheme(UNSAFE_BaseButton_themes_redwood_BaseButtonTheme.BaseButtonRedwoodTheme, {
        size,
        variant,
        unstyled: isStyled('unstyled') ? 'isUnstyled' : 'notUnstyled',
        embedded: isStyled('embedded') ? 'isEmbedded' : 'notEmbedded',
        container: isStyled('container') ? 'isContainer' : 'notContainer',
        noBorderRadiusStart: isStyled('noBorderRadiusStart')
            ? 'isNoBorderRadiusStart'
            : 'notNoBorderRadiusStart',
        noBorderRadiusEnd: isStyled('noBorderRadiusEnd')
            ? 'isNoBorderRadiusEnd'
            : 'notNoBorderRadiusEnd',
        noBorderStart: isStyled('noBorderStart') ? 'isNoBorderStart' : 'notNoBorderStart',
        noBorderEnd: isStyled('noBorderEnd') ? 'isNoBorderEnd' : 'notNoBorderEnd',
        activeStyle: isStyled('active') ? 'isActiveStyle' : 'notActiveStyle',
        selectedStyle: isStyled('selected') ? 'isSelectedStyle' : 'notSelectedStyle',
        buttonSetItem: isStyled('buttonSetItem') ? 'isButtonSetItem' : 'notButtonSetItem',
        trailingItem: isStyled('trailingItem') ? 'isTrailingItem' : 'notTrailingItem',
        fill: isStyled('fill') ? 'isFill' : 'notFill',
        min: isStyled('min') ? 'isMin' : 'notMin',
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        edge,
        hybrid: isHybrid ? 'isHybrid' : 'notHybrid',
        active: isActive ? 'isActive' : 'notActive',
        hover: isHover ? 'isHover' : 'notHover',
        span: isSpan ? 'isSpan' : 'notSpan',
        widthEqual: isStyled('widthEqual') ? 'isWidthEqual' : 'notWidthEqual',
        widthAuto: isStyled('widthAuto') ? 'isWidthAuto' : 'notWidthAuto',
        hiddenLabel: isStyled('hiddenLabel') ? 'isHidden' : 'notHidden',
        noHoverActive: isStyled('noHoverActive') ? 'isNoHoverActive' : 'notNoHoverActive',
        newStackingContext: isStyled('newStackingContext')
            ? 'isNewStackingContext'
            : 'notNewStackingContext',
        colorScheme
    });
    const legacyGlobals = {
        autofocus: props.autofocus,
        title: props.title
    };
    const mergedProps = mergeProps.mergeProps(iosProps, ariaProps, hoverProps, isStyled('container') ? {} : activeProps, globalEventProps, legacyGlobals);
    const wrapperClasses = classNames.classNames([cls, variantClasses]);
    return (jsxRuntime.jsx(BareButton.BareButton, { ref: ref, isDisabled: isDisabled || isReadOnly, isClick: isClick, classNames: wrapperClasses, style: styles, testId: testId, onAction: onAction, autofocus: props['autofocus'], elementDetails: elementDetails, isRepeat: isRepeat, ...ariaProps, ...mergedProps, children: props.children }));
});
BaseButton.displayName = 'BaseButton';

exports.BaseButton = BaseButton;
