// @oracle/oraclejet-preact/cjs/ColorSwatch-8410a4cd.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var preact = require('preact');
var hooks = require('preact/hooks');
var compat = require('preact/compat');
var BareButton = require('./BareButton-814ae4b7.js');
var IconButton = require('./IconButton-217314e6.js');
var useToggleAction = require('./useToggleAction-e4f15550.js');
var useTooltip = require('./useTooltip-8728f2ff.js');
var useHover = require('./useHover-49b0430d.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
require('./LayerHost-aee0e070.js');
var useColorScheme = require('./useColorScheme-23e4aab4.js');
var classNames = require('./classNames-c14c6ef3.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-ae48fed6.js');
var useActive = require('./useActive-6770f917.js');
var clientHints = require('./clientHints-c40c5250.js');
var UNSAFE_ColorSwatch_themes_redwood_ColorSwatchTheme = require('./UNSAFE_ColorSwatch/themes/redwood/ColorSwatchTheme.js');
var buttonUtils = require('./buttonUtils-dff2ea3e.js');
var swatchData = require('./swatchData-17c1d335.js');
var resources_nls_en_bundle = require('./resources/nls/en/bundle.js');
var CheckboxCrossed = require('./CheckboxCrossed-276de4f4.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * The NoFill image
 */
const NoFill = () => (jsxRuntime.jsx("svg", { xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 48 48", preserveAspectRatio: "none", width: "100%", height: "100%", children: jsxRuntime.jsx("path", { "stroke-width": "2", "vector-effect": "non-scaling-stroke", d: "M0 0L48 48" }) }));

/**
 * ColorSwatches render a color, showing focus, hover, and selection variations.
 */
const ColorSwatch = compat.forwardRef(({ size = 'sm', isSelected = false, isButton = false, onToggle, isDisabled = false, isTooltipDisabled = false, testId, value, ...props }, ref) => {
    const clientHints$1 = clientHints.getClientHints();
    const isHybrid = clientHints$1.hoverSupport === 'events';
    const valueText = value;
    const hasValue = valueText && valueText.length;
    const lowerValueText = hasValue ? valueText?.toLocaleLowerCase() : value;
    const isNoFill = hasValue ? lowerValueText === 'transparent' : false;
    const hasAlpha = value && hasValue ? swatchData.isEightDigitHex(value) : false;
    // const isSmallButton = isButton && size === 'xs'; // if used as xs button, the icon is 20x20, while the target size is 24x24
    const selectedTranslated = resources_nls_en_bundle["default"].vis_stateSelected();
    const noneTranslated = resources_nls_en_bundle["default"].labelNone();
    const colorSwatchTranslated = resources_nls_en_bundle["default"].accColorSwatch();
    const { globalEventProps, ariaProps: { 'aria-label': ariaLabel, 'aria-hidden': ariaHidden, 'aria-roledescription': ariaRoleDescription, ...swatchAriaProps } } = buttonUtils.processGlobals(props);
    const { hoverProps, isHover } = useHover.useHover({ isDisabled: !isHybrid });
    const { activeProps, isActive } = useActive.useActive({ isDisabled: isDisabled });
    // Cope with rgb or palette colors or illegal colors
    const getHexValue = hooks.useCallback((raw) => {
        if (raw?.toLowerCase() === 'transparent') {
            return noneTranslated;
        }
        return swatchData.parseColorValue(raw);
    }, [noneTranslated]);
    const [accessibleLabel, setAccessibleLabel] = hooks.useState(ariaLabel
        ? ariaLabel
        : getHexValue(value) + isSelected
            ? ' ' + selectedTranslated
            : '');
    const colorScheme = useColorScheme.useColorScheme();
    const { classes, styles: { alphaHalfStyles, swatchStyles, buttonStyles, noFillStyles, noFillDisabledStyles, noFillHoverStyles, smallButtonContainerStyles } } = useComponentTheme.useComponentTheme(UNSAFE_ColorSwatch_themes_redwood_ColorSwatchTheme.ColorSwatchRedwoodTheme, {
        size: size,
        selected: isSelected ? 'isSelected' : 'notSelected',
        active: isActive ? 'isActive' : 'notActive',
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        hover: isHover ? 'isHover' : 'notHover',
        noFill: isNoFill ? 'isNoFill' : 'notNoFill',
        hasValue: hasValue ? 'hasValue' : 'notHasValue',
        button: isButton ? 'isButton' : 'notButton',
        colorScheme
    });
    hooks.useEffect(() => {
        if (!ariaLabel) {
            setAccessibleLabel((isSelected ? selectedTranslated + ' ' : '') + getHexValue(value));
        }
    }, [isSelected, value, getHexValue, selectedTranslated, ariaLabel]);
    const { triggerProps } = useToggleAction.useToggleAction({
        onToggle: onToggle
    });
    const { tooltipContent, tooltipProps } = useTooltip.useTooltip({
        text: accessibleLabel,
        isDisabled: isDisabled || isTooltipDisabled
    });
    const elementDetails = {
        type: 'div'
    };
    const splitSwatch = () => {
        return (jsxRuntime.jsxs(preact.Fragment, { children: [jsxRuntime.jsx("div", { className: alphaHalfStyles, style: isDisabled ? {} : { backgroundColor: swatchData.getSixDigitHex(value) } }), jsxRuntime.jsx("div", { className: alphaHalfStyles, style: isDisabled ? {} : { backgroundColor: value } })] }));
    };
    const noFill = () => {
        return (jsxRuntime.jsx("div", { className: isDisabled ? noFillDisabledStyles : noFillStyles, children: jsxRuntime.jsx(NoFill, {}) }));
    };
    const valueSwatch = () => {
        return (jsxRuntime.jsx("div", { className: classNames.classNames([
                smallButtonContainerStyles,
                isButton && buttonStyles,
                !isButton && swatchStyles
            ]), style: isNoFill || isDisabled || hasAlpha ? {} : { backgroundColor: value }, children: isNoFill ? noFill() : hasAlpha ? splitSwatch() : undefined }));
    };
    const noValueSwatch = () => {
        return (jsxRuntime.jsx("div", { className: classNames.classNames([
                !isButton && swatchStyles,
                isDisabled ? noFillDisabledStyles : noFillHoverStyles
            ]), children: jsxRuntime.jsx(CheckboxCrossed.SvgCheckboxCrossed, {}) }));
    };
    const swatchContents = () => {
        return hasValue ? valueSwatch() : noValueSwatch();
    };
    const buttonContents = () => {
        return (jsxRuntime.jsx(IconButton.IconButton, { "aria-label": accessibleLabel, "aria-roledescription": ariaRoleDescription ?? colorSwatchTranslated, ...swatchAriaProps, isDisabled: isDisabled, ref: ref, size: "sm", testId: testId, variant: "ghost", onAction: () => onToggle?.({ value: isSelected }), children: swatchContents() }));
    };
    return isButton ? (buttonContents()) : (jsxRuntime.jsxs(preact.Fragment, { children: [jsxRuntime.jsx(BareButton.BareButton, { ref: ref, elementDetails: elementDetails, ...mergeProps.mergeProps(triggerProps, globalEventProps, hoverProps, activeProps, globalEventProps, tooltipProps), testId: testId, isDisabled: isDisabled, "aria-label": accessibleLabel, "aria-roledescription": ariaRoleDescription ?? colorSwatchTranslated, ...swatchAriaProps, classNames: classes, onAction: () => onToggle?.({ value: isSelected }), children: swatchContents() }), tooltipContent] }));
});
ColorSwatch.displayName = 'ColorSwatch';

exports.ColorSwatch = ColorSwatch;
