// @oracle/oraclejet-preact/cjs/ComponentMessageContainer-0a029bed.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
require('preact/hooks');
require('./Flex-6ca216a7.js');


require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js');
require('./logger-2b636482.js');
require('preact');
var useTestId = require('./useTestId-9093a54b.js');
require('./LayerHost-aee0e070.js');
require('preact/compat');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-ae48fed6.js');

require('./IconButton-217314e6.js');

require('./MessageFormattingUtils-539d75b0.js');
var MessagesManager = require('./MessagesManager-5f1e36da.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var classNames = require('./classNames-c14c6ef3.js');
var clientHints = require('./clientHints-c40c5250.js');
require('./TransitionGroup-b239d98f.js');
require('./MessagesContext-4e939750.js');
var HiddenAccessible = require('./HiddenAccessible-087eb49a.js');
var UNSAFE_ComponentMessage_themes_redwood_ComponentMessageTheme = require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js');
var UNSAFE_UserAssistance_themes_redwood_UserAssistanceTheme = require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js');
var UNSAFE_ComponentMessage_themes_ComponentMessageStyles_css = require('./UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js');

/**
 * The component that renders an individual message for inline component messaging.
 */
function ComponentMessage({ detail, fieldLabel, severity = 'error', testId }) {
    const { classes } = useComponentTheme.useComponentTheme(UNSAFE_ComponentMessage_themes_redwood_ComponentMessageTheme.ComponentMessageRedwoodTheme, {
        severity: severity ?? 'none'
    });
    const { baseTheme } = useComponentTheme.useComponentTheme(UNSAFE_UserAssistance_themes_redwood_UserAssistanceTheme.UserAssistanceRedwoodTheme);
    const styleClasses = classNames.classNames([classes, baseTheme]);
    const testIdProps = useTestId.useTestId(testId);
    // aria-live assertive works on ios VoiceOver, but not on Jaws for our non-input-element form components like checkbox-set and input-date-mask,
    // and role='alert' works on Jaws but not ios Voiceover due to VO bug.
    // JET-65574 error message is not read on focus in voiceover due to VoiceOver bug with role: 'alert'.
    // JET-72375 Accessibility - Checkbox, Error message is not read by JAWS due to aria-live: 'assertive'.
    return (jsxRuntime.jsxs("div", { class: styleClasses, ...(isIosSafari() ? { 'aria-live': 'assertive' } : { role: 'alert' }), "aria-atomic": "true", ...testIdProps, children: [fieldLabel && jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { children: fieldLabel }), MessagesManager.isSeverityIconNeeded(severity) && (jsxRuntime.jsx(MessagesManager.MessageStartIcon, { severity: severity, testId: testId ? `${testId}-icon` : undefined, variant: "inline" })), jsxRuntime.jsx(MessagesManager.MessageSummary, { testId: testId ? `${testId}-summary` : undefined, text: detail, variant: "inline" })] }));
}
/**
 * Helper function to determine if the platform is iOS and the Browser is Safari.
 * Exported for usage in local tests.
 * @returns true if the platform is iOS, false otherwise
 */
function isIosSafari() {
    return clientHints.getClientHints().platform === 'ios' && clientHints.getClientHints().browser === 'safari';
}

/**
 * Converts the messages data into Item
 *
 * @param messages The messages data passed down as prop
 * @returns The messages data as Item
 */
function generateMessagesData(messages) {
    return messages.map((message, index) => {
        return {
            key: index,
            data: {
                closeAffordance: 'off',
                severity: message.severity || 'error',
                detail: message.detail
            }
        };
    });
}
function ComponentMessageContainer({ fieldLabel, messages = [], testId }) {
    const testIdProps = useTestId.useTestId(testId);
    return (jsxRuntime.jsx("div", { class: UNSAFE_ComponentMessage_themes_ComponentMessageStyles_css.componentMessageContainerBase, ...testIdProps, children: jsxRuntime.jsx(MessagesManager.MessagesManager, { data: generateMessagesData(messages), children: ({ item }) => (jsxRuntime.jsx(ComponentMessage, { detail: item.data.detail, fieldLabel: fieldLabel, severity: item.data.severity, testId: testId ? `${testId}-message-${item.key}` : undefined }, item.key)) }) }));
}

exports.ComponentMessage = ComponentMessage;
exports.ComponentMessageContainer = ComponentMessageContainer;
