// @oracle/oraclejet-preact/cjs/LayerHost-aee0e070.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var index = require('./index-15e13649.js');
var preact = require('preact');
var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var UNSAFE_Layer_themes_LayerHostStyles_css = require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Default environment created for the application
 */
const DefaultEnvironment = {
    user: {
        locale: document.documentElement.getAttribute('lang') || 'en',
        direction: document.documentElement.getAttribute('dir')?.toLowerCase() === 'rtl' ? 'rtl' : 'ltr',
        forcedColors: window.matchMedia?.('(forced-colors: active)')?.matches ? 'active' : 'none'
    },
    theme: index.t,
    colorScheme: 'light',
    scale: 'lg',
    currentBgColor: undefined,
    mode: 'production',
    density: 'standard'
};
const EnvironmentContext = preact.createContext(DefaultEnvironment);

const LayerContext = preact.createContext({});

const LayerHost = compat.forwardRef((_props, ref) => {
    return jsxRuntime.jsx("div", { id: "__root_layer_host", role: "presentation", ref: ref, class: UNSAFE_Layer_themes_LayerHostStyles_css.baseStyle });
});
LayerHost.displayName = 'Forwarded<LayerHost>';

exports.DefaultEnvironment = DefaultEnvironment;
exports.EnvironmentContext = EnvironmentContext;
exports.LayerContext = LayerContext;
exports.LayerHost = LayerHost;
