// @oracle/oraclejet-preact/cjs/NavigationListLinkItem-984bf9b0.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var Close = require('./Close-8eecf245.js');
var EmbeddedIconButton = require('./EmbeddedIconButton-a5a8938b.js');
var hooks = require('preact/hooks');
var useHover = require('./useHover-49b0430d.js');
var useTranslationBundle = require('./useTranslationBundle-18b7bf8b.js');
var Text = require('./Text-436e8b56.js');
var usePress = require('./usePress-886180e4.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var classNames = require('./classNames-c14c6ef3.js');
var useInteractionStyle = require('./useInteractionStyle-c203a8a0.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var preact = require('preact');
var mergeInterpolations = require('./mergeInterpolations-6727b536.js');
var flexitem = require('./flexitem-fee13e26.js');
var UNSAFE_NavigationListCommon_themes_redwood_NavigationListItemTheme = require('./UNSAFE_NavigationListCommon/themes/redwood/NavigationListItemTheme.js');
var useUser = require('./useUser-f900ddf1.js');
var Flex = require('./Flex-6ca216a7.js');
var WarningS = require('./WarningS-2243c6d7.js');
var ErrorS = require('./ErrorS-14a49cb6.js');
var InformationS = require('./InformationS-b6024498.js');
var SuccessS = require('./SuccessS-7e6dbebc.js');
var Badge = require('./Badge-0e522f12.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Context used to pass navlist information without having to pass it to navlist children props.
 * We want to communicate information down to any interested navlist item children.
 */
const NavigationListContext = preact.createContext(null);

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Utility hook for consuming the NavigationListContext
 *
 * @returns The value of NavigationListContext provider
 */
function useNavigationListContext() {
    return hooks.useContext(NavigationListContext);
}

const flexItemStyleInterpolations = mergeInterpolations.mergeInterpolations([
    ...Object.values(flexitem.flexitemInterpolations)
]);
const { class: labelContainerFlexItemClasses, ...labelContainerFlexItemStyle } = flexItemStyleInterpolations({
    alignSelf: 'center'
});
function useNavigationListItem({ itemKey }) {
    const { onSelectionChange, selection, currentKey, showFocusRing, navigationListItemPrefix } = useNavigationListContext();
    //Selection and focus state
    const isCurrent = currentKey === itemKey;
    const isSelected = selection === itemKey;
    //Styling
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle, applyActiveStyle } = useInteractionStyle.useInteractionStyle();
    const { direction } = useUser.useUser();
    const isRtl = direction === 'rtl';
    const { classes: listItemClasses, styles: { labelContainerClasses } } = useComponentTheme.useComponentTheme(UNSAFE_NavigationListCommon_themes_redwood_NavigationListItemTheme.NavigationListItemRedwoodTheme, {
        focusRing: showFocusRing ? 'isFocusRing' : 'notFocusRing',
        current: isCurrent ? 'isCurrent' : 'notCurrtent',
        selected: isSelected ? 'isSelected' : 'notSelected',
        direction: isRtl ? 'isRtl' : 'notRtl',
        hybridHover: applyHoverStyle ? 'isHybridHover' : 'notHybridHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        active: applyActiveStyle ? 'isActive' : 'notActive'
    });
    //Item handlers
    const handlePress = hooks.useCallback(() => {
        onSelectionChange?.({ value: itemKey, reason: 'pointer' });
    }, [itemKey, onSelectionChange]);
    //Legacy navlist makes selection on keyDown, but since preact tabbar made selection after keyUp, will do the same
    //for preact navlist
    const { pressProps } = usePress.usePress(handlePress);
    return {
        itemId: navigationListItemPrefix + itemKey,
        itemHandlers: mergeProps.mergeProps(interactionProps, pressProps /*,onFocusProps*/),
        itemClasses: classNames.classNames([listItemClasses]),
        isSelected,
        labelContainerClasses: classNames.classNames([labelContainerClasses, labelContainerFlexItemClasses]),
        labelContainerStyle: labelContainerFlexItemStyle
    };
}

const onClickLink = (event) => {
    event.preventDefault();
};
/*TODO: JET-58535. Add ripple effect when touching navigationListItem. */
function BaseNavigationListItem({ itemKey, label, badge, metadata, severity = 'none', removeIcon, href, 'aria-controls': ariaControls }) {
    const itemRef = hooks.useRef(null);
    //Logic of the component handled here
    const { itemId, itemHandlers, itemClasses, isSelected, labelContainerClasses, labelContainerStyle } = useNavigationListItem({
        itemKey
    });
    const severityRequirement = !(severity === 'none' || severity === undefined);
    const IconComponent = (severityValue) => {
        switch (severityValue) {
            case 'error':
                return jsxRuntime.jsx(ErrorS.SvgErrorS, { color: "danger" });
            case 'warning':
                return jsxRuntime.jsx(WarningS.SvgWarningS, { color: "warning" });
            case 'info':
                return jsxRuntime.jsx(InformationS.SvgInformationS, { color: "info" });
            case 'confirmation':
                return jsxRuntime.jsx(SuccessS.SvgSuccessS, { color: "success" });
            default:
                return null;
        }
    };
    const { styles: { iconLabelContainer, iconContainer, navigationItemMetadata, navigationListItemLink } } = useComponentTheme.useComponentTheme(UNSAFE_NavigationListCommon_themes_redwood_NavigationListItemTheme.NavigationListItemRedwoodTheme);
    const navigationItemMatadata = classNames.classNames([navigationItemMetadata]);
    const navigationItemLinkClass = classNames.classNames([navigationListItemLink]);
    const endIconContainerClasses = classNames.classNames([iconLabelContainer, iconContainer]);
    const { styles: { navigationlistRemoveIcon } } = useComponentTheme.useComponentTheme(UNSAFE_NavigationListCommon_themes_redwood_NavigationListItemTheme.NavigationListItemRedwoodTheme);
    const removeClass = classNames.classNames([navigationlistRemoveIcon]);
    const isComplexItem = metadata || severityRequirement || badge;
    const isLinkItem = href != null;
    const labelRenderer = () => {
        return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsxRuntime.jsx(Text.Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) }), jsxRuntime.jsx(Flex.Flex, { gap: "2x", align: "center", justify: "end", flex: "1 1 auto", children: removeIcon && !isLinkItem && jsxRuntime.jsx("span", { class: removeClass, children: removeIcon }) })] }));
    };
    const complexItemRenderer = () => {
        return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [label && (jsxRuntime.jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsxRuntime.jsx(Text.Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) })), (metadata || severityRequirement || badge || removeIcon) && (jsxRuntime.jsxs(Flex.Flex, { gap: "2x", align: "center", justify: "end", flex: "1 1 auto", children: [' ', badge && (jsxRuntime.jsx("span", { class: endIconContainerClasses, children: jsxRuntime.jsx(Badge.Badge, { size: "md", variant: "neutralSubtle", children: badge }) })), metadata && (jsxRuntime.jsx("span", { class: navigationItemMatadata, children: jsxRuntime.jsx(Text.Text, { children: '( ' + metadata + ' )' }) })), severityRequirement && (jsxRuntime.jsx("span", { class: endIconContainerClasses, children: IconComponent(severity) })), removeIcon && !isLinkItem && jsxRuntime.jsx("span", { class: removeClass, children: removeIcon })] }))] }));
    };
    const navlistItemRenderer = () => {
        return (jsxRuntime.jsx("div", { role: "tab", ...itemHandlers, "data-oj-key": itemKey, id: itemId, class: itemClasses, "aria-selected": isSelected, ref: itemRef, "aria-controls": ariaControls, children: isComplexItem ? complexItemRenderer() : labelRenderer() }));
    };
    const navlistLinkItemRenderer = () => {
        return (jsxRuntime.jsx("div", { ref: itemRef, ...itemHandlers, class: itemClasses, children: jsxRuntime.jsx("a", { tabindex: -1, role: "tab", "data-oj-key": itemKey, id: itemId, class: navigationItemLinkClass, "aria-selected": isSelected, "aria-controls": ariaControls, href: href, onClick: onClickLink, children: isComplexItem ? complexItemRenderer() : labelRenderer() }) }));
    };
    return isLinkItem ? navlistLinkItemRenderer() : navlistItemRenderer();
}

function RemovableNavigationListItem({ itemKey, label, badge, metadata, severity }) {
    const { selection, currentKey, onRemove } = useNavigationListContext();
    const isSelected = selection === itemKey;
    const isCurrent = currentKey === itemKey;
    const handleRemove = hooks.useCallback(() => {
        onRemove?.({ value: itemKey });
    }, [onRemove, itemKey]);
    const { hoverProps, isHover } = useHover.useHover({});
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    //instead of specifying flex: '1 0 auto' in cssProps var
    return (jsxRuntime.jsx("div", { ...hoverProps, children: jsxRuntime.jsx(BaseNavigationListItem, { itemKey: itemKey, label: label, badge: badge, metadata: metadata, severity: severity, removeIcon: jsxRuntime.jsx("span", { "data-oj-navigationlist-item-remove-icon": true, children: jsxRuntime.jsx(EmbeddedIconButton.EmbeddedIconButton, { "aria-label": translations.tabBarNavigationList_removeCueText(), size: "xs", onAction: handleRemove, children: (isHover || isSelected || isCurrent) && jsxRuntime.jsx(Close.SvgClose, {}) }) }) }) }));
}

/**
 * NavigationListLinkItem is used to render links in a NavigationList. It allows to view the browser's built in link context menu to open the link in new browser tab.
 */
function NavigationListLinkItem({ itemKey, badge, metadata, severity, href, label, 'aria-controls': ariaControls }) {
    return (jsxRuntime.jsx(BaseNavigationListItem, { "aria-controls": ariaControls, itemKey: itemKey, label: label, badge: badge, metadata: metadata, severity: severity, href: href }));
}

exports.BaseNavigationListItem = BaseNavigationListItem;
exports.NavigationListContext = NavigationListContext;
exports.NavigationListLinkItem = NavigationListLinkItem;
exports.RemovableNavigationListItem = RemovableNavigationListItem;
