// @oracle/oraclejet-preact/cjs/padding-398f986c.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var size = require('./size-0717c0bd.js');
var arrayUtils = require('./arrayUtils-7d8dcfc3.js');

const paddingInterpolations = {
    padding: ({ padding }) => {
        if (padding === undefined) {
            return {};
        }
        else {
            const [topPadding, rightPadding = topPadding, bottomPadding = topPadding, leftPadding = rightPadding] = arrayUtils.coerceArray(padding);
            return {
                padding: `${size.sizeToCSS(topPadding)} ${size.sizeToCSS(rightPadding)} 
        ${size.sizeToCSS(bottomPadding)} ${size.sizeToCSS(leftPadding)}`
            };
        }
    },
    paddingBlockStart: ({ paddingBlockStart }) => paddingBlockStart === undefined
        ? {}
        : {
            paddingBlockStart: `${size.sizeToCSS(paddingBlockStart)}`
        },
    paddingBlockEnd: ({ paddingBlockEnd }) => paddingBlockEnd === undefined
        ? {}
        : {
            paddingBlockEnd: `${size.sizeToCSS(paddingBlockEnd)}`
        },
    paddingInlineStart: ({ paddingInlineStart }) => paddingInlineStart === undefined
        ? {}
        : {
            paddingInlineStart: `${size.sizeToCSS(paddingInlineStart)}`
        },
    paddingInlineEnd: ({ paddingInlineEnd }) => paddingInlineEnd === undefined
        ? {}
        : {
            paddingInlineEnd: `${size.sizeToCSS(paddingInlineEnd)}`
        }
};

exports.paddingInterpolations = paddingInterpolations;
