// @oracle/oraclejet-preact/cjs/swatchData-17c1d335.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var colorUtils = require('./colorUtils-01f3caa2.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Convert a color keyword to hex value.
 * If the keyword is not valid, return an empty string.
 * @param value
 * @returns
 */
function convertKeywordToHex(value) {
    try {
        return colorUtils.getNamedColor(value);
    }
    catch {
        return '';
    }
}
/**
 * Returns true if the value is 'transparent'
 * @param value
 * @returns
 */
function isTransparent(value) {
    if (value && value.length && value.toUpperCase() === 'TRANSPARENT') {
        return true;
    }
    else {
        return false;
    }
}
/**
 * Check if the value is a valid named color.
 * @param value
 * @returns
 */
function isNamedColor(value) {
    try {
        return colorUtils.getNamedColor(value).length > 0;
    }
    catch {
        return false;
    }
}
/**
 * Parse a color value and return a hex string or 'TRANSPARENT'.
 * ie, 'blue' => '#0000FF'
 * If not valid keyword, returns ''
 * @param value
 * @returns
 */
function parseColorValue(value) {
    if (!value) {
        return '';
    }
    value = value.trim();
    // check for 'transparent' or 'TRANSPARENT' a valid CSS keyword
    if (isTransparent(value)) {
        return value;
    }
    // check for named color
    if (isNamedColor(value)) {
        return convertKeywordToHex(value).toUpperCase();
    }
    // check for string '#xxxxxx' with '#' up to 3, 4,  6, 8 hex digits all valid css colors
    if (/^([#]([0-9A-Fa-f]{3}|[0-9A-Fa-f]{4}|[0-9A-Fa-f]{6}|[0-9A-Fa-f]{8}))$/.test(value)) {
        return value.toUpperCase();
    }
    return '';
}
/** Get the ColorValue from a string
 */
function getColorValue(value) {
    return parseColorValue(value);
}
/**
 * Can convert array to all hex values and 'transparent'
 * as desired.
 * @param palette
 * @returns
 */
function convertPalette(palette) {
    return palette.map((x) => getColorValue(x));
}
/**
 * More 'transparent' to first position
 * @param palette
 * @returns
 */
function sortPalette(palette) {
    const uniqueArray = [...new Set(palette)];
    const p = uniqueArray.map((str) => str?.toLowerCase());
    const i = p.indexOf('transparent');
    switch (i) {
        case 0:
        case -1:
            return p;
        default: {
            const f = p.filter((str) => str?.toLowerCase() !== 'transparent');
            return ['transparent', ...f];
        }
    }
}
/**
 * Return true if the color params represent 2 different colors.
 */
const colorsAreDifferent = (c1, c2, isExpandable = false) => {
    const c1_empty = !c1 || !c1.length;
    const c2_empty = !c2 || !c2.length;
    // If both colors are empty, return false.
    if (c1_empty && c2_empty) {
        return false;
    }
    // If one is empty and the other isn't, return true.
    if ((c1_empty && !c2_empty) || (!c1_empty && c2_empty)) {
        return true;
    }
    if (isExpandable &&
        getExpandedColorValue(c1, true) !== getExpandedColorValue(c2, true)) {
        return true;
    }
    if (c1 !== c2) {
        return true;
    }
    return false;
};
/**
 * Return true if the color value is a valid hex color or 'TRANSPARENT'
 * Valid values to type include
 * - '#xxxxxx'
 * - '#xxxxxxxx'
 * - 'TRANSPARENT'
 * - 'transparent'
 * - named color (e.g. 'blue', 'red', etc.)
 * @param value
 * @returns
 */
const isValidColorValue = (value) => {
    const rgb = value ? parseColorValue(value) : '';
    return rgb ? rgb.length === 7 || rgb.length === 9 || isTransparent(rgb) : false;
};
/**
 * Convert a color value to an 8 digit hex value.
 * If the value is a valid hex color, it will return the value with an alpha channel of 'FF'.
 * If the value is not valid, it will return an empty string.
 * @param value
 * @returns
 */
const getEightDigitHex = (value) => {
    // First check if this is a complete value and return that
    if (isValidColorValue(value)) {
        const retVal = getColorValue(value);
        if (retVal.length === 9) {
            return retVal;
        }
        else if (retVal.length === 7) {
            return (retVal + 'FF'); // add alpha channel
        }
    }
    return ''; // return empty string for invalid values
};
/**
 * Convert a color value to an 6 digit hex value.
 * If any other alpha, return empty string.
 * @param value
 * @returns
 */
const getSixDigitHex = (value) => {
    // First check if this is a complete value and return that
    if (isValidColorValue(value)) {
        const retVal = getColorValue(value);
        if (/^([#]([0-9A-Fa-f]{6}))$/.test(retVal)) {
            return retVal;
        }
        else if (/^([#]([0-9A-Fa-f]{8}))$/.test(retVal)) {
            return retVal.substring(0, 7);
        }
    }
    return ''; // return empty string for invalid values
};
/**
 * Returns true if value is an 8 digit hex value
 * @param value
 * @returns
 */
const isEightDigitHex = (value) => {
    return /^([#]([0-9A-Fa-f]{8}))$/.test(value);
};
// /**
//  * Returns true if value is an 6 digit hex value
//  * @param value
//  * @returns
//  */
// const isSixDigitHex = (value: ColorValue): boolean => {
//   return (/^([#]([0-9A-Fa-f]{6}))$/.test(value));
// }
/**
 * Returns the intended color value for a partially specified color value.
 * Can request 8 or 6 digit hex rgb return values.  Note that 'transparent' is
 * always returned unconverted as it is a valid css value and requested to stay
 * untouched for ux reasons.  Finds first valid hex substring and applies
 * heuristics.
 *
 * Returns '' if the value is not valid.
 * @param value
 * @returns
 */
const getExpandedColorValue = (value, hasEightDigitHex = false) => {
    if (!value) {
        return '';
    }
    if (isTransparent(value)) {
        return value;
    }
    // Return eight digit hex results
    if (hasEightDigitHex) {
        // First check if this is a complete value and return that
        if (isValidColorValue(value)) {
            return getEightDigitHex(getColorValue(value));
        }
        // Now implement desired behavior for partially specified color values
        // strip leading hash
        const nohash = value[0] === '#' ? value.substring(1) : value;
        // Ux request is to strip all non-hex characters first , upper-case, and then apply heuristics
        const val = nohash?.match(/[0-9A-Fa-f]+/)?.[0].toUpperCase() ?? '';
        // A -> #AAAAAAFF
        if (/^([0-9A-F]{1})$/.test(val)) {
            return ('#' + val + val + val + val + val + val + 'FF');
        }
        // AB -> #ABABABFF
        if (/^([0-9A-F]{2})$/.test(val)) {
            return ('#' + val[0] + val[1] + val[0] + val[1] + val[0] + val[1] + 'FF');
        }
        // ABC -> #AABBCCFF
        if (/^([0-9A-F]{3})$/.test(val)) {
            return ('#' + val[0] + val[0] + val[1] + val[1] + val[2] + val[2] + 'FF');
        }
        // ABCF -> #AABBCCFF
        if (/^([0-9A-F]{4})$/.test(val)) {
            return ('#' +
                val[0] +
                val[0] +
                val[1] +
                val[1] +
                val[2] +
                val[2] +
                val[3] +
                val[3]);
        }
        // ABCDEF -> #ABCDEFFF
        if (/^([0-9A-F]{6})$/.test(val)) {
            return ('#' + val[0] + val[1] + val[2] + val[3] + val[4] + val[5] + 'FF');
        }
        // ABCDEFFF -> #ABCDEFFF
        if (/^([0-9A-F]{8})$/.test(val)) {
            return ('#' +
                val[0] +
                val[1] +
                val[2] +
                val[3] +
                val[4] +
                val[5] +
                val[6] +
                val[7]);
        }
    }
    else {
        // 6 digit hex path return
        // First check if this is a complete value and return that
        if (isValidColorValue(value)) {
            return getSixDigitHex(getColorValue(value));
        }
        // strip leading hash
        const nohash = value[0] === '#' ? value.substring(1) : value;
        // Ux request is to strip all non-hex characters first , upper-case, and then apply heuristics
        const val = nohash?.match(/[0-9A-Fa-f]+/)?.[0].toUpperCase() ?? '';
        // A -> #AAAAAA
        if (/^([0-9A-F]{1})$/.test(val)) {
            return ('#' + val + val + val + val + val + val);
        }
        // AB -> #ABABAB
        if (/^([0-9A-F]{2})$/.test(val)) {
            return ('#' + val[0] + val[1] + val[0] + val[1] + val[0] + val[1]);
        }
        // ABC -> #AABBCC
        if (/^([0-9A-F]{3})$/.test(val)) {
            return ('#' + val[0] + val[0] + val[1] + val[1] + val[2] + val[2]);
        }
        // ABCDEF -> #ABCDEF
        if (/^([0-9A-F]{6})$/.test(val)) {
            return ('#' + val[0] + val[1] + val[2] + val[3] + val[4] + val[5]);
        }
    }
    return '';
};

/**
 * unique colormap const
 */
const uniqueColorMap = {
    transparent: 'transparent',
    aliceblue: '#f0f8ff',
    antiquewhite: '#faEbd7',
    aqua: '#00ffff',
    aquamarine: '#7fffd4',
    azure: '#f0ffff',
    beige: '#f5f5dc',
    bisque: '#ffE4c4',
    black: '#000000',
    blanchedalmond: '#ffEbcd',
    blue: '#0000ff',
    blueviolet: '#8a2bE2',
    brown: '#a52a2a',
    burlywood: '#dEb887',
    cadetblue: '#5f9Ea0',
    chartreuse: '#7fff00',
    chocolate: '#d2691E',
    coral: '#ff7f50',
    cornflowerblue: '#6495Ed',
    cornsilk: '#fff8dc',
    crimson: '#dc143c',
    darkblue: '#00008b',
    darkcyan: '#008b8b',
    darkgoldenrod: '#b8860b',
    darkgray: '#a9a9a9',
    darkgreen: '#006400',
    darkkhaki: '#bdb76b',
    darkmagenta: '#8b008b',
    darkolivegreen: '#556b2f',
    darkorange: '#ff8c00',
    darkorchid: '#9932cc',
    darkred: '#8b0000',
    darksalmon: '#E9967a',
    darkseagreen: '#8fbc8f',
    darkslateblue: '#483d8b',
    darkslategray: '#2f4f4f',
    darkturquoise: '#00cEd1',
    darkviolet: '#9400d3',
    deeppink: '#ff1493',
    deepskyblue: '#00bfff',
    dimgray: '#696969',
    dodgerblue: '#1E90ff',
    firebrick: '#b22222',
    floralwhite: '#fffaf0',
    forestgreen: '#228b22',
    gainsboro: '#dcdcdc',
    ghostwhite: '#f8f8ff',
    gold: '#ffd700',
    goldenrod: '#daa520',
    gray: '#808080',
    green: '#008000',
    greenyellow: '#adff2f',
    honeydew: '#f0fff0',
    hotpink: '#ff69b4',
    indianred: '#cd5c5c',
    indigo: '#4b0082',
    ivory: '#fffff0',
    khaki: '#f0E68c',
    lavender: '#E6E6fa',
    lavenderblush: '#fff0f5',
    lawngreen: '#7cfc00',
    lemonchiffon: '#fffacd',
    lightblue: '#add8E6',
    lightcoral: '#f08080',
    lightcyan: '#E0ffff',
    lightgoldenrodyellow: '#fafad2',
    lightgray: '#d3d3d3',
    lightgreen: '#90EE90',
    lightpink: '#ffb6c1',
    lightsalmon: '#ffa07a',
    lightseagreen: '#20b2aa',
    lightskyblue: '#87cEfa',
    lightslategray: '#778899',
    lightsteelblue: '#b0c4dE',
    lightyellow: '#ffffE0',
    lime: '#00ff00',
    limegreen: '#32cd32',
    linen: '#faf0E6',
    magenta: '#ff00ff',
    maroon: '#800000',
    mediumaquamarine: '#66cdaa',
    mediumblue: '#0000cd',
    mediumorchid: '#ba55d3',
    mediumpurple: '#9370db',
    mediumseagreen: '#3cb371',
    mediumslateblue: '#7b68EE',
    mediumspringgreen: '#00fa9a',
    mediumturquoise: '#48d1cc',
    mediumvioletred: '#c71585',
    midnightblue: '#191970',
    mintcream: '#f5fffa',
    mistyrose: '#ffE4E1',
    moccasin: '#ffE4b5',
    navajowhite: '#ffdEad',
    navy: '#000080',
    oldlace: '#fdf5E6',
    olive: '#808000',
    olivedrab: '#6b8E23',
    orange: '#ffa500',
    orangered: '#ff4500',
    orchid: '#da70d6',
    palegoldenrod: '#EEE8aa',
    palegreen: '#98fb98',
    paleturquoise: '#afEEEE',
    palevioletred: '#db7093',
    papayawhip: '#ffEfd5',
    peachpuff: '#ffdab9',
    peru: '#cd853f',
    pink: '#ffc0cb',
    plum: '#dda0dd',
    powderblue: '#b0E0E6',
    purple: '#800080',
    rebeccapurple: '#663399',
    red: '#ff0000',
    rosybrown: '#bc8f8f',
    royalblue: '#4169E1',
    saddlebrown: '#8b4513',
    salmon: '#fa8072',
    sandybrown: '#f4a460',
    seagreen: '#2E8b57',
    seashell: '#fff5EE',
    sienna: '#a0522d',
    silver: '#c0c0c0',
    skyblue: '#87cEEb',
    slateblue: '#6a5acd',
    slategray: '#708090',
    snow: '#fffafa',
    springgreen: '#00ff7f',
    steelblue: '#4682b4',
    tan: '#d2b48c',
    teal: '#008080',
    thistle: '#d8bfd8',
    tomato: '#ff6347',
    turquoise: '#40E0d0',
    violet: '#EE82EE',
    wheat: '#f5dEb3',
    white: '#ffffff',
    whitesmoke: '#f5f5f5',
    yellow: '#ffff00',
    yellowgreen: '#9acd32'
};
const defaultRedwoodNoFillPaletteValues = [
    'transparent',
    '#FFFFFF',
    '#F4F4F4',
    '#D5D5D5',
    '#909090',
    '#4B4B4B',
    '#000000',
    '#E6B0AA',
    '#EDBB99',
    '#F9E79F',
    '#A9DFBF',
    '#AED6F1',
    '#D7BDE2',
    '#F9B3E2',
    '#CD6155',
    '#F98C45',
    '#F4D03F',
    '#52BE80',
    '#5DADE2',
    '#AF7AC5',
    '#E55FB9',
    '#81261D',
    '#853500',
    '#A3850A',
    '#19713E',
    '#246A98',
    '#6C3E7F',
    '#932970'
];
const defaultRedwoodPaletteValues = [
    '#FFFFFF',
    '#F4F4F4',
    '#D5D5D5',
    '#909090',
    '#6E6E6E',
    '#4B4B4B',
    '#000000',
    '#E6B0AA',
    '#EDBB99',
    '#F9E79F',
    '#A9DFBF',
    '#AED6F1',
    '#D7BDE2',
    '#F9B3E2',
    '#CD6155',
    '#F98C45',
    '#F4D03F',
    '#52BE80',
    '#5DADE2',
    '#AF7AC5',
    '#E55FB9',
    '#81261D',
    '#853500',
    '#A3850A',
    '#19713E',
    '#246A98',
    '#6C3E7F',
    '#932970'
];
Object.values(uniqueColorMap);

exports.colorsAreDifferent = colorsAreDifferent;
exports.convertPalette = convertPalette;
exports.defaultRedwoodNoFillPaletteValues = defaultRedwoodNoFillPaletteValues;
exports.defaultRedwoodPaletteValues = defaultRedwoodPaletteValues;
exports.getColorValue = getColorValue;
exports.getEightDigitHex = getEightDigitHex;
exports.getExpandedColorValue = getExpandedColorValue;
exports.getSixDigitHex = getSixDigitHex;
exports.isEightDigitHex = isEightDigitHex;
exports.isValidColorValue = isValidColorValue;
exports.parseColorValue = parseColorValue;
exports.sortPalette = sortPalette;
exports.uniqueColorMap = uniqueColorMap;
