// @oracle/oraclejet-preact/cjs/UNSAFE_Theme.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Theme = require('./Theme-f06687af.js');
var utils = require('./utils-a66430fc.js');
var mergeDeepRight = require('./mergeDeepRight-2767a52a.js');
require('./themeContract.css-a53fd740.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./_curry2-c15d89cd.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_isObject-28636267.js');



exports.COLORSCHEME_DEPENDENT_CLASS = Theme.COLORSCHEME_DEPENDENT_CLASS;
exports.DARK_CLASS = Theme.DARK_CLASS;
exports.DENSITY_COMPACT_CLASS = Theme.DENSITY_COMPACT_CLASS;
exports.DENSITY_COMPACT_SELECTORS = Theme.DENSITY_COMPACT_SELECTORS;
exports.DENSITY_DEPENDENT_CLASS = Theme.DENSITY_DEPENDENT_CLASS;
exports.DENSITY_DEPENDENT_SELECTOR = Theme.DENSITY_DEPENDENT_SELECTOR;
exports.DENSITY_SELECTORS = Theme.DENSITY_SELECTORS;
exports.DENSITY_STANDARD_CLASS = Theme.DENSITY_STANDARD_CLASS;
exports.DENSITY_STANDARD_SELECTORS = Theme.DENSITY_STANDARD_SELECTORS;
exports.INVERT_CLASS = Theme.INVERT_CLASS;
exports.LIGHT_CLASS = Theme.LIGHT_CLASS;
exports.ROOT_SELECTOR = Theme.ROOT_SELECTOR;
exports.SCALE_DEFAULT_SELECTORS = Theme.SCALE_DEFAULT_SELECTORS;
exports.SCALE_DEPENDENT_CLASS = Theme.SCALE_DEPENDENT_CLASS;
exports.SCALE_DEPENDENT_SELECTOR = Theme.SCALE_DEPENDENT_SELECTOR;
exports.SCALE_DEPENDENT_SELECTORS = Theme.SCALE_DEPENDENT_SELECTORS;
exports.SCALE_LG_CLASS = Theme.SCALE_LG_CLASS;
exports.SCALE_LG_SELECTORS = Theme.SCALE_LG_SELECTORS;
exports.SCALE_MD_CLASS = Theme.SCALE_MD_CLASS;
exports.SCALE_MD_SELECTORS = Theme.SCALE_MD_SELECTORS;
exports.SCALE_SM_CLASS = Theme.SCALE_SM_CLASS;
exports.SCALE_SM_SELECTORS = Theme.SCALE_SM_SELECTORS;
exports.colorThemes = Theme.colorThemes;
exports.colorMix = utils.colorMix;
exports.densityUnits = utils.densityUnits;
exports.fallbackVar = utils.fallbackVar;
exports.max = utils.max;
exports.min = utils.min;
exports.rgb = utils.rgb;
exports.rgba = utils.rgba;
exports.xUnits = utils.xUnits;
exports.extendThemeContractVars = mergeDeepRight.extendThemeContractVars;
