// @oracle/oraclejet-preact/cjs/UNSAFE_FormLayout.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var vanillaExtractDynamic_esm = require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var FormContext = require('./FormContext-d1d05967.js');
var useFormContext = require('./useFormContext-a794f7b6.js');
var UNSAFE_FormLayout_themes_FormLayoutStyles_css = require('./UNSAFE_FormLayout/themes/FormLayoutStyles.css.js');
var LayoutStyles_css = require('./LayoutStyles.css-b4025421.js');
var UNSAFE_FormLayout_themes_redwood_FormLayoutTheme = require('./UNSAFE_FormLayout/themes/redwood/FormLayoutTheme.js');
var classNames = require('./classNames-c14c6ef3.js');
var useTestId = require('./useTestId-9093a54b.js');
require('./logger-2b636482.js');
require('preact');
require('preact/hooks');



require('./UNSAFE_FormLayout/themes/redwood/FormLayoutBaseTheme.css.js');
require('./UNSAFE_FormLayout/themes/redwood/FormLayoutVariants.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./LayerHost-aee0e070.js');
require('./index-15e13649.js');
require('preact/compat');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');


const FormLayout = ({ columns = 1, columnBehavior = 'responsive', columnSpan = 1, direction = 'row', isFullWidth, isReadonly: propIsReadonly, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, labelWrapping: propLabelWrapping, userAssistanceDensity: propUserAssistanceDensity, children, testId }) => {
    const { isDisabled, isFormLayout: formIsFormLayout, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, labelWrapping: formLabelWrapping, 
    // Components in a FormLayout should default to textAlign="start" in all configurations
    textAlign = 'start', userAssistanceDensity: formUserAssistanceDensity } = useFormContext.useFormContext();
    const testIdProps = useTestId.useTestId(testId);
    const layoutElementTestIdProps = useTestId.useTestId(testId ? `${testId}_layout` : undefined);
    const isReadonly = propIsReadonly ?? isFormReadonly;
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    const labelWrapping = propLabelWrapping ?? formLabelWrapping;
    // If we have the prop, pass it on, or if we are in a form layout, use the context value, otherwise default to 'efficient'
    const userAssistanceDensity = propUserAssistanceDensity ?? (formIsFormLayout ? formUserAssistanceDensity : 'efficient');
    // We need to set the localVars.minColumnCount and localVars.maxColumnCount so that we can correctly set the max width to the current column count
    const assignVarStyles = vanillaExtractDynamic_esm.assignInlineVars({
        [LayoutStyles_css.layoutLocalVars.maxColumnCount]: `${columns}`,
        [LayoutStyles_css.layoutLocalVars.minColumnCount]: `${columnBehavior === 'responsive' ? 1 : columns}`
    });
    const { baseTheme, variantClasses } = useComponentTheme.useComponentTheme(UNSAFE_FormLayout_themes_redwood_FormLayoutTheme.FormLayoutRedwoodTheme, {
        columnBehavior,
        direction
    });
    const { variantClasses: rootMaxWidthVariant } = useComponentTheme.useComponentTheme(UNSAFE_FormLayout_themes_redwood_FormLayoutTheme.FormLayoutRedwoodTheme, {
        labelEdge,
        fullWidth: isFullWidth ? 'isFullWidth' : 'notFullWidth'
    });
    return (jsxRuntime.jsx(FormContext.FormContext.Provider, { value: {
            isDisabled,
            isFormLayout: true,
            isReadonly,
            labelEdge,
            labelStartWidth,
            labelWrapping,
            textAlign,
            userAssistanceDensity
        }, children: jsxRuntime.jsx("div", { class: classNames.classNames([
                baseTheme,
                rootMaxWidthVariant,
                UNSAFE_FormLayout_themes_FormLayoutStyles_css.styles.rootWrapperStyle,
                LayoutStyles_css.layoutSpanStyles.layoutSpanColumn[columnSpan]
            ]), ...testIdProps, style: assignVarStyles, children: jsxRuntime.jsx("div", { class: variantClasses, ...layoutElementTestIdProps, children: children }) }) }));
};

exports.compoundVariants = UNSAFE_FormLayout_themes_FormLayoutStyles_css.compoundVariants;
exports.styles = UNSAFE_FormLayout_themes_FormLayoutStyles_css.styles;
exports.variants = UNSAFE_FormLayout_themes_FormLayoutStyles_css.variants;
exports.FormLayout = FormLayout;
