// @oracle/oraclejet-preact/cjs/UNSAFE_BarChart.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var utils = require('./utils-7c1b4abe.js');
var useUser = require('./useUser-f900ddf1.js');
var BarGroup = require('./BarGroup-234f35bf.js');
var dataStyleUtils = require('./dataStyleUtils-8c59d570.js');
var Chart = require('./Chart-376785c5.js');
require('./themeContract.css-a53fd740.js');
require('./utils-a66430fc.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./getLocale-8e17dc46.js');
require('./stringUtils-3e19c8af.js');
require('./labelUtils-a3deed57.js');
require('./textUtils-e779c845.js');
require('./accUtils-5c7d3255.js');
require('./clientHints-c40c5250.js');
require('./utils-ca441171.js');
require('./datatipUtils-d1ca81a1.js');
require('./clientUtils-a4619fcd.js');
require('./dimensionUtils-3919753d.js');
require('./scale-9e43a5f6.js');
require('preact/hooks');
require('./LayerHost-aee0e070.js');
require('./index-15e13649.js');
require('preact');
require('preact/compat');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./PRIVATE_Chart/themes/ChartStyles.css.js');

require('./classNames-c14c6ef3.js');
require('./colorUtils-c97cf90b.js');
require('./TrackResizeContainer-100be103.js');
require('./useSize-8115eae0.js');
require('./useResizeObserver-a9e97180.js');
require('./useAddBusyState-d19ae1fa.js');
require('./BusyStateContext-86f40d3c.js');
require('./useBusyStateContext-362eee2a.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./size-0717c0bd.js');
require('./useTextDimensions-c8b41697.js');
require('./useCssVars-cc025013.js');
require('./useTheme-09dfbb78.js');
require('./useColorScheme-23e4aab4.js');
require('./useScale-adc62f41.js');
require('./Gridlines-3bdce0b7.js');
require('./PRIVATE_PlotArea/themes/PlotAreaStyles.css.js');

require('./hooks/PRIVATE_useVisDnd/themes/VisDndStyles.css.js');

require('./useId-6c0eeb27.js');
require('./Axis-2d648884.js');
require('./PRIVATE_Axis/themes/AxisStyles.css.js');

require('./MultilineText-837e9afd.js');
require('./PRIVATE_Axis/themes/redwood/AxisTheme.js');
require('./PRIVATE_Axis/themes/redwood/AxisBaseTheme.css.js');

require('./PRIVATE_Axis/themes/redwood/AxisVariants.css.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./useComponentTheme-082fc8e4.js');
require('./logger-2b636482.js');
require('./useChartViewport-d2e8c46f.js');
require('./eventsUtils-d34f8647.js');
require('./usePinch-741e9343.js');
require('./usePointerGesture-0b9cc1b7.js');
require('./mergeProps-e3da7237.js');
require('./useChartNav-b079ab1e.js');
require('./layoutUtils-e52005dc.js');
require('./useVisHover-ad36112d.js');
require('./useSelectDrill-b818e267.js');
require('./useChartDataCursor-363a537d.js');
require('./hooks/PRIVATE_useChartDataCursor/themes/DataCursorStyles.css.js');

require('./Marker-b40392cc.js');

require('./useChartDatatip-76d4aeaa.js');
require('./useDatatip-4f6d3759.js');
require('./Layer-28b1afce.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useThemeInterpolations-4faece98.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-022a8f80.js');
require('./Floating-5a704a27.js');
require('./useFloating-13101293.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-c3802f86.js');
require('./arrayUtils-7d8dcfc3.js');
require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./hooks/PRIVATE_useDatatip/themes/useDatatip.css.js');

require('./VisTabularDatatip-9e44a7dd.js');
require('./PRIVATE_VisTabularDatatip/themes/VisTabularDatatipStyles.css.js');

require('./useChartDrill-8ada5349.js');
require('./useVisDrill-61cccf0f.js');
require('./useDoubleTap-d3937fb6.js');
require('./clientUtils-ef5e20ce.js');
require('./useSelection-b5b55bdd.js');
require('./BaseButton-f6eb6de1.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useActive-6770f917.js');
require('./dimensions-2fcc0acf.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-814ae4b7.js');
require('./useBareButton-8c4f3b41.js');
require('./usePress-886180e4.js');
require('./usePressClick-421494df.js');
require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-a275583f.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./useTestId-9093a54b.js');
require('./Text-436e8b56.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useTranslationBundle-18b7bf8b.js');
require('./hooks/PRIVATE_useVisSelection/themes/ClearSelection.css.js');

require('./cssUtils-3318fbe2.js');
require('./useChartContextMenu-a8c6d928.js');
require('./useContextMenu-e90f401c.js');
require('./SelectMenuGroupContext-ae5e1373.js');
require('./Sheet-44b25f1d.js');
require('./Modal-0cea796e.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./tabbableUtils-7e41d383.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./popupUtils-488fe8f7.js');
require('./WindowOverlay-a10f995e.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./UNSAFE_Sheet/themes/SheetStyles.css.js');

require('./useAnimationStatus-c0c14bcc.js');
require('./useAnimation-fb11e4cc.js');
require('./animationUtils-54852d03.js');
require('./Dropdown-b2579045.js');
require('./keyboardUtils-b42fe7d5.js');
require('./usePopupAnimation-18040396.js');
require('./popupAnimationUtils-dcace427.js');
require('./CollectionInteractionContext-f84ff7a1.js');
require('./UNSAFE_Dropdown/themes/dropdownStyles.css.js');

require('./useOutsideMousedown-8f913d1a.js');
require('./UNSAFE_Menu/themes/MenuStyles.css.js');

require('./UNSAFE_Menu/themes/DropdownMenuStyles.css.js');

require('./Flex-6ca216a7.js');
require('./boxalignment-6dde2812.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');


require('./Skeleton-e746e396.js');
require('./borders-4b8488cb.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');
require('./UNSAFE_Menu/themes/MenuSkeletonStyles.css.js');

require('./LiveRegion-e410e187.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemTheme.js');
require('./UNSAFE_Menu/themes/MenuItemStyles.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js');

require('./useInteractionStyle-c203a8a0.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./tooltipUtils-ae48fed6.js');
require('./EnvironmentProvider-8696a5b8.js');
require('./LayerManager-fc4df14f.js');

require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');



require('./UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js');
require('./useContextMenuGesture-03ba6353.js');
require('./textSelectionUtils-432ab66b.js');
require('./useChartDnd-2c152d6b.js');
require('./utils-6f32ac40.js');
require('./useDnd-a7d9b60c.js');
require('./useTooltip-8728f2ff.js');
require('./useTooltipControlled-3c2afa53.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');
require('./useChartMarquee-02a04456.js');
require('./useVisDragModeControls-56c2f3c3.js');
require('./Button-51364b4d.js');
require('./Icon-bc05e7bd.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./Move-7e3d4692.js');
require('./hooks/PRIVATE_useVisDragModeControls/themes/useVisDragModeControls.css.js');

require('./ZoomAlt-fc9f2cd5.js');
require('./ButtonSetSingle-885ce1b6.js');
require('./useButtonSetContext-a2f31c8e.js');
require('./ButtonSet-a1f0f8b9.js');
require('./UNSAFE_ButtonSet/themes/ButtonSetStyles.css.js');

require('./ButtonSetPositionContext-6a358667.js');
require('./toggleButtonUtils-245111a7.js');
require('./useRovingTabIndexContainer-a3594203.js');
require('./ButtonSetItem-7c81cf63.js');
require('./ButtonSetButton-9106a899.js');
require('./ToggleButtonLabel-cd470bcb.js');
require('./ButtonLabelLayout-8981b39a.js');
require('./useButtonLabelLayoutTheme-b7b3bdb6.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./HiddenAccessible-087eb49a.js');

require('./ButtonSetIconButton-8c52036a.js');
require('./useButtonSetPositionContext-409e4cd6.js');
require('./IconToggleButton-52c12187.js');
require('./useToggleAction-e4f15550.js');
require('./hooks/PRIVATE_useChartMarquee/themes/useChartMarquee.css.js');

require('./useVisTouchResponse-376e11d5.js');
require('./useOverviewContext-1732a3f1.js');
require('./OverviewSlidingWindow-09c08dc9.js');
require('./PRIVATE_VisOverview/themes/VisOverview.css.js');

require('./DragH-7667516a.js');
require('./ScrollBar-76e58249.js');
require('./PRIVATE_ChartScrollBar/themes/ScrollBarStyles.css.js');


/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Get next item in unstacked bar chart.
 * @param info
 * @param viewportEndIndex
 * @param numSeries
 * @param isNavigable
 * @returns
 */
function getNextItem(info, viewportEndIndex, numSeries, isNavigable) {
    const nextItemInGroup = dataStyleUtils.getNextItemInGroup(info, numSeries, isNavigable);
    if (nextItemInGroup)
        return nextItemInGroup;
    if (info.groupIndex < viewportEndIndex) {
        let nextItem = { seriesIndex: 0, groupIndex: info.groupIndex + 1 };
        if (isNavigable(nextItem.seriesIndex, nextItem.groupIndex))
            return nextItem;
        nextItem = getNextItem(nextItem, viewportEndIndex, numSeries, isNavigable);
        if (nextItem)
            return nextItem;
    }
    return;
}
/**
 * Get prev item in unstacked bar chart.
 * @param info
 * @param viewportStartIndex
 * @param numSeries
 * @param isNavigable
 * @returns
 */
function getPreviousItem(info, viewportStartIndex, numSeries, isNavigable) {
    const prevItemInGroup = dataStyleUtils.getPrevItemInGroup(info, isNavigable);
    if (prevItemInGroup)
        return prevItemInGroup;
    if (info.groupIndex > viewportStartIndex) {
        let prevItem = { groupIndex: info.groupIndex - 1, seriesIndex: numSeries - 1 };
        if (isNavigable(prevItem.seriesIndex, prevItem.groupIndex))
            return prevItem;
        prevItem = getPreviousItem(prevItem, viewportStartIndex, numSeries, isNavigable);
        if (prevItem)
            return prevItem;
    }
    return info;
}
function getBarChartNavUtil(getDataItem, numSeries, viewportStartIndex, viewportEndIndex, isStacked, isRtl, isHoriz) {
    const isItemNavigable = (seriesIndex, groupIndex) => {
        const barItem = getDataItem(seriesIndex, groupIndex);
        return Boolean(barItem);
    };
    const getNextChartItem = (itemInfo, arrow) => {
        if (!isStacked &&
            (((arrow === 'ArrowDown' || arrow === 'ArrowUp') && !isHoriz) ||
                ((arrow === 'ArrowLeft' || arrow === 'ArrowRight') && isHoriz))) {
            return itemInfo;
        }
        let newInfo;
        // TODO: this is just to keep typescript happy.
        // figure out consistent typing while doing type cleanup.
        const info = {
            groupIndex: itemInfo.groupIndex,
            seriesIndex: itemInfo.seriesIndex
        };
        if (info.groupIndex < viewportStartIndex) {
            info.groupIndex = viewportStartIndex;
            return info;
        }
        else if (info.groupIndex > viewportEndIndex) {
            info.groupIndex = viewportEndIndex;
            return info;
        }
        if (isStacked) {
            if ((arrow === 'ArrowLeft' && !isHoriz) || (arrow === 'ArrowUp' && isHoriz)) {
                newInfo =
                    isRtl && !isHoriz
                        ? dataStyleUtils.getNextItemInSeries(info, viewportEndIndex, isItemNavigable)
                        : dataStyleUtils.getPrevItemInSeries(info, viewportStartIndex, isItemNavigable);
            }
            else if ((arrow === 'ArrowRight' && !isHoriz) || (arrow === 'ArrowDown' && isHoriz)) {
                newInfo =
                    isRtl && !isHoriz
                        ? dataStyleUtils.getPrevItemInSeries(info, viewportStartIndex, isItemNavigable)
                        : dataStyleUtils.getNextItemInSeries(info, viewportEndIndex, isItemNavigable);
            }
            else if ((arrow === 'ArrowDown' && !isHoriz) || (arrow === 'ArrowLeft' && isHoriz)) {
                newInfo =
                    isRtl && isHoriz
                        ? dataStyleUtils.getNextItemInGroup(info, numSeries, isItemNavigable)
                        : dataStyleUtils.getPrevItemInGroup(info, isItemNavigable);
            }
            else if ((arrow === 'ArrowUp' && !isHoriz) || (arrow === 'ArrowRight' && isHoriz)) {
                newInfo =
                    isRtl && isHoriz
                        ? dataStyleUtils.getPrevItemInGroup(info, isItemNavigable)
                        : dataStyleUtils.getNextItemInGroup(info, numSeries, isItemNavigable);
            }
        }
        else {
            if (!isHoriz) {
                if (arrow === 'ArrowRight') {
                    newInfo = isRtl
                        ? getPreviousItem(info, viewportStartIndex, numSeries, isItemNavigable)
                        : getNextItem(info, viewportEndIndex, numSeries, isItemNavigable);
                }
                else if (arrow === 'ArrowLeft') {
                    newInfo = isRtl
                        ? getNextItem(info, viewportEndIndex, numSeries, isItemNavigable)
                        : getPreviousItem(info, viewportStartIndex, numSeries, isItemNavigable);
                }
            }
            else {
                if (arrow === 'ArrowDown') {
                    newInfo = getNextItem(info, viewportEndIndex, numSeries, isItemNavigable);
                }
                else if (arrow === 'ArrowUp') {
                    newInfo = getPreviousItem(info, viewportStartIndex, numSeries, isItemNavigable);
                }
            }
        }
        return newInfo || info;
    };
    return getNextChartItem;
}

function OverviewBarChart(props) {
    return (jsxRuntime.jsx(BarChart, { datatip: () => ({ content: '' }), zoomAndScroll: "off", plotArea: {
            yMajorTick: {
                isRendered: false
            },
            xMajorTick: {}
        }, xAxis: {
            tickLabel: {
                autoRotate: false,
                // @ts-ignore
                isRendered: props.xAxis?.timeAxisType ? true : false
            }
        }, series: props.series, groups: props.groups, getDataItem: props.getDataItem, yAxis: {
            tickLabel: {
                isRendered: false
            }
        }, overview: "off" }));
}

/**
 * A Bar chart displays information graphically using rectangular bars with heights or lengths proportional to the values that they represent,
 * making relationships among the data easier to understand.
 */
function BarChart({ width, height, selectedIds = [], hiddenIds = [], highlightedIds, hideAndShowBehavior = 'none', orientation = 'vertical', yAxis, series, groups, isStacked = false, getDataItem, selectionMode = 'none', drilling = 'off', dataItemGaps = 0.5, ...props }) {
    const { direction } = useUser.useUser();
    const isRtl = direction === 'rtl';
    const hiddenSet = new Set(hiddenIds);
    const isHoriz = orientation === 'horizontal';
    const timeAxisType = props.xAxis?.timeAxisType;
    const selectedSet = new Set(selectedIds);
    const highlightedSet = new Set(highlightedIds);
    const getGapRatio = (startIndex, endIndex) => {
        return utils.getBarGapRatio(isStacked, startIndex, endIndex);
    };
    const getNextChartItemFunc = (startIndex, endIndex) => {
        return getBarChartNavUtil(getDataItem, series.length, startIndex, endIndex, isStacked, isRtl, isHoriz);
    };
    const isSplitDualY = typeof props.y2Axis?.split === 'number';
    const getDataItemPos = (xScale, yScale, y2Scale, axisStepWidth, averageGroupZ, gapRatio) => {
        return BarGroup.getBarPosition(series, groups, getDataItem, xScale, yScale, y2Scale, hiddenSet, dataItemGaps, axisStepWidth, averageGroupZ, gapRatio, yAxis?.scale === 'log', isStacked, isRtl, isHoriz, isSplitDualY, timeAxisType);
    };
    const findNearest = (xScale, yScale, y2Scale, axisStepWidth, averageGroupZ, gapRatio) => {
        const getSeriesDataInfo = BarGroup.getSeriesOffsetAndWidth(axisStepWidth, averageGroupZ, series, getDataItem, gapRatio, dataItemGaps, isStacked, isRtl, isHoriz);
        return utils.findNearestDataPoint(series, groups, xScale, yScale, getDataItem, isStacked, getSeriesDataInfo, BarGroup.getBarCenterCoordinate, yAxis?.scale === 'log', dataItemGaps, hiddenSet, timeAxisType, y2Scale);
    };
    const getMarkerColorAndType = (seriesIndex, groupIndex) => {
        return {
            color: dataStyleUtils.getItemColor(seriesIndex, groupIndex, series, getDataItem),
            type: 'circle'
        };
    };
    const getOffset = (groupCount) => {
        return utils.getStartAndEndAxesOffset(false, false, isHoriz, groupCount, true);
    };
    return (jsxRuntime.jsx(Chart.Chart, { width: width, height: height, selectedIds: selectedSet, hiddenIds: hiddenIds, hideAndShowBehavior: hideAndShowBehavior, orientation: orientation, yAxis: yAxis, series: series, groups: groups, isStacked: isStacked, getDataItem: getDataItem, selectionMode: selectionMode, drilling: drilling, renderGridLinesInFront: false, findNearest: findNearest, getDataItemPos: getDataItemPos, getNavUtil: getNextChartItemFunc, getGapRatio: getGapRatio, getMarkerInfo: getMarkerColorAndType, isRtl: isRtl, getStartAndEndAxesOffset: getOffset, defaultOverviewContentRenderer: () => {
            return jsxRuntime.jsx(OverviewBarChart, { series: series, groups: groups, getDataItem: getDataItem });
        }, ...props, children: ({ groupsInfo, xStartIndex, xEndIndex, xScale, yScale, y2Scale, gapRatio, activeId, focusedItemInfo, hoveredItemInfo, axisStepWidth, averageGroupZ, isPointInsideMarquee, getTextDimensions, defaultFontSize, colors, getItemAriaLabel, plotAreaClipPathId }) => {
            return (jsxRuntime.jsx("g", { clipPath: `url(#${plotAreaClipPathId})`, children: groupsInfo
                    .slice(xStartIndex, xEndIndex + 1)
                    .map(({ item: group }, index) => (jsxRuntime.jsx(BarGroup.BarGroup, { yScale: yScale, splitDualY: isSplitDualY, y2Scale: y2Scale, xScale: xScale, groupIndex: index + xStartIndex, selectedIds: selectedSet, hiddenIds: hiddenSet, highlightedIds: highlightedIds ? highlightedSet : highlightedIds, isDrillEnabled: drilling === 'on', series: series, barGapRatio: gapRatio, dataItemGaps: dataItemGaps, isPointInsideMarquee: isPointInsideMarquee, isStacked: isStacked, axisStepWidth: axisStepWidth, averageGroupZ: averageGroupZ, group: group, timeAxisType: props.xAxis?.timeAxisType, colors: colors, activeId: activeId, isLog: yAxis?.scale === 'log', orientation: orientation, getDataItem: getDataItem, isSelectionEnabled: selectionMode != 'none', isRtl: isRtl, focusedItemInfo: focusedItemInfo, hoveredItemInfo: hoveredItemInfo, getTextDimensions: getTextDimensions, defaultFontSize: defaultFontSize, getStackLabel: props.getStackLabel, getItemAriaLabel: getItemAriaLabel }))) }));
        } }));
}

exports.BarChart = BarChart;
