// @oracle/oraclejet-preact/cjs/UNSAFE_LineAreaChart.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var utils = require('./utils-7c1b4abe.js');
var chartUtils = require('./chartUtils-84b671f7.js');
var LineSeries = require('./LineSeries-4c97a793.js');
var dataStyleUtils = require('./dataStyleUtils-8c59d570.js');
var useUser = require('./useUser-f900ddf1.js');
var colorUtils = require('./colorUtils-c97cf90b.js');
require('./accUtils-5c7d3255.js');
var themeContract_css = require('./themeContract.css-a53fd740.js');
var Chart = require('./Chart-376785c5.js');
require('./utils-a66430fc.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./getLocale-8e17dc46.js');
require('./stringUtils-3e19c8af.js');
require('./labelUtils-a3deed57.js');
require('./textUtils-e779c845.js');
require('./utils-ca441171.js');
require('./datatipUtils-d1ca81a1.js');
require('./clientUtils-a4619fcd.js');
require('./clientHints-c40c5250.js');
require('./dimensionUtils-3919753d.js');
require('./scale-9e43a5f6.js');
require('./lineUtils-34a63fb2.js');
require('./Marker-b40392cc.js');

require('./PRIVATE_Chart/themes/ChartStyles.css.js');

require('./classNames-c14c6ef3.js');
require('preact/hooks');
require('./LayerHost-aee0e070.js');
require('./index-15e13649.js');
require('preact');
require('preact/compat');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./TrackResizeContainer-100be103.js');
require('./useSize-8115eae0.js');
require('./useResizeObserver-a9e97180.js');
require('./useAddBusyState-d19ae1fa.js');
require('./BusyStateContext-86f40d3c.js');
require('./useBusyStateContext-362eee2a.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./size-0717c0bd.js');
require('./useTextDimensions-c8b41697.js');
require('./useCssVars-cc025013.js');
require('./useTheme-09dfbb78.js');
require('./useColorScheme-23e4aab4.js');
require('./useScale-adc62f41.js');
require('./Gridlines-3bdce0b7.js');
require('./PRIVATE_PlotArea/themes/PlotAreaStyles.css.js');

require('./hooks/PRIVATE_useVisDnd/themes/VisDndStyles.css.js');

require('./useId-6c0eeb27.js');
require('./Axis-2d648884.js');
require('./PRIVATE_Axis/themes/AxisStyles.css.js');

require('./MultilineText-837e9afd.js');
require('./PRIVATE_Axis/themes/redwood/AxisTheme.js');
require('./PRIVATE_Axis/themes/redwood/AxisBaseTheme.css.js');

require('./PRIVATE_Axis/themes/redwood/AxisVariants.css.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./useComponentTheme-082fc8e4.js');
require('./logger-2b636482.js');
require('./useChartViewport-d2e8c46f.js');
require('./eventsUtils-d34f8647.js');
require('./usePinch-741e9343.js');
require('./usePointerGesture-0b9cc1b7.js');
require('./mergeProps-e3da7237.js');
require('./useChartNav-b079ab1e.js');
require('./layoutUtils-e52005dc.js');
require('./useVisHover-ad36112d.js');
require('./useSelectDrill-b818e267.js');
require('./useChartDataCursor-363a537d.js');
require('./hooks/PRIVATE_useChartDataCursor/themes/DataCursorStyles.css.js');

require('./useChartDatatip-76d4aeaa.js');
require('./useDatatip-4f6d3759.js');
require('./Layer-28b1afce.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useThemeInterpolations-4faece98.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-022a8f80.js');
require('./Floating-5a704a27.js');
require('./useFloating-13101293.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-c3802f86.js');
require('./arrayUtils-7d8dcfc3.js');
require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./hooks/PRIVATE_useDatatip/themes/useDatatip.css.js');

require('./VisTabularDatatip-9e44a7dd.js');
require('./PRIVATE_VisTabularDatatip/themes/VisTabularDatatipStyles.css.js');

require('./useChartDrill-8ada5349.js');
require('./useVisDrill-61cccf0f.js');
require('./useDoubleTap-d3937fb6.js');
require('./clientUtils-ef5e20ce.js');
require('./useSelection-b5b55bdd.js');
require('./BaseButton-f6eb6de1.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useActive-6770f917.js');
require('./dimensions-2fcc0acf.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-814ae4b7.js');
require('./useBareButton-8c4f3b41.js');
require('./usePress-886180e4.js');
require('./usePressClick-421494df.js');
require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-a275583f.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./useTestId-9093a54b.js');
require('./Text-436e8b56.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useTranslationBundle-18b7bf8b.js');
require('./hooks/PRIVATE_useVisSelection/themes/ClearSelection.css.js');

require('./cssUtils-3318fbe2.js');
require('./useChartContextMenu-a8c6d928.js');
require('./useContextMenu-e90f401c.js');
require('./SelectMenuGroupContext-ae5e1373.js');
require('./Sheet-44b25f1d.js');
require('./Modal-0cea796e.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./tabbableUtils-7e41d383.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./popupUtils-488fe8f7.js');
require('./WindowOverlay-a10f995e.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./UNSAFE_Sheet/themes/SheetStyles.css.js');

require('./useAnimationStatus-c0c14bcc.js');
require('./useAnimation-fb11e4cc.js');
require('./animationUtils-54852d03.js');
require('./Dropdown-b2579045.js');
require('./keyboardUtils-b42fe7d5.js');
require('./usePopupAnimation-18040396.js');
require('./popupAnimationUtils-dcace427.js');
require('./CollectionInteractionContext-f84ff7a1.js');
require('./UNSAFE_Dropdown/themes/dropdownStyles.css.js');

require('./useOutsideMousedown-8f913d1a.js');
require('./UNSAFE_Menu/themes/MenuStyles.css.js');

require('./UNSAFE_Menu/themes/DropdownMenuStyles.css.js');

require('./Flex-6ca216a7.js');
require('./boxalignment-6dde2812.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');


require('./Skeleton-e746e396.js');
require('./borders-4b8488cb.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');
require('./UNSAFE_Menu/themes/MenuSkeletonStyles.css.js');

require('./LiveRegion-e410e187.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemTheme.js');
require('./UNSAFE_Menu/themes/MenuItemStyles.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js');

require('./useInteractionStyle-c203a8a0.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./tooltipUtils-ae48fed6.js');
require('./EnvironmentProvider-8696a5b8.js');
require('./LayerManager-fc4df14f.js');

require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');



require('./UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js');
require('./useContextMenuGesture-03ba6353.js');
require('./textSelectionUtils-432ab66b.js');
require('./useChartDnd-2c152d6b.js');
require('./utils-6f32ac40.js');
require('./useDnd-a7d9b60c.js');
require('./useTooltip-8728f2ff.js');
require('./useTooltipControlled-3c2afa53.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');
require('./useChartMarquee-02a04456.js');
require('./useVisDragModeControls-56c2f3c3.js');
require('./Button-51364b4d.js');
require('./Icon-bc05e7bd.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./Move-7e3d4692.js');
require('./hooks/PRIVATE_useVisDragModeControls/themes/useVisDragModeControls.css.js');

require('./ZoomAlt-fc9f2cd5.js');
require('./ButtonSetSingle-885ce1b6.js');
require('./useButtonSetContext-a2f31c8e.js');
require('./ButtonSet-a1f0f8b9.js');
require('./UNSAFE_ButtonSet/themes/ButtonSetStyles.css.js');

require('./ButtonSetPositionContext-6a358667.js');
require('./toggleButtonUtils-245111a7.js');
require('./useRovingTabIndexContainer-a3594203.js');
require('./ButtonSetItem-7c81cf63.js');
require('./ButtonSetButton-9106a899.js');
require('./ToggleButtonLabel-cd470bcb.js');
require('./ButtonLabelLayout-8981b39a.js');
require('./useButtonLabelLayoutTheme-b7b3bdb6.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./HiddenAccessible-087eb49a.js');

require('./ButtonSetIconButton-8c52036a.js');
require('./useButtonSetPositionContext-409e4cd6.js');
require('./IconToggleButton-52c12187.js');
require('./useToggleAction-e4f15550.js');
require('./hooks/PRIVATE_useChartMarquee/themes/useChartMarquee.css.js');

require('./useVisTouchResponse-376e11d5.js');
require('./useOverviewContext-1732a3f1.js');
require('./OverviewSlidingWindow-09c08dc9.js');
require('./PRIVATE_VisOverview/themes/VisOverview.css.js');

require('./DragH-7667516a.js');
require('./ScrollBar-76e58249.js');
require('./PRIVATE_ChartScrollBar/themes/ScrollBarStyles.css.js');


const LINE_WITH_AREA_OPACITY = 0.2;
/**
 * The LineAreaChart displays information graphically using lines and filled areas, making relationships among the data easier to understand.
 */
function LineAreaChart({ width, height, selectedIds = [], hiddenIds = [], highlightedIds, hideAndShowBehavior = 'none', orientation = 'vertical', yAxis, y2Axis, series, groups, isStacked = false, getDataItem, selectionMode = 'none', drilling = 'off', type = 'line', overview = 'off', dataLabelEffect = 'outline', ...props }) {
    const { direction, forcedColors } = useUser.useUser();
    const isHighContrast = forcedColors === 'active';
    const isTestEnv = props.testId !== undefined;
    const isRtl = direction === 'rtl';
    const isLog = yAxis?.scale === 'log';
    const timeAxisType = props.xAxis?.timeAxisType;
    const hiddenSet = new Set(hiddenIds);
    const selectedSet = new Set(selectedIds);
    const highlightedSet = new Set(highlightedIds);
    const isHoriz = orientation === 'horizontal';
    const colors = colorUtils.getColorRamp();
    let renderedSeries = isStacked
        ? utils.getCumulativeSeries(series.length, groups.length, getDataItem, hiddenSet, isLog)
        : utils.getSeriesData(series?.length, groups?.length, getDataItem, hiddenSet, isLog);
    const renderedGetDataItem = (seriesIndex, groupIndex) => {
        if (!isStacked)
            return getDataItem(seriesIndex, groupIndex);
        return {
            ...getDataItem(seriesIndex, groupIndex),
            value: renderedSeries[seriesIndex][groupIndex]
        };
    };
    const getGapRatio = () => {
        return 1;
    };
    const getNextChartItemFunc = (startIndex, endIndex) => {
        // TODO: JET-71332
        //@ts-ignore
        return dataStyleUtils.getLineAreaChartNavUtil(getDataItem, series.length, startIndex, endIndex, isStacked, isRtl, isHoriz);
    };
    const getDataItemPos = (xScale, yScale, y2Scale) => {
        return (seriesIndex, groupIndex) => {
            const isAssociatedToY2 = series[seriesIndex].associatedYAxis === 'y2';
            return chartUtils.getLineAreaChartItemPosition(seriesIndex, groupIndex, renderedSeries, getDataItem, groups, isHoriz, xScale, isAssociatedToY2 ? y2Scale : yScale, timeAxisType);
        };
    };
    const getMarkerColorAndType = (seriesIndex, groupIndex) => {
        return {
            color: dataStyleUtils.getItemColor(seriesIndex, groupIndex, series, getDataItem),
            type: dataStyleUtils.getMarkerType(seriesIndex, groupIndex, getDataItem)
        };
    };
    const findNearest = (xScale, yScale, y2Scale) => {
        const getSeriesDataInfo = () => {
            return {
                offset: 0,
                dataWidth: 0
            };
        };
        const getYCoord = (value, yScale) => {
            return yScale.transform(value);
        };
        return utils.findNearestDataPoint(series, groups, xScale, yScale, renderedGetDataItem, isStacked, getSeriesDataInfo, getYCoord, isLog, 0, hiddenSet, timeAxisType, y2Scale);
    };
    const { hasCenteredSeries, hasUncenteredSeries } = series.reduce((acc, s) => {
        return {
            hasCenteredSeries: acc.hasCenteredSeries ||
                s.lineType === 'centeredSegmented' ||
                s.lineType === 'centeredStepped',
            hasUncenteredSeries: acc.hasUncenteredSeries || s.lineType === 'segmented' || s.lineType === 'stepped'
        };
    }, { hasCenteredSeries: false, hasUncenteredSeries: false });
    const getOffset = (groupCount) => {
        return utils.getStartAndEndAxesOffset(hasCenteredSeries, hasUncenteredSeries, isHoriz, groupCount);
    };
    const markers = utils.getMarkers();
    const areaOpacity = type === 'lineWithArea' ? LINE_WITH_AREA_OPACITY : 1;
    const isAreaRendered = type === 'area' || type === 'lineWithArea';
    const isContrastLineNeeded = type === 'area';
    const allLineSegmentCoords = [];
    return (jsxRuntime.jsx(Chart.Chart, { width: width, height: height, selectedIds: selectedSet, hiddenIds: hiddenIds, hideAndShowBehavior: hideAndShowBehavior, orientation: orientation, yAxis: yAxis, y2Axis: y2Axis, series: series, groups: groups, selectionMode: selectionMode, drilling: drilling, isStacked: isStacked, getDataItem: getDataItem, findNearest: findNearest, getDataItemPos: getDataItemPos, getMarkerInfo: getMarkerColorAndType, renderGridLinesInFront: isAreaRendered, getNavUtil: getNextChartItemFunc, getGapRatio: getGapRatio, getStartAndEndAxesOffset: getOffset, isRtl: isRtl, overview: overview, defaultOverviewContentRenderer: () => {
            return (jsxRuntime.jsx(OverviewLineAreaChart, { series: series, groups: groups, getDataItem: getDataItem, type: type }));
        }, ...props, children: ({ getItemAriaLabel, groupsInfo, xStartIndex, xEndIndex, xScale, yScale, y2Scale, activeId, focusedItemInfo, hoveredItemInfo, isPointInsideMarquee, defaultFontSize, getTextDimensions, plotAreaClipPathId, axisStepWidth }) => {
            const leafGroups = groupsInfo.map(({ item }) => item);
            renderedSeries = isStacked
                ? utils.getCumulativeSeries(series.length, leafGroups.length, getDataItem, hiddenSet, isLog)
                : utils.getSeriesData(series?.length, leafGroups?.length, getDataItem, hiddenSet, isLog);
            return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [renderedSeries.map((lineAreaSeries, index) => {
                        if (lineAreaSeries.length === 0) {
                            /* TODO: need a better way to handle when a series is hidden. */
                            return;
                        }
                        const seriesItem = series[index];
                        const isAssociatedToY2 = seriesItem.associatedYAxis === 'y2';
                        const lineSegmentCoords = utils.getLineAreaSegments(isAssociatedToY2 ? y2Scale : yScale, xScale, lineAreaSeries, xStartIndex, xEndIndex, isLog, leafGroups, index, renderedGetDataItem, timeAxisType);
                        allLineSegmentCoords[index] = lineSegmentCoords[0];
                        const isBottomSegmentCurved = seriesItem.lineType === 'curved' &&
                            isStacked &&
                            renderedSeries.length > 1 &&
                            index > 0;
                        const isHighlighted = highlightedIds === undefined ||
                            highlightedIds.length === 0 ||
                            highlightedSet.has(seriesItem.id);
                        const prevCoords = isStacked && index != 0
                            ? allLineSegmentCoords.reduce((prevValue, newValue, j) => {
                                // find the index i such that i < index, and series i  has at least
                                // one or more data item unhidden. Only such series will can be
                                // used as bottomCoord for current series. Otherwise we have to use
                                // baseline coord
                                if (j >= index) {
                                    return prevValue;
                                }
                                if (newValue.length > 0) {
                                    return j;
                                }
                                return prevValue;
                            }, 0)
                            : undefined;
                        return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [isAreaRendered && (jsxRuntime.jsx(chartUtils.AreaSeries, { yScale: isAssociatedToY2 ? y2Scale : yScale, lineType: seriesItem.lineType, color: seriesItem.areaColor || colors[index % colors.length], isHorizontal: isHoriz, areaColorOpacity: areaOpacity, isBottomSegmentCurved: isBottomSegmentCurved, bottomCoords: prevCoords !== undefined &&
                                        allLineSegmentCoords[prevCoords].some((value) => value != null)
                                        ? allLineSegmentCoords[prevCoords]
                                        : undefined, lineSegments: lineSegmentCoords, isLog: isLog, isHighlighted: isHighlighted, seriesIndex: index, groupWidth: axisStepWidth, isRtl: isRtl, timeAxisType: timeAxisType, plotAreaClipPathId: plotAreaClipPathId })), jsxRuntime.jsx(LineSeries.LineSeries, { lineType: seriesItem.lineType, 
                                    /* TODO: confirm with UX if this is contrastLine (changes depending on dark/light mode) or always white.
                                        It's always white in legacy.*/
                                    lineColor: isContrastLineNeeded ? themeContract_css.colorSchemeVars.dvt.contrastLine : seriesItem.lineColor, 
                                    /* The area generated in legacy are bit apart and not touching when stacked. Preact areas are a bit closer than legacy when curved.
                                      Increased the contrast linewidth (to 2.5 from legacy 1.25) to give legacy visual treatment.
                                      TODO: investivage why legacy areas are not touching. maybe we want to do the same in preact svg utils that
                                      generates this curve.
                                     */
                                    lineWidth: isContrastLineNeeded ? (isStacked ? 2.5 : 1.25) : seriesItem.lineWidth, lineStyle: seriesItem.lineStyle, color: seriesItem.lineColor || colors[index % colors.length], lineSegmentCoords: lineSegmentCoords, isHorizontal: isHoriz, isHighlighted: isHighlighted, seriesIndex: index, groupWidth: axisStepWidth, isRtl: isRtl, timeAxisType: timeAxisType, plotAreaClipPathId: plotAreaClipPathId }), 
                                /*for non stacked charts, markers should render in the order of series.
                                 some markers might get hidden behind area but this preserves the legacy behavior. */
                                !isStacked && (jsxRuntime.jsx(LineSeries.MarkersWithLabel, { yScale: isAssociatedToY2 ? y2Scale : yScale, xScale: xScale, groups: leafGroups, series: seriesItem, timeAxisType: timeAxisType, color: seriesItem.lineColor ||
                                        seriesItem.areaColor ||
                                        colors[index % colors.length] /* TODO : make this use getItemColor function from dataStyleUtils. */, seriesIndex: index, startIndex: xStartIndex, endIndex: xEndIndex, markerType: markers[index % markers.length], orientation: orientation, isLog: isLog, activeId: activeId, getDataItem: getDataItem, isPointInsideMarquee: isPointInsideMarquee, selectedIds: selectedSet, highlightedIds: highlightedIds !== undefined ? highlightedSet : undefined, focusedItemIndex: focusedItemInfo.seriesIndex === index &&
                                        focusedItemInfo.isFocusVisible &&
                                        !(props.isDataCursorEnabled && props.dataCursorStyle?.isMarkerDisplayed)
                                        ? focusedItemInfo.groupIndex
                                        : undefined, hoveredItemIndex: hoveredItemInfo?.isCurrent && hoveredItemInfo.seriesIndex === index
                                        ? hoveredItemInfo.groupIndex
                                        : undefined, isSelectionEnabled: selectionMode === 'single' || selectionMode === 'multiple', isDrillEnabled: drilling === 'on', defaultFontSize: defaultFontSize, dataLabelEffect: dataLabelEffect, isHighContrast: isHighContrast, isTestEnv: isTestEnv, getTextDimensions: getTextDimensions, getItemAriaLabel: getItemAriaLabel }))] }));
                    }), isStacked &&
                        renderedSeries.map((lineAreaSeries, index) => {
                            if (lineAreaSeries.length === 0) {
                                /* TODO: need a better way to handle when a series is hidden.*/
                                return;
                            }
                            const seriesItem = series[index];
                            const isAssociatedToY2 = seriesItem.associatedYAxis === 'y2';
                            return (jsxRuntime.jsx(LineSeries.MarkersWithLabel, { yScale: isAssociatedToY2 ? y2Scale : yScale, series: series[index], xScale: xScale, groups: leafGroups, timeAxisType: timeAxisType, color: seriesItem.lineColor || seriesItem.areaColor || colors[index % colors.length], seriesIndex: index, isPointInsideMarquee: isPointInsideMarquee, startIndex: xStartIndex, endIndex: xEndIndex, markerType: markers[index % markers.length], orientation: orientation, isLog: isLog, activeId: activeId, getDataItem: renderedGetDataItem, selectedIds: selectedSet, focusedItemIndex: focusedItemInfo.seriesIndex === index && focusedItemInfo.isFocusVisible
                                    ? focusedItemInfo.groupIndex
                                    : undefined, hoveredItemIndex: hoveredItemInfo?.isCurrent && hoveredItemInfo.seriesIndex === index
                                    ? hoveredItemInfo.groupIndex
                                    : undefined, isSelectionEnabled: selectionMode === 'single' || selectionMode === 'multiple', isDrillEnabled: drilling === 'on', defaultFontSize: defaultFontSize, dataLabelEffect: dataLabelEffect, isHighContrast: isHighContrast, isTestEnv: isTestEnv, getTextDimensions: getTextDimensions, getItemAriaLabel: getItemAriaLabel }));
                        })] }));
        } }));
}
function OverviewLineAreaChart(props) {
    return (jsxRuntime.jsx(LineAreaChart, { datatip: () => ({ content: '' }), zoomAndScroll: "off", overview: "off", plotArea: {
            yMajorTick: {
                isRendered: false
            }
        }, xAxis: {
            tickLabel: {
                autoRotate: false,
                // @ts-ignore
                isRendered: props.xAxis?.timeAxisType ? true : false
            }
        }, series: props.series, groups: props.groups, getDataItem: props.getDataItem, yAxis: {
            tickLabel: {
                isRendered: false
            }
        }, type: props.type }));
}

exports.LineAreaChart = LineAreaChart;
