// @oracle/oraclejet-preact/cjs/UNSAFE_ButtonSetMultiple.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var useButtonSetContext = require('./useButtonSetContext-a2f31c8e.js');
require('preact/hooks');
var ButtonSet = require('./ButtonSet-a1f0f8b9.js');
var buttonUtils = require('./buttonUtils-dff2ea3e.js');
var useId = require('./useId-6c0eeb27.js');
require('preact');
require('./dimensions-2fcc0acf.js');
require('./size-0717c0bd.js');
require('./utils-a66430fc.js');
require('./themeContract.css-a53fd740.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./mergeInterpolations-6727b536.js');
require('./classNames-c14c6ef3.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./UNSAFE_ButtonSet/themes/ButtonSetStyles.css.js');

require('./ButtonSetPositionContext-6a358667.js');
require('./toggleButtonUtils-245111a7.js');
require('./useTestId-9093a54b.js');
require('./LayerHost-aee0e070.js');
require('./index-15e13649.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./useRovingTabIndexContainer-a3594203.js');
require('./useUser-f900ddf1.js');
require('./clientHints-c40c5250.js');
require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-a275583f.js');
require('./refUtils-a9872e75.js');
require('./id-86356250.js');

/**
 * ButtonSet allow users to select the state of one or more related options.
 */
const ButtonSetMultiple = compat.forwardRef(({ children, value: buttonSetItemValue, onCommit, display = 'all', testId, size = 'md', isDisabled = false, variant = 'outlined', layoutWidth = 'auto', 'aria-label': accessibleLabel, 'aria-labelledby': ariaLabelledBy, 'aria-controls': ariaControls, 'aria-describedby': ariaDescribedBy, width, ...props }, ref) => {
    const { globalEventProps } = buttonUtils.processGlobals(props);
    return (jsxRuntime.jsx(useButtonSetContext.ButtonSetContext.Provider, { value: {
            variant: variant,
            size: size,
            isDisabled: isDisabled,
            layoutWidth: layoutWidth,
            inputType: 'checkbox',
            inputName: useId.useId(),
            display: display,
            width: width,
            buttonSetValue: buttonSetItemValue,
            onCommit: onCommit
        }, children: jsxRuntime.jsx(ButtonSet.ButtonSet, { ref: ref, width: width, testId: testId, "aria-label": accessibleLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, "aria-controls": ariaControls, ...globalEventProps, children: children }) }));
});
ButtonSetMultiple.displayName = 'ButtonSetMultiple';

exports.ButtonSetMultiple = ButtonSetMultiple;
