// @oracle/oraclejet-preact/cjs/useButtonLabelLayoutTheme-b7b3bdb6.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme = require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useButtonLabelLayoutTheme({ display, styling, size, children, startIcon, endIcon, suffix }) {
    const isShowIcons = display === 'all' || display === 'icons';
    const isShowText = display === 'all' || display === 'label';
    const isOnlyIcon = (startIcon && (display === 'icons' || !children) && !endIcon);
    const hasTextStartIcon = !isNullOrUndefined(startIcon) && display === 'all';
    const hasTextEndIcon = !isNullOrUndefined(endIcon) && isShowIcons;
    const isTwoIcons = !isNullOrUndefined(startIcon) && !isNullOrUndefined(endIcon);
    const isTwoIconsOnly = isTwoIcons && display === 'icons';
    const { classes: startIconClasses } = useComponentTheme.useComponentTheme(UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme.ButtonLabelLayoutRedwoodTheme, {
        target: 'startIcon',
        size: styling == 'default' ? size : ('embedded' + size),
        onlyIcon: isOnlyIcon ? 'isOnlyIcon' : 'notOnlyIcon',
        twoIconsOnly: isTwoIconsOnly ? 'isTwoIconsOnly' : 'notTwoIconsOnly'
    });
    const { classes: labelClasses, styles: labelStyles } = useComponentTheme.useComponentTheme(UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme.ButtonLabelLayoutRedwoodTheme, {
        target: 'label',
        size: styling == 'embedded' ? ('embedded' + size) : size,
        fill: styling == 'fill' ? 'isFill' : 'notFill',
        textEndIcon: hasTextEndIcon ? 'hasTextEndIcon' : 'notTextEndIcon',
        textStartIcon: hasTextStartIcon ? 'hasTextStartIcon' : 'notTextStartIcon',
        suffix: suffix ? 'hasSuffix' : 'notSuffix',
        twoIcons: isTwoIcons ? 'hasTwoIcons' : 'notTwoIcons',
        loading: 'notLoading'
    });
    const { classes: suffixClasses } = useComponentTheme.useComponentTheme(UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme.ButtonLabelLayoutRedwoodTheme, {
        target: 'suffix',
        size: styling == 'default' ? size : ('embedded' + size),
        textStartIcon: hasTextStartIcon ? 'hasTextStartIcon' : 'notTextStartIcon',
        twoIcons: isTwoIcons ? 'hasTwoIcons' : 'notTwoIcons'
    });
    const { classes: endIconClasses } = useComponentTheme.useComponentTheme(UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme.ButtonLabelLayoutRedwoodTheme, {
        target: 'endIcon',
        textEndIcon: hasTextEndIcon ? 'hasTextEndIcon' : 'notTextEndIcon',
        twoIcons: isTwoIcons ? 'hasTwoIcons' : 'notTwoIcons',
        size: styling == 'default' ? size : ('embedded' + size)
    });
    return {
        startIconClasses,
        endIconClasses,
        labelClasses,
        labelStyles,
        suffixClasses,
        containerClasses: labelStyles.container,
        isShowIcons,
        isShowText,
        isOnlyIcon,
        hasTextStartIcon,
        hasTextEndIcon,
        isTwoIcons,
        isTwoIconsOnly
    };
}
const isNullOrUndefined = (value) => {
    return value === null || value === undefined;
};

exports.useButtonLabelLayoutTheme = useButtonLabelLayoutTheme;
