// @oracle/oraclejet-preact/cjs/useBareButton-8c4f3b41.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var usePress = require('./usePress-886180e4.js');
var usePressClick = require('./usePressClick-421494df.js');
require('./TabbableModeContext-e99d527e.js');
var useTabbableMode = require('./useTabbableMode-a275583f.js');
var clientHints = require('./clientHints-c40c5250.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var buttonUtils = require('./buttonUtils-dff2ea3e.js');
var useTestId = require('./useTestId-9093a54b.js');

const roleProps = (elType, hidden, isFocusable, isLabel, isSwitch) => {
    switch (elType) {
        case 'link':
            return { role: 'link' };
        case 'button':
            return isSwitch ? { role: 'switch' } : {};
        case 'div':
            return hidden ? {} : { role: 'button' };
        case 'span':
            return isLabel
                ? {}
                : hidden
                    ? { role: 'none' }
                    : isFocusable
                        ? isSwitch
                            ? { role: 'switch' }
                            : { role: 'button' }
                        : { role: 'img' };
        default:
            return hidden ? { role: 'none' } : isFocusable ? { role: 'button' } : { role: 'img' };
    }
};
/**
 * BareButtons are unformatted bare buttons that direct users to initiate or take actions and work with a single tap, click, or keystroke.
 */
function useBareButton({ onAction, isDisabled, isRepeat, isClick, elementDetails = { type: 'button', buttonType: 'button' }, testId, ...props }) {
    const clientHints$1 = clientHints.getClientHints();
    // :active only works on IOS devices if a touch handler exists
    const iosProps = clientHints$1.platform === 'ios' ? { ontouchstart: function () { } } : {};
    const pressHandler = (e) => {
        if (onAction) {
            onAction({
                reason: e.pointerType == '' ? 'keyboard' : 'pointer'
            });
        }
    };
    const { globalEventProps, ariaProps: { 'aria-disabled': ariaDisabled, 'aria-hidden': ariaHidden, ...ariaOtherProps } } = buttonUtils.processGlobals(props);
    const { pressProps } = usePress.usePress(pressHandler, {
        isDisabled: isDisabled,
        isRepeat
    });
    const { pressProps: pressClickProps } = usePressClick.usePressClick(pressHandler, {
        isDisabled: isDisabled
    });
    const { type: elementType = 'button', isFocusable: isElementFocusable = true, isLabel: isElementLabel = false, isSwitch: isElementSwitch = false, ...elementProps } = { ...elementDetails };
    const buttonType = elementDetails.type === 'button' ? elementDetails.buttonType ?? 'button' : undefined;
    const typeProps = elementDetails.type === 'button' && elementDetails.buttonType === 'button'
        ? { type: buttonType }
        : {};
    // 'role' should be 'none' when aria-hidden is true or 'img' when false and not focusable
    const role = roleProps(elementType, Boolean(ariaHidden), isElementFocusable ?? false, isElementLabel ?? false, isElementSwitch ?? false);
    const ElementType = elementDetails?.type;
    const isSpan = ElementType == 'span';
    const isBareButton = ElementType == 'button';
    const { tabbableModeProps } = useTabbableMode.useTabbableMode(isDisabled);
    const arDisabled = ElementType == 'button'
        ? ariaDisabled
            ? { 'aria-disabled': ariaDisabled }
            : {}
        : isDisabled
            ? { 'aria-disabled': isDisabled }
            : {};
    const arHidden = ElementType === 'div' ? {} : { 'aria-hidden': ariaHidden };
    const adjustedAriaProps = role.role === 'none' ? {} : mergeProps.mergeProps(arDisabled, arHidden, ariaOtherProps);
    const testIdProps = useTestId.useTestId(testId);
    const getTabIndex = () => {
        if ((isSpan && (isElementFocusable === false || isDisabled)) || (isBareButton && isDisabled)) {
            // no tabIndex value should be defined for disabled button elements and non-focusable or disabled span elements
            return { tabIndex: undefined };
        }
        return tabbableModeProps;
    };
    const legacyGlobals = {
        autofocus: props.autofocus,
        title: props.title,
        disabled: isDisabled
    };
    return mergeProps.mergeProps(elementProps, onAction ? (isRepeat || (isClick ?? isBareButton) ? pressClickProps : pressProps) : {}, iosProps, globalEventProps, adjustedAriaProps, getTabIndex(), typeProps, testIdProps, role, legacyGlobals);
}

exports.useBareButton = useBareButton;
