// @oracle/oraclejet-preact/cjs/useProgressButtonLabelLayoutTheme-9c9f986f.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme = require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useButtonLabelLayoutTheme = require('./useButtonLabelLayoutTheme-b7b3bdb6.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useProgressButtonLabelLayoutTheme({ display, styling, size, children, startIcon, isLoading = false }) {
    const { startIconClasses, containerClasses, isShowIcons, isShowText, hasTextStartIcon } = useButtonLabelLayoutTheme.useButtonLabelLayoutTheme({ display, size, styling, children, startIcon });
    const { classes: labelClasses, styles: labelStyles } = useComponentTheme.useComponentTheme(UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme.ButtonLabelLayoutRedwoodTheme, {
        target: 'label',
        size: size,
        fill: styling == 'fill' ? 'isFill' : 'notFill',
        textEndIcon: 'notTextEndIcon',
        textStartIcon: hasTextStartIcon ? 'hasTextStartIcon' : 'notTextStartIcon',
        suffix: 'notSuffix',
        twoIcons: 'notTwoIcons',
        loading: isLoading && !hasTextStartIcon ? 'isLoading' : 'notLoading'
    });
    const { classes: progressIconClasses } = useComponentTheme.useComponentTheme(UNSAFE_ButtonLabelLayout_themes_redwood_ButtonLabelLayoutTheme.ButtonLabelLayoutRedwoodTheme, {
        target: 'progressIcon',
        size: size
    });
    return {
        startIconClasses,
        labelClasses,
        labelStyles,
        progressIconClasses,
        containerClasses,
        isShowIcons,
        isShowText,
        hasTextStartIcon
    };
}

exports.useProgressButtonLabelLayoutTheme = useProgressButtonLabelLayoutTheme;
