// @oracle/oraclejet-preact/cjs/useTextField-6ecf97b7.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var hooks = require('preact/hooks');
var stringUtils = require('./stringUtils-3e19c8af.js');
var truncationUtils = require('./truncationUtils-2d44b04b.js');
var useMessageSeverity = require('./useMessageSeverity-cfc4ae9d.js');
var useFocusWithin = require('./useFocusWithin-eb7f956a.js');
var useFormFieldContextProps = require('./useFormFieldContextProps-0fa67f6c.js');
var useId = require('./useId-6c0eeb27.js');

/**
 * A custom hook to determine the props for a component that renders
 * a text field
 */
function useTextField({ ariaDescribedBy, contentVariant, helpSourceLink, helpSourceText, isDisabled, isFocused, isLoading, isReadonly, isRequiredShown, labelEdge, messages, styleVariant, userAssistanceDensity, value }) {
    const baseId = useId.useId();
    const labelId = labelEdge !== 'none' ? `${baseId}-label` : undefined;
    const inputId = !isReadonly ? `${baseId}-input` : undefined;
    const uaId = !isDisabled ? `${baseId}-ua` : undefined;
    const helpIconId = userAssistanceDensity === 'compact' ? `${baseId}-helpIcon` : undefined;
    // JET-73197: Truncated Text Enhancements for Core Pack
    // Create a ref to the Label element. When we see focusWithin the input, measure the label
    // to determine if the text is truncated and if so, set showTruncatedTooltip to true.
    // If we see a blurWithin event and the input is no longer focused, set the flag to false.
    const labelRef = hooks.useRef(null);
    const [showTruncatedTooltip, setShowTruncatedTooltip] = hooks.useState(false);
    const onBlurWithin = hooks.useCallback(() => {
        setShowTruncatedTooltip(false);
    }, []);
    const onFocusWithin = hooks.useCallback(() => {
        const isTruncated = labelEdge === 'inside' && !isDisabled ? truncationUtils.detectTruncation(labelRef) : false;
        setShowTruncatedTooltip(isTruncated);
    }, [isDisabled, labelEdge, labelRef]);
    const { focusProps, isFocused: isInputFocused } = useFocusWithin.useFocusWithin({
        isDisabled,
        onBlurWithin,
        onFocusWithin
    });
    const hasValue = value !== undefined && (typeof value !== 'string' || value !== '');
    // Create the form field context
    const formFieldContext = useFormFieldContextProps.useFormFieldContextProps({
        hasTruncatedLabel: showTruncatedTooltip,
        hasValue,
        isDisabled,
        isFocused,
        isInputFocused,
        isLoading,
        isReadonly
    });
    // Determine the highest severity from the messages
    const messageSeverity = useMessageSeverity.useMessageSeverity(messages);
    // Determine the label variant
    const labelVariant = labelEdge !== 'none' ? labelEdge : undefined;
    return {
        baseId,
        formFieldContext,
        inputProps: {
            id: inputId,
            'aria-describedby': stringUtils.l([helpIconId, uaId, ariaDescribedBy]),
            'aria-invalid': messageSeverity === 'error'
                ? 'true'
                : undefined,
            variant: styleVariant,
            onBlur: focusProps?.onFocusOut,
            onFocus: focusProps?.onFocusIn
        },
        labelProps: {
            forId: inputId,
            helpIconId,
            helpSourceLink,
            helpSourceText,
            id: labelId,
            isRequiredShown,
            labelRef,
            userAssistanceDensity,
            variant: labelVariant
        },
        textFieldProps: {
            contentVariant,
            statusVariant: /error|warning/.test(messageSeverity ?? '')
                ? messageSeverity
                : undefined,
            styleVariant
        },
        userAssistanceProps: {
            id: uaId
        }
    };
}

exports.useTextField = useTextField;
