// @oracle/oraclejet-webdriver/src/elements/oj-drawer-popup/OjDrawerPopupBase.ts
import { OjWebElement } from '../';

/**
 * This is the base class for oj-drawer-popup WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjDrawerPopup.ts.
 */
export class OjDrawerPopupBase extends OjWebElement {
  /**
   * Gets the value of <code>opened</code> property.
   * Specifies the 'opened' state of the drawer.
   * @return The value of <code>opened</code> property.
   *
   */
  public getOpened(): Promise<boolean> {
    return this.getProperty<boolean>('opened');
  }

  /**
   * Gets the value of <code>edge</code> property.
   * Specifies at what edge the drawer opens.
   * @return The value of <code>edge</code> property.
   *
   */
  public getEdge(): Promise<string> {
    return this.getProperty<string>('edge');
  }

  /**
   * Gets the value of <code>modality</code> property.
   * Controls the modality of the drawer.
   * @return The value of <code>modality</code> property.
   *
   */
  public getModality(): Promise<string> {
    return this.getProperty<string>('modality');
  }

  /**
   * Gets the value of <code>autoDismiss</code> property.
   * Controls the close auto-dismiss behaviour to close the drawer.
   * @return The value of <code>autoDismiss</code> property.
   *
   */
  public getAutoDismiss(): Promise<string> {
    return this.getProperty<string>('autoDismiss');
  }

  /**
   * Gets the value of <code>closeGesture</code> property.
   * Controls the close gesture to close the drawer.
   * @return The value of <code>closeGesture</code> property.
   *
   */
  public getCloseGesture(): Promise<string> {
    return this.getProperty<string>('closeGesture');
  }
}
