// @oracle/oraclejet-webdriver/src/elements/oj-swipe-actions/OjSwipeActionsBase.ts
import { OjWebElement } from '../';

/**
 * This is the base class for oj-swipe-actions WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjSwipeActions.ts.
 */
export class OjSwipeActionsBase extends OjWebElement {
  /**
   * Gets the value of <code>translations</code> property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of <code>translations</code> property.
   *
   */
  public getTranslations(): Promise<Translations | null> {
    return this.getProperty<Translations | null>('translations');
  }
}
export interface Translations {
  /**
   * Label assigned to an anchor tag used for hidding start/end swipe actions using keyboard or accessibility agents such as VoiceOver.
   */
  ariaHideActionsDescription: string;
  /**
   * Label assigned to an anchor tag used for showing end swipe actions using keyboard or accessibility agents such as VoiceOver.
   */
  ariaShowEndActionsDescription: string;
  /**
   * Label assigned to an anchor tag used for showing start swipe actions using keyboard or accessibility agents such as VoiceOver.
   */
  ariaShowStartActionsDescription: string;
}
