// @oracle/oraclejet-preact/cjs/utils/PRIVATE_dayUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dayUtils = require('../dayUtils-78df6937.js');
var maskUtils = require('../maskUtils-3252a439.js');
require('../getFormatParse-88bacc78.js');
require('../types-55573a46.js');
require('../logger-2b636482.js');
require('../supplementalData-cdc1a84a.js');
require('../calendarDateUtils-32adc676.js');



exports.getDatesInWeek = dayUtils.getDatesInWeek;
exports.getDayOfWeek = dayUtils.getDayOfWeek;
exports.getDayOfWeekForCalendarGrid = dayUtils.getDayOfWeekForCalendarGrid;
exports.getDaysOfWeekNames = dayUtils.getDaysOfWeekNames;
exports.getRegion = dayUtils.getRegion;
exports.getStartOfWeek = dayUtils.getStartOfWeek;
exports.getWeekStart = dayUtils.getWeekStart;
exports.getWeeksInMonth = dayUtils.getWeeksInMonth;
exports.toIsoCalendarDateRequired = dayUtils.toIsoCalendarDateRequired;
exports.calendarDatesAreDifferent = maskUtils.calendarDatesAreDifferent;
exports.getMonthName = maskUtils.getMonthName;
exports.getTodayAsCalendarDate = maskUtils.getTodayAsCalendarDate;
exports.isNumberOnlyString = maskUtils.isNumberOnlyString;
exports.translateCustomMasks = maskUtils.translateCustomMasks;
