// @oracle/oraclejet-preact/cjs/utils/PRIVATE_tabbableUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tabbableUtils = require('../tabbableUtils-7e41d383.js');
require('../clientHints-c40c5250.js');
require('../head-68d0992f.js');
require('../_arity-c228159c.js');
require('../_curry3-18677bca.js');
require('../_curry1-e8f0d7ea.js');
require('../_curry2-c15d89cd.js');
require('../_isArray-73160ad5.js');
require('../_isString-f4443c9e.js');



exports.allTabbableElements = tabbableUtils.allTabbableElements;
exports.focusOn = tabbableUtils.focusOn;
exports.focusOnElementOrNearestAncestor = tabbableUtils.focusOnElementOrNearestAncestor;
exports.focusOnEnd = tabbableUtils.focusOnEnd;
exports.focusOnStart = tabbableUtils.focusOnStart;
exports.focusWithin = tabbableUtils.focusWithin;
exports.getActiveElement = tabbableUtils.getActiveElement;
exports.getAutofocusFocusables = tabbableUtils.getAutofocusFocusables;
exports.getBodyElement = tabbableUtils.getBodyElement;
exports.isTabbableElement = tabbableUtils.isTabbableElement;
