// @oracle/oraclejet-preact/cjs/utils/PRIVATE_collectionUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var collectionUtils = require('../collectionUtils-fe12edac.js');
require('../keys-25710458.js');
require('../clientHints-c40c5250.js');



exports.ITEM_SELECTOR = collectionUtils.ITEM_SELECTOR;
exports.compareListItemContext = collectionUtils.compareListItemContext;
exports.dragToScroll = collectionUtils.dragToScroll;
exports.findElementByKey = collectionUtils.findElementByKey;
exports.getElementContainsFunc = collectionUtils.getElementContainsFunc;
exports.getFirstVisibleKey = collectionUtils.getFirstVisibleKey;
exports.getItemKeysArray = collectionUtils.getItemKeysArray;
exports.getKey = collectionUtils.getKey;
exports.getPrevNextKey = collectionUtils.getPrevNextKey;
exports.getPrevNextKeyByCount = collectionUtils.getPrevNextKeyByCount;
exports.getPrevNextKeyUsingRef = collectionUtils.getPrevNextKeyUsingRef;
exports.getViewportConfig = collectionUtils.getViewportConfig;
exports.handleEnterActionableMode = collectionUtils.handleEnterActionableMode;
exports.handleSelectionRange = collectionUtils.handleSelectionRange;
exports.handleWrapActionableFocus = collectionUtils.handleWrapActionableFocus;
exports.isMobile = collectionUtils.isMobile;
exports.keyExtractor = collectionUtils.keyExtractor;
