// @oracle/oraclejet-preact/cjs/utils/PRIVATE_animationUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var animationUtils = require('../animationUtils-54852d03.js');



exports.ANCHOR_BOTTOM_CENTER = animationUtils.ANCHOR_BOTTOM_CENTER;
exports.ANCHOR_BOTTOM_LEFT = animationUtils.ANCHOR_BOTTOM_LEFT;
exports.ANCHOR_BOTTOM_RIGHT = animationUtils.ANCHOR_BOTTOM_RIGHT;
exports.ANCHOR_DEFAULT = animationUtils.ANCHOR_DEFAULT;
exports.ANCHOR_LEFT_CENTER = animationUtils.ANCHOR_LEFT_CENTER;
exports.ANCHOR_RIGHT_CENTER = animationUtils.ANCHOR_RIGHT_CENTER;
exports.ANCHOR_TOP_CENTER = animationUtils.ANCHOR_TOP_CENTER;
exports.ANCHOR_TOP_LEFT = animationUtils.ANCHOR_TOP_LEFT;
exports.ANCHOR_TOP_RIGHT = animationUtils.ANCHOR_TOP_RIGHT;
exports.DELAY_LARGE = animationUtils.DELAY_LARGE;
exports.DELAY_MEDIUM = animationUtils.DELAY_MEDIUM;
exports.DELAY_NONE = animationUtils.DELAY_NONE;
exports.DELAY_SMALL = animationUtils.DELAY_SMALL;
exports.DURATION_LARGE = animationUtils.DURATION_LARGE;
exports.DURATION_MEDIUM = animationUtils.DURATION_MEDIUM;
exports.DURATION_SMALL = animationUtils.DURATION_SMALL;
exports.DURATION_XLARGE = animationUtils.DURATION_XLARGE;
exports.DURATION_XSMALL = animationUtils.DURATION_XSMALL;
exports.DURATION_XXLARGE = animationUtils.DURATION_XXLARGE;
exports.DURATION_XXSMALL = animationUtils.DURATION_XXSMALL;
exports.EASING_ENTER = animationUtils.EASING_ENTER;
exports.EASING_EXIT = animationUtils.EASING_EXIT;
exports.EASING_LiNEAR = animationUtils.EASING_LiNEAR;
exports.EASING_PERSIST = animationUtils.EASING_PERSIST;
exports.FADE_IN_FULL = animationUtils.FADE_IN_FULL;
exports.FADE_IN_SMALL = animationUtils.FADE_IN_SMALL;
exports.FADE_OUT_FULL = animationUtils.FADE_OUT_FULL;
exports.FADE_OUT_SMALL = animationUtils.FADE_OUT_SMALL;
exports.SCALE_DOWN_FULL = animationUtils.SCALE_DOWN_FULL;
exports.SCALE_DOWN_MEDIUM = animationUtils.SCALE_DOWN_MEDIUM;
exports.SCALE_DOWN_SMALL = animationUtils.SCALE_DOWN_SMALL;
exports.SCALE_UP_FULL = animationUtils.SCALE_UP_FULL;
exports.SCALE_UP_MEDIUM = animationUtils.SCALE_UP_MEDIUM;
exports.SCALE_UP_SMALL = animationUtils.SCALE_UP_SMALL;
exports.SLIDE_DOWN_LARGE = animationUtils.SLIDE_DOWN_LARGE;
exports.SLIDE_DOWN_MEDIUM = animationUtils.SLIDE_DOWN_MEDIUM;
exports.SLIDE_DOWN_SMALL = animationUtils.SLIDE_DOWN_SMALL;
exports.SLIDE_DOWN_XLARGE = animationUtils.SLIDE_DOWN_XLARGE;
exports.SLIDE_END_LARGE = animationUtils.SLIDE_END_LARGE;
exports.SLIDE_END_MEDIUM = animationUtils.SLIDE_END_MEDIUM;
exports.SLIDE_END_SMALL = animationUtils.SLIDE_END_SMALL;
exports.SLIDE_END_XLARGE = animationUtils.SLIDE_END_XLARGE;
exports.SLIDE_START_LARGE = animationUtils.SLIDE_START_LARGE;
exports.SLIDE_START_MEDIUM = animationUtils.SLIDE_START_MEDIUM;
exports.SLIDE_START_SMALL = animationUtils.SLIDE_START_SMALL;
exports.SLIDE_START_XLARGE = animationUtils.SLIDE_START_XLARGE;
exports.SLIDE_UP_LARGE = animationUtils.SLIDE_UP_LARGE;
exports.SLIDE_UP_MEDIUM = animationUtils.SLIDE_UP_MEDIUM;
exports.SLIDE_UP_SMALL = animationUtils.SLIDE_UP_SMALL;
exports.SLIDE_UP_XLARGE = animationUtils.SLIDE_UP_XLARGE;
exports.mergeAnimationStateObjects = animationUtils.mergeAnimationStateObjects;
