// @oracle/oraclejet-preact/cjs/utils/PRIVATE_meterUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var meterUtils = require('../meterUtils-dc96aab8.js');
var circleUtils = require('../circleUtils-4af93948.js');
require('../utils-50db2fcf.js');
require('../themeContract.css-a53fd740.js');
require('../utils-a66430fc.js');
require('../colorUtils-01f3caa2.js');
require('../_curry1-e8f0d7ea.js');



exports.findThreshold = meterUtils.findThreshold;
exports.getMeterAriaProps = meterUtils.getMeterAriaProps;
exports.getThresholdColorFromValue = meterUtils.getThresholdColorFromValue;
exports.getTrackAndIndicatorColor = meterUtils.getTrackAndIndicatorColor;
exports.validateRange = meterUtils.validateRange;
exports.convertToPolar = circleUtils.convertToPolar;
exports.getCenterCoord = circleUtils.getCenterCoord;
exports.getCircleSection = circleUtils.getCircleSection;
exports.getClipPath = circleUtils.getClipPath;
exports.getPositiveAngle = circleUtils.getPositiveAngle;
exports.getProgressClipPath = circleUtils.getProgressClipPath;
