// @oracle/oraclejet-preact/cjs/utils/UNSAFE_color.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var swatchData = require('../swatchData-17c1d335.js');
require('../colorUtils-01f3caa2.js');
require('../_curry1-e8f0d7ea.js');



exports.colorsAreDifferent = swatchData.colorsAreDifferent;
exports.convertPalette = swatchData.convertPalette;
exports.defaultRedwoodNoFillPaletteValues = swatchData.defaultRedwoodNoFillPaletteValues;
exports.defaultRedwoodPaletteValues = swatchData.defaultRedwoodPaletteValues;
exports.getColorValue = swatchData.getColorValue;
exports.getEightDigitHex = swatchData.getEightDigitHex;
exports.getExpandedColorValue = swatchData.getExpandedColorValue;
exports.getSixDigitHex = swatchData.getSixDigitHex;
exports.isEightDigitHex = swatchData.isEightDigitHex;
exports.isValidColorValue = swatchData.isValidColorValue;
exports.parseColorValue = swatchData.parseColorValue;
exports.sortPalette = swatchData.sortPalette;
exports.uniqueColorMap = swatchData.uniqueColorMap;
