// @oracle/oraclejet-preact/cjs/utils/UNSAFE_timeUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var timeUtils = require('../timeUtils-ce57624b.js');
require('../getFormatParse-88bacc78.js');
require('../types-55573a46.js');
require('../logger-2b636482.js');
require('../calendarDateUtils-32adc676.js');



exports.AM_VALUE = timeUtils.AM_VALUE;
exports.PM_VALUE = timeUtils.PM_VALUE;
exports.compareTimes = timeUtils.compareTimes;
exports.displayedTimeSegmentsAreDifferent = timeUtils.displayedTimeSegmentsAreDifferent;
exports.formatIsoTimeStrAsExample = timeUtils.formatIsoTimeStrAsExample;
exports.formatTimeFromMasks = timeUtils.formatTimeFromMasks;
exports.getAmPmMatchType = timeUtils.getAmPmMatchType;
exports.getAmPmStringsForLocale = timeUtils.getAmPmStringsForLocale;
exports.getDayPeriodValueFromHour = timeUtils.getDayPeriodValueFromHour;
exports.getDayPeriodValueStr = timeUtils.getDayPeriodValueStr;
exports.getHourValueStr = timeUtils.getHourValueStr;
exports.getLeadingZeroForHour = timeUtils.getLeadingZeroForHour;
exports.getTimeMasksFromLocaleAndOptions = timeUtils.getTimeMasksFromLocaleAndOptions;
exports.getTimeObjFromTimeOnlyISOStr = timeUtils.getTimeObjFromTimeOnlyISOStr;
exports.getTimeOnlyISOStrFromTimeObj = timeUtils.getTimeOnlyISOStrFromTimeObj;
exports.getTimePlaceholdersFromPattern = timeUtils.getTimePlaceholdersFromPattern;
exports.hasSegmentsNotDisplayed = timeUtils.hasSegmentsNotDisplayed;
exports.padWithZero = timeUtils.padWithZero;
exports.replaceCustomMasksValuesWithPlaceholders = timeUtils.replaceCustomMasksValuesWithPlaceholders;
exports.timesAreDifferent = timeUtils.timesAreDifferent;
