// @oracle/oraclejet-preact/cjs/utils/UNSAFE_visUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var eventsUtils = require('../eventsUtils-d34f8647.js');
var textUtils = require('../textUtils-e779c845.js');
var colorUtils = require('../colorUtils-c97cf90b.js');
var accUtils = require('../accUtils-5c7d3255.js');
var utils = require('../utils-6f32ac40.js');
require('../utils-a66430fc.js');
require('../themeContract.css-a53fd740.js');
require('../colorUtils-01f3caa2.js');
require('../_curry1-e8f0d7ea.js');
require('../clientHints-c40c5250.js');



exports.cancelEvent = eventsUtils.cancelEvent;
exports.getElementData = eventsUtils.getElementData;
exports.getInfo = eventsUtils.getInfo;
exports.isHeightOverlap = textUtils.isHeightOverlap;
exports.isWidthOverlap = textUtils.isWidthOverlap;
exports.getColorRamp = colorUtils.getColorRamp;
exports.generateAriaLabel = accUtils.generateAriaLabel;
exports.supportsMobileScreenReader = accUtils.supportsMobileScreenReader;
exports.inSpace = utils.inSpace;
