// @oracle/oraclejet-preact/cjs/PRIVATE_ContainerDiagram/themes/DiagramStyles.css.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });



var buttonStyle = 'DiagramStyles_buttonStyle__amen395';
var childContentStyle = 'DiagramStyles_childContentStyle__amen397';
var dimensionStyle = 'DiagramStyles_dimensionStyle__amen39a';
var itemStyles = {nodeStyles:'DiagramStyles_nodeStyles__amen392',linkStyles:'DiagramStyles_linkStyles__amen393',dimmedItemStyle:'DiagramStyles_dimmedItemStyle__amen398'};
var labelStyles = {labelStyle:'DiagramStyles_labelStyle__amen394',labelBorderStyle:'DiagramStyles_labelBorderStyle__amen396'};
var layerStyles = 'DiagramStyles_layerStyles__amen391';
var outerStyles = 'DiagramStyles_outerStyles__amen390';
var panZoomStyles = 'DiagramStyles_panZoomStyles__amen399';

exports.buttonStyle = buttonStyle;
exports.childContentStyle = childContentStyle;
exports.dimensionStyle = dimensionStyle;
exports.itemStyles = itemStyles;
exports.labelStyles = labelStyles;
exports.layerStyles = layerStyles;
exports.outerStyles = outerStyles;
exports.panZoomStyles = panZoomStyles;
