// @oracle/oraclejet-preact/es/UNSAFE_BaseButton/themes/stable/BaseButtonTheme.lg.tokens.js
/* @oracle/oraclejet-preact: undefined */
import { x as xUnits, b as max, d as densityUnits } from '../../../utils-cfef1929.js';
import { b as baseVars, t as typographyVars } from '../../../themeContract.css-2b5a42e2.js';
import '../../../colorUtils-b3804ee8.js';
import '../../../_curry1-63949a9b.js';

const scaleSchemeLg = {
    borderRadius: baseVars.borderRadius.md,
    fontSize: typographyVars.body.sm.fontSize,
    sizes: {
        xs: {
            textToEdgePadding: xUnits(4),
            // 24px x 24px is the minimum size per WCAG Success Criterion 2.5.8 Target Size (Minimum)
            height: max(`calc(${xUnits(6)} + ${densityUnits(3)})`, '24px')
        },
        sm: {
            textToEdgePadding: xUnits(4),
            // 24px x 24px is the minimum size per WCAG Success Criterion 2.5.8 Target Size (Minimum)
            height: max(`calc(${xUnits(9)} + ${densityUnits(3)})`, '24px')
        },
        md: {
            textToEdgePadding: xUnits(4),
            height: `calc(${xUnits(11)} + ${densityUnits(2)})`
        },
        lg: {
            textToEdgePadding: xUnits(6),
            height: `calc(${xUnits(13)} + ${densityUnits(2)})`
        }
    }
};

export { scaleSchemeLg };
