// @oracle/oraclejet-preact/es/UNSAFE_BaseButton/themes/stable/BaseButtonTheme.light.tokens.js
/* @oracle/oraclejet-preact: undefined */
import { a as rgb, r as rgba } from '../../../utils-cfef1929.js';
import { c as colorSchemeVars } from '../../../themeContract.css-2b5a42e2.js';
import '../../../colorUtils-b3804ee8.js';
import '../../../_curry1-63949a9b.js';

const colorSchemeLight = {
    variants: {
        outlined: {
            base: {
                backgroundColor: 'transparent',
                borderColor: colorSchemeVars.border.enabled,
                color: colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: 'transparent',
                borderColor: colorSchemeVars.border.disabled,
                color: colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: colorSchemeVars.overlay.hover
            },
            active: {
                backgroundOverlayColor: colorSchemeVars.overlay.active
            }
        },
        callToAction: {
            base: {
                backgroundColor: rgb(colorSchemeVars.palette.neutral[170]),
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.inverse
            },
            disabled: {
                backgroundColor: colorSchemeVars.surface.disabled,
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: colorSchemeVars.overlay.inverseHover
            },
            active: {
                backgroundOverlayColor: colorSchemeVars.overlay.inverseActive
            }
        },
        borderless: {
            base: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: colorSchemeVars.overlay.hover
            },
            active: {
                backgroundOverlayColor: colorSchemeVars.overlay.active
            }
        },
        ghost: {
            base: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: 'transparent',
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: 'transparent'
            },
            active: {
                backgroundOverlayColor: 'transparent'
            }
        },
        solid: {
            base: {
                backgroundColor: rgba(colorSchemeVars.palette.neutral['190'], 0.08),
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.primary
            },
            disabled: {
                backgroundColor: colorSchemeVars.surface.disabled,
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: colorSchemeVars.overlay.hover
            },
            active: {
                backgroundOverlayColor: colorSchemeVars.overlay.active
            }
        },
        danger: {
            base: {
                backgroundColor: colorSchemeVars.surface.danger.strong,
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.onColor
            },
            disabled: {
                backgroundColor: colorSchemeVars.surface.disabled,
                borderColor: 'transparent',
                color: colorSchemeVars.textIcon.disabled
            },
            hover: {
                backgroundOverlayColor: colorSchemeVars.overlay.dangerHover
            },
            active: {
                backgroundOverlayColor: colorSchemeVars.overlay.dangerActive
            }
        }
    }
};

export { colorSchemeLight };
