// @oracle/oraclejet-preact/es/PRIVATE_MonthView/themes/redwood/MonthViewTheme.tokens.js
/* @oracle/oraclejet-preact: undefined */
import { t as typographyVars, c as colorSchemeVars } from '../../../themeContract.css-2b5a42e2.js';
import { x as xUnits } from '../../../utils-cfef1929.js';
import { e as extendThemeContractVars } from '../../../mergeDeepRight-f0f42e67.js';
import '../../../colorUtils-b3804ee8.js';
import '../../../_curry1-63949a9b.js';
import '../../../_curry2-01c87f61.js';
import '../../../mergeDeepWithKey-33e3d1be.js';
import '../../../_curry3-577eb45e.js';
import '../../../_isObject-1fab0f5b.js';

// color scheme default vars
const monthViewColorSchemeDefault = {
    dayOfTheWeek: {
        color: colorSchemeVars.textIcon.secondary
    }
};
// color scheme light extensions
const monthViewColorSchemeLight = extendThemeContractVars(monthViewColorSchemeDefault, {}); // No overrides
// color scheme dark extensions
const monthViewColorSchemeDark = extendThemeContractVars(monthViewColorSchemeDefault, {}); // No overrides
// scale default vars
const monthViewScaleSchemeDefault = {
    rowGap: xUnits(3),
    columnGap: xUnits(0),
    dayOfTheWeek: {
        lineHeight: `calc(2 * ${typographyVars.body.xs.fontSize})`,
        fontSize: typographyVars.body.xs.fontSize,
        fontWeight: typographyVars.body.xs.fontWeight.semibold
    }
};
// scale lg extensions
const monthViewScaleSchemeLg = extendThemeContractVars(monthViewScaleSchemeDefault, {}); // No overrides
// scale md extensions
const monthViewScaleSchemeMd = extendThemeContractVars(monthViewScaleSchemeDefault, {}); // No overrides
// scale sm extensions
const monthViewScaleSchemeSm = extendThemeContractVars(monthViewScaleSchemeDefault, {}); // No overrides

export { monthViewColorSchemeDark, monthViewColorSchemeDefault, monthViewColorSchemeLight, monthViewScaleSchemeDefault, monthViewScaleSchemeLg, monthViewScaleSchemeMd, monthViewScaleSchemeSm };
