// @oracle/oraclejet-preact/es/resources/nls/no/bundle.js
/* @oracle/oraclejet-preact: undefined */
// This file is auto-generated by @oracle/oraclejet-icu-l10n from src/resources/icu/bundle.json
// and should not be edited by hand. Update the JSON file and rerun the build to regenerate this content.
const bundle = {
    accAnnounceMovedKey: (p) => 'Flyttet ' + p['itemKey'] + ' ' + p['position'] + ' ' + p['referenceKey'] + '.',
    accButtonProgressComplete: (p) => p['label'] + ' – handlingen er fullført',
    accButtonProgressStatus: (p) => p['label'] + ' – handlingen pågår',
    accColorPalette: () => 'Fargepalett',
    accColorSwatch: () => 'Fargekart',
    accCommandKey: () => 'Kommando',
    accContextMenuLabel: () => 'handlinger',
    accControlKey: () => 'Lukk',
    accPositionAfter: () => 'etter',
    accPositionBefore: () => 'før',
    accReorderInstructions: (p) => 'Trykk på og hold nede tastene ' +
        p['keyPressed'] +
        ' og Skift, og bruk deretter piltastene for å endre rekkefølgen.',
    calendarDateConverter_parseError: (p) => 'Angi en fullstendig dato som dette: ' + p['dateExample'] + '.',
    calendarMonthConverter_parseError: (p) => 'Angi en måned og et år som dette: ' + p['dateExample'] + '.',
    chart_labelGroup: () => 'Gruppe',
    chart_labelSeries: () => 'Serie',
    chart_labelValue: () => 'Verdi',
    chart_labelWidthWeight: () => 'Z',
    chart_labelX: () => 'X',
    chart_labelY: () => 'Y',
    checkboxSet_requiredMessageDetail: () => 'Velg en verdi.',
    checkbox_requiredMessageDetail: () => 'Velg en verdi.',
    close: () => 'Lukk',
    collection_apply: () => 'Bruk',
    collection_cancel: () => 'Avbryt',
    collection_loaded: () => 'Lastet',
    collection_loading: () => 'Laster',
    collection_noData: () => 'Ingen data å vise.',
    collection_pixels: () => 'piksler',
    collection_resizeColumnDialog: () => 'Endre størrelse på kolonne',
    collection_resizeColumnMenu: () => 'Endre størrelse på kolonne...',
    collection_resizeMessageDetail: (p) => 'Angi et tall mellom ' + p['MIN'] + ' og ' + p['MAX'] + '.',
    collection_selectAllRows: () => 'Velg alle rader',
    collection_selectRow: (p) => 'Velg raden ' + p['ROW_NAME'],
    collection_sortAscendingSubMenu: () => 'Stigende',
    collection_sortColumnMenu: () => 'Sorter kolonne',
    collection_sortDescendingSubMenu: () => 'Synkende',
    collection_sortDisabled: () => 'Ikke sorterbar',
    collection_sortEnabled: () => 'Sorterbar',
    collection_width: () => 'Bredde',
    colorPicker_label: () => 'Fargevelger',
    confirmation: () => 'Bekreftelse',
    datePicker_currentDate: () => 'Dagens dato',
    datePicker_goToToday: () => 'Gå til i dag',
    datePicker_label: () => 'Datovelger',
    datePicker_month: () => 'Måned',
    datePicker_months: () => 'Måneder',
    datePicker_next: () => 'Neste',
    datePicker_nextDecade: () => 'Neste tiår',
    datePicker_nextMonth: () => 'Neste måned',
    datePicker_prev: () => 'Forrige',
    datePicker_prevDecade: () => 'Forrige tiår',
    datePicker_prevMonth: () => 'Forrige måned',
    datePicker_restricted: () => 'Begrenset',
    datePicker_selected: () => 'Valgt',
    datePicker_today: () => 'I dag',
    datePicker_weekNum: (p) => 'Uke ' + p['NUM'],
    datePicker_weekNumColLabel: () => 'Ukenummer',
    datePicker_year: () => 'År',
    datePicker_years: () => 'År',
    error: () => 'Feil',
    expandableList_expandCollapseInstructionText: () => 'Bruk piltastene for å utvide og trekke sammen',
    expandableList_groupCollapse: () => 'Sammentrukket',
    expandableList_groupExpand: () => 'Utvidet',
    filePicker_addFiles: () => 'Legg til filer',
    filePicker_dropzonePrimaryText: () => 'Dra og slipp',
    filePicker_dropzoneSecondaryText: () => 'Velg en fil, eller slipp en fil her',
    filePicker_dropzoneSecondaryTextMultiple: () => 'Velg eller slipp filer her',
    filePicker_multipleFileTypeUploadError: (p) => 'Du kan ikke laste filer av følgende typer: ' + p['fileTypes'],
    filePicker_singleFileUploadError: () => 'Last én fil om gangen',
    filePicker_singleTypeUploadError: (p) => 'Du kan ikke laste filer av typen ' + p['fileType'],
    filePicker_unknownFileTypeUploadError: () => 'ukjent',
    formControl_clear: () => 'Nullstill',
    formControl_day: () => 'Dag',
    formControl_empty_segment: () => 'Tomt',
    formControl_helpAvailable: () => 'Hjelp – tilgjengelig',
    formControl_hide: () => 'Skjul',
    formControl_hour: () => 'Time',
    formControl_limitReached: (p) => 'Du har nådd grensen på ' +
        ({ one: new Intl.NumberFormat('no').format(p['CHARACTER_LIMIT']) + '-tegn' }[new Intl.PluralRules('no').select(p['CHARACTER_LIMIT'])] ||
            new Intl.NumberFormat('no').format(p['CHARACTER_LIMIT']) + '-tegn') +
        '.',
    formControl_loading: () => 'Laster',
    formControl_maxLength: (p) => 'Du kan angi opptil ' + p['MAX_LENGTH'] + ' tegn.',
    formControl_maxLengthExceeded: (p) => 'Maksimumslengden, ' + p['MAX_LENGTH'] + ', er overskredet.',
    formControl_maxLengthRemaining: (p) => p['CHARACTER_COUNT'] + ' tegn igjen.',
    formControl_millisecond: () => 'Millisekund',
    formControl_minute: () => 'Minutt',
    formControl_month: () => 'Måned',
    formControl_readOnly: () => 'Skrivebeskyttet',
    formControl_requiredMessageDetail: () => 'Angi en verdi.',
    formControl_second: () => 'Sekund',
    formControl_selectionExactMessageDetail: (p) => 'Velg ' + p['n'] + ' verdier.',
    formControl_selectionOverflowMessageDetail: (p) => 'Du har nådd utvalgsgrensen på ' + p['max'] + '.',
    formControl_selectionRangeMessageDetail: (p) => 'Velg fra ' + p['min'] + ' til ' + p['max'] + ' verdier.',
    formControl_selectionUnderflowMessageDetail: (p) => 'Velg minst ' + p['min'] + ' verdier.',
    formControl_show: () => 'Vis',
    formControl_year: () => 'År',
    gantt_labelDate: () => 'Dato',
    gantt_labelEnd: () => 'Slutt',
    gantt_labelLabel: () => 'Etikett',
    gantt_labelRow: () => 'Rad',
    gantt_labelStart: () => 'Start',
    indexer_disabledLabel: (p) => p['SECTION'] + '. Finner ingen samsvarende seksjon.',
    indexer_keyboardInstructionText: () => 'Trykk på Enter hvis du vil velge verdien.',
    indexer_othersLabel: () => '#',
    indexer_separatorLabel: (p) => 'Mellom ' + p['FROM_SECTION'] + ' og ' + p['TO_SECTION'] + '.',
    indexer_touchInstructionText: () => 'Dobbelttrykk og hold, slik at bevegelsesmodus aktiveres, og dra deretter opp og ned hvis du vil justere verdien.',
    info: () => 'Opplysninger',
    inputColor_selectColor: () => 'Velg en farge',
    inputDateMask_date_cleared: () => 'Dato nullstilt',
    inputDateMask_dayPlaceholder: () => 'dd',
    inputDateMask_empty_segment: () => 'Tomt',
    inputDateMask_monthPlaceholder: () => 'mm',
    inputDateMask_yearPlaceholder: () => 'åååå',
    inputDatePicker_dateRangeOverflowMessageDetail: (p) => 'Angi en dato som er på eller før tidspunktet ' + p['max'] + '.',
    inputDatePicker_dateRangeUnderflowMessageDetail: (p) => 'Angi en dato som er på eller etter tidspunktet ' + p['min'] + '.',
    inputDatePicker_dateRestrictionMessageDetail: () => 'Angi en tilgjengelig dato.',
    inputDatePicker_instruction: () => 'Trykk på tabulatortasten hvis du vil gå til kalenderen.',
    inputDatePicker_selectDate: () => 'Velg en dato',
    inputMonthMask_dateRangeOverflowMessageDetail: (p) => 'Angi en måned og et år som er på eller før tidspunktet ' + p['max'] + '.',
    inputMonthMask_dateRangeUnderflowMessageDetail: (p) => 'Angi en måned og et år som er på eller etter tidspunktet ' + p['min'] + '.',
    inputNumber_converterParseError: () => 'Angi et tall.',
    inputNumber_decrease: () => 'Reduser',
    inputNumber_increase: () => 'Øk',
    inputPassword_hidden: () => 'Passord skjult',
    inputPassword_hide: () => 'Skjul passord',
    inputPassword_show: () => 'Vis passord',
    inputSensitiveText_hidden: () => 'Sensitiv tekst skjult',
    inputTimeMask_timeRangeOverflowMessageDetail: (p) => 'Angi et klokkeslett som er ' + p['max'] + ' eller tidligere.',
    inputTimeMask_timeRangeUnderflowMessageDetail: (p) => 'Angi et klokkeslett som er ' + p['min'] + ' eller senere.',
    inputTimeMask_time_cleared: () => 'Tidspunkt for nullstilling',
    labelNone: () => 'Ingen',
    list_msgFetchCompleted: () => 'Lastet',
    list_msgFetchingData: () => 'Laster',
    list_suggestion: () => 'Forslag',
    menu_stateLoading: () => 'Laster',
    messageToast_allMessagesClosed: () => 'Alle sprettoppvarsler er lukket.',
    message_close: () => 'Lukk',
    message_confirmation: () => 'Vellykket',
    message_error: () => 'Feil',
    message_info: () => 'Opplysninger',
    message_navigationFromMessagesRegion: () => 'Åpner meldingsområde. Trykk på F6 hvis du vil navigere tilbake til tidligere fokusert element.',
    message_navigationToMessagesRegion: () => 'Meldingsområdet inneholder nye meldinger. Trykk på F6 hvis du vil navigere til de nyeste meldingene.',
    message_warning: () => 'Advarsel',
    noData_message: () => 'Ingen elementer å vise',
    overflowItemLabel: () => 'Flere faner',
    plural_separator: () => ', ',
    popup_close_skipLink: () => 'Dobbelttrykk hvis du vil lukke det åpne hurtigvinduet.',
    popup_liveRegion_F6_initial_focus_firstFocusable: () => 'Åpner hurtigvindu. Trykk på F6 hvis du vil navigere mellom hurtigvinduet og den tilhørende kontrollen.',
    popup_liveRegion_F6_initial_focus_none: () => 'Hurtigvinduet er åpnet. Trykk på F6 hvis du vil navigere mellom hurtigvinduet og den tilhørende kontrollen.',
    progressIndeterminate: () => 'Pågår',
    radio_helpForRadio: () => 'Hjelp for alternativknapp',
    radio_requiredMessageDetail: () => 'Velg en verdi.',
    selectMultiple_apply: () => 'Bruk',
    selectMultiple_back: () => 'Tilbake',
    selectMultiple_countPlus: (p) => p['COUNT'] + '+',
    selectMultiple_highlightSelectedTagsInstructionText: () => 'Bruk venstre og høyre piltast for å utheve valgte verdier.',
    selectMultiple_removeSelectedTagInstructionText: () => 'Trykk på Tilbake-tasten eller slettetasten hvis du vil fjerne den valgte verdien.',
    selectMultiple_selectedValues: () => 'Valgte verdier',
    selectMultiple_showSelectedValues: () => 'Vis bare valgte verdier.',
    selectMultiple_valuesSelected: (p) => p['VALUE_COUNT'] + ' verdier er valgt.',
    select_addToList: () => 'Legg til i liste',
    select_addToListAvailable: () => 'Legg til i liste – tilgjengelig',
    select_moreSearchOptions: () => 'Flere søkevalg',
    select_moreSearchOptionsAvailable: () => 'Flere søkevalg – tilgjengelig',
    select_noMatchesFound: () => 'Ingen samsvar funnet.',
    select_oneMatchFound: () => 'Finner ett treff',
    select_requiredMessageDetail: () => 'Velg en verdi.',
    select_sizeMatchesFound: (p) => p['TOTAL_SIZE'] + ' samsvar funnet',
    select_sizeOrMoreMatchesFound: (p) => p['TOTAL_SIZE'] + ' eller flere samsvar funnet',
    selectorAll_deselectAll: () => 'Opphev alle valg',
    selectorAll_selectAll: () => 'Velg alle',
    selector_selected: () => 'Avmerket',
    selector_unselected: () => 'Ikke avmerket',
    sheet_close: () => 'Lukk',
    tabBarNavigationList_removeCueText: () => 'Kan fjernes',
    tabBar_tabContainsErrors: () => 'Inneholder feil',
    tabbar_accAnnounceRemovedKey: (p) => 'Fjernet ' + p['itemKey'],
    tabbar_labelContextMenu: () => 'har kontekstmeny',
    tabbar_labelMoveLeft: () => 'Flytt til venstre',
    tabbar_labelMoveRight: () => 'Flytt til høyre',
    tabbar_labelRemove: () => 'Fjern',
    timeComponent_tooltipZoomIn: () => 'Zoom inn',
    timeComponent_tooltipZoomOut: () => 'Zoom ut',
    timeConverter_parseError: () => 'Angi et fullstendig klokkeslett.',
    train_current: () => 'Gjeldende',
    train_message_type: () => 'Meldingstype',
    train_not_visited: () => 'Ikke besøkt',
    train_status: (p) => 'Trinn ' + p['currentStep'] + ' av ' + p['numberOfSteps'],
    train_visited: () => 'Besøkt',
    userAssistance_learnMore: () => 'Jeg vil vite mer',
    userAssistance_required: () => 'Nødvendig',
    vis_clearSelection: () => 'Nullstill valg',
    vis_drillable: () => 'Drillbar',
    vis_labelAndValue: (p) => p['LABEL'] + ': ' + p['VALUE'],
    vis_labelCountWithTotal: (p) => p['itemCount'] + ' av ' + p['totalCount'],
    vis_marqueeSelectTooltip: () => 'Valg med markeringshåndtak',
    vis_marqueeZoomTooltip: () => 'Zoom med markeringshåndtak',
    vis_noData: () => 'Ingen data å vise',
    vis_panTooltip: () => 'Panorer',
    vis_stateLoaded: () => 'Lastet',
    vis_stateLoading: () => 'Laster',
    vis_stateSelected: () => 'Valgt',
    vis_stateUnselected: () => 'Ikke valgt',
    warn: () => 'Advarsel'
};

export { bundle as default };
