// @oracle/oraclejet-jest-preset/src/preset.js
const jestPreset = require('jest-preset-preact');
const path = require('path');

const mappedModules = {
  // Alias to allow ojthemeutils to fetch CSS w/o being intercepted by
  // identify-obj-proxy mapping below
  '^__oj-theme-css$': '@oracle/oraclejet/dist/css/redwood/oj-redwood.css',
  // Noop style files
  '^.+\\.(css|sass|scss|less)$': 'identity-obj-proxy',
  '^css!.*': 'identity-obj-proxy',
  '^ojcss!.*': 'identity-obj-proxy',
  // Remove text! prefix and use loaders to transform content to text for things
  // typically loaded with this plugin (text & html)
  '^text!(.*)': '$1',
  // Mock modules
  touchr: path.resolve(__dirname, 'mocks', 'blank'),
  ojdnd: path.resolve(__dirname, 'mocks', 'blank'),
  'ojs/(.*)': [
    // Resolve ojs/* to any mock modules first
    path.resolve(__dirname, 'mocks', '$1'),
    '@oracle/oraclejet/dist/js/libs/oj/debug/$1'
  ],
  'ojL10n!./(.*)': './$1',
  'ojL10n!ojtranslations/nls/(.*)': '@oracle/oraclejet/dist/js/libs/oj/resources/nls/$1',
  'jqueryui-amd/(.*)': '@oracle/oraclejet/dist/js/libs/jquery/jqueryui-amd-1.14.1/$1',
  '^oj-c/(.*)': '@oracle/oraclejet-core-pack/oj-c/$1'
};

Object.assign(jestPreset, {
  modulePaths: ['@oracle/oraclejet/dist/js/libs/oj/'],
  transform: {
    [`\\${path.sep}resources\\${path.sep}nls\\${path.sep}`]:
      require.resolve('./loaders/ojL10n-loader'),
    // Use our own Babel configuration
    '^.+\\.(mjs|js|jsx|ts|tsx)$': require.resolve('./babel.js'),
    // Handle loader.ts which loads component.json and expects string instead of
    // JSON object
    'component\\.json$': require.resolve('./loaders/json-text-loader'),
    '\\.html$': require.resolve('./loaders/jest-raw-loader')
  },
  transformIgnorePatterns: [
    'node_modules/(?!@oracle/(oraclejet|oraclejet-core-pack)/)',
    '^.+\\.(css|sass|scss|less)$'
  ],
  setupFiles: [require.resolve('./setup.js')]
});
Object.assign(jestPreset.moduleNameMapper, mappedModules);

module.exports = jestPreset;
