// @oracle/oraclejet-webdriver/elements/oj-drawer-popup/OjDrawerPopupBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjDrawerPopupBase = void 0;
const __1 = require("../");
/**
 * This is the base class for oj-drawer-popup WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjDrawerPopup.ts.
 */
class OjDrawerPopupBase extends __1.OjWebElement {
    /**
     * Gets the value of <code>opened</code> property.
     * Specifies the 'opened' state of the drawer.
     * @return The value of <code>opened</code> property.
     *
     */
    getOpened() {
        return this.getProperty('opened');
    }
    /**
     * Gets the value of <code>edge</code> property.
     * Specifies at what edge the drawer opens.
     * @return The value of <code>edge</code> property.
     *
     */
    getEdge() {
        return this.getProperty('edge');
    }
    /**
     * Gets the value of <code>modality</code> property.
     * Controls the modality of the drawer.
     * @return The value of <code>modality</code> property.
     *
     */
    getModality() {
        return this.getProperty('modality');
    }
    /**
     * Gets the value of <code>autoDismiss</code> property.
     * Controls the close auto-dismiss behaviour to close the drawer.
     * @return The value of <code>autoDismiss</code> property.
     *
     */
    getAutoDismiss() {
        return this.getProperty('autoDismiss');
    }
    /**
     * Gets the value of <code>closeGesture</code> property.
     * Controls the close gesture to close the drawer.
     * @return The value of <code>closeGesture</code> property.
     *
     */
    getCloseGesture() {
        return this.getProperty('closeGesture');
    }
}
exports.OjDrawerPopupBase = OjDrawerPopupBase;
