// @oracle/oraclejet-preact/es/resources/nls/he/bundle.js
/* @oracle/oraclejet-preact: undefined */
// This file is auto-generated by @oracle/oraclejet-icu-l10n from src/resources/icu/bundle.json
// and should not be edited by hand. Update the JSON file and rerun the build to regenerate this content.
const bundle = {
    accAnnounceMovedKey: (p) => 'הועבר ' + p['itemKey'] + ' ' + p['position'] + ' ' + p['referenceKey'] + '.',
    accButtonProgressComplete: (p) => 'פעולה ' + p['label'] + ' הושלמה',
    accButtonProgressStatus: (p) => 'פעולה ' + p['label'] + ' בתהליך',
    accColorPalette: () => 'לוח צבעים',
    accColorSwatch: () => 'Swatch של צבעים',
    accCommandKey: () => 'פקודה',
    accContextMenuLabel: () => 'פעולות',
    accControlKey: () => 'סגור',
    accPositionAfter: () => 'אחרי',
    accPositionBefore: () => 'לפני',
    accReorderInstructions: (p) => 'הקש והחזק את ' +
        p['keyPressed'] +
        ' ואת מקשי Shift, ולאחר מכן השתמש במקשי החצים כדי לשנות את הסדר.',
    calendarDateConverter_parseError: (p) => 'הזן תאריך מלא כך: ' + p['dateExample'] + '.',
    calendarMonthConverter_parseError: (p) => 'הזן חודש ושנה כך: ' + p['dateExample'] + '.',
    chart_labelGroup: () => 'קבוצה',
    chart_labelSeries: () => 'סדרה',
    chart_labelValue: () => 'ערך',
    chart_labelWidthWeight: () => 'Z',
    chart_labelX: () => 'X',
    chart_labelY: () => 'Y',
    checkboxSet_requiredMessageDetail: () => 'בחר ערך',
    checkbox_requiredMessageDetail: () => 'בחר ערך',
    close: () => 'סגור',
    collection_apply: () => 'החל',
    collection_cancel: () => 'בטל',
    collection_loaded: () => 'נטען',
    collection_loading: () => 'טוען',
    collection_noData: () => 'אין נתונים להצגה.',
    collection_pixels: () => 'פיקסל',
    collection_resizeColumnDialog: () => 'שנה גודל עמודה',
    collection_resizeColumnMenu: () => 'שנה גודל עמודה...',
    collection_resizeMessageDetail: (p) => 'הזן מספר בין ' + p['MIN'] + ' לבין ' + p['MAX'] + '.',
    collection_selectAllRows: () => 'בחר את כל השורות',
    collection_selectRow: (p) => 'בחר את שורה ' + p['ROW_NAME'],
    collection_sortAscendingSubMenu: () => 'סדר עולה',
    collection_sortColumnMenu: () => 'מיין עמודה',
    collection_sortDescendingSubMenu: () => 'סדר יורד',
    collection_sortDisabled: () => 'לא ניתן למיון',
    collection_sortEnabled: () => 'ניתן למיון',
    collection_width: () => 'רוחב',
    colorPicker_label: () => 'בורר צבעים',
    confirmation: () => 'אישור',
    datePicker_currentDate: () => 'תאריך נוכחי',
    datePicker_goToToday: () => 'עבור להיום',
    datePicker_label: () => 'בורר תאריכים',
    datePicker_month: () => 'חודש',
    datePicker_months: () => 'חודשים',
    datePicker_next: () => 'הבא',
    datePicker_nextDecade: () => 'העשור הבא',
    datePicker_nextMonth: () => 'החודש הבא',
    datePicker_prev: () => 'הקודם',
    datePicker_prevDecade: () => 'העשור הקודם',
    datePicker_prevMonth: () => 'החודש הקודם',
    datePicker_restricted: () => 'מוגבל',
    datePicker_selected: () => 'נבחר',
    datePicker_today: () => 'היום',
    datePicker_weekNum: (p) => 'שבוע ' + p['NUM'],
    datePicker_weekNumColLabel: () => 'מספר שבוע',
    datePicker_year: () => 'שנה',
    datePicker_years: () => 'שנים',
    error: () => 'שגיאה',
    expandableList_expandCollapseInstructionText: () => 'יש להשתמש במקשי החצים כדי להרחיב ולצמצם',
    expandableList_groupCollapse: () => 'מכווץ',
    expandableList_groupExpand: () => 'מורחב',
    filePicker_addFiles: () => 'הוסף קבצים',
    filePicker_dropzonePrimaryText: () => 'גרור ושחרר',
    filePicker_dropzoneSecondaryText: () => 'בחר קובץ או שחרר אותו כאן',
    filePicker_dropzoneSecondaryTextMultiple: () => 'בחר קבצים או שחרר אותם כאן',
    filePicker_multipleFileTypeUploadError: (p) => 'אי אפשר להעלות קבצים מסוג: ' + p['fileTypes'],
    filePicker_singleFileUploadError: () => 'טען קובץ אחד בכל פעם',
    filePicker_singleTypeUploadError: (p) => 'אי אפשר להעלות קבצים מסוג ' + p['fileType'],
    filePicker_unknownFileTypeUploadError: () => 'לא ידוע',
    formControl_clear: () => 'מחק',
    formControl_day: () => 'יום',
    formControl_empty_segment: () => 'ריק',
    formControl_helpAvailable: () => 'עזרה זמינה',
    formControl_hide: () => 'הסתר',
    formControl_hour: () => 'שעה',
    formControl_limitReached: (p) => 'הגעת למגבלה של ' +
        ({
            one: new Intl.NumberFormat('he').format(p['CHARACTER_LIMIT']) + '-תווים',
            two: new Intl.NumberFormat('he').format(p['CHARACTER_LIMIT']) + '-תווים'
        }[new Intl.PluralRules('he').select(p['CHARACTER_LIMIT'])] ||
            new Intl.NumberFormat('he').format(p['CHARACTER_LIMIT']) + '-תווים') +
        '.',
    formControl_loading: () => 'טוען',
    formControl_maxLength: (p) => 'אתה יכול להזין עד ' + p['MAX_LENGTH'] + ' תווים.',
    formControl_maxLengthExceeded: (p) => 'חריגה מהאורך המרבי ' + p['MAX_LENGTH'] + '.',
    formControl_maxLengthRemaining: (p) => p['CHARACTER_COUNT'] + ' תווים נותרו.',
    formControl_millisecond: () => 'מילישניות',
    formControl_minute: () => 'דקה',
    formControl_month: () => 'חודש',
    formControl_readOnly: () => 'קריאה בלבד',
    formControl_requiredMessageDetail: () => 'הזן ערך.',
    formControl_second: () => 'שנייה',
    formControl_selectionExactMessageDetail: (p) => 'בחר ' + p['n'] + ' ערכים.',
    formControl_selectionOverflowMessageDetail: (p) => 'הגעת למגבלה של ' + p['max'] + '-בחירה.',
    formControl_selectionRangeMessageDetail: (p) => 'בחר ערכי ' + p['min'] + ' עד ' + p['max'] + '.',
    formControl_selectionUnderflowMessageDetail: (p) => 'בחר לפחות ' + p['min'] + ' ערכים.',
    formControl_show: () => 'הצגה',
    formControl_year: () => 'שנה',
    gantt_labelDate: () => 'תאריך',
    gantt_labelEnd: () => 'סיום',
    gantt_labelLabel: () => 'תווית',
    gantt_labelRow: () => 'שורה',
    gantt_labelStart: () => 'התחלה',
    indexer_disabledLabel: (p) => p['SECTION'] + '. לא נמצא פרק תואם.',
    indexer_keyboardInstructionText: () => 'לחץ על Enter כדי לבחור ערך.',
    indexer_othersLabel: () => '#',
    indexer_separatorLabel: (p) => 'בין ' + p['FROM_SECTION'] + ' ו-' + p['TO_SECTION'] + '.',
    indexer_touchInstructionText: () => 'הקש פעמיים והחזק כדי להיכנס למצב מחוות, ולאחר מכן גרור מעלה ומטה כדי להתאים את הערך.',
    info: () => 'מידע',
    inputColor_selectColor: () => 'בחר צבע',
    inputDateMask_date_cleared: () => 'תאריך נמחק',
    inputDateMask_dayPlaceholder: () => 'dd',
    inputDateMask_empty_segment: () => 'ריק',
    inputDateMask_monthPlaceholder: () => 'mm',
    inputDateMask_yearPlaceholder: () => 'yyyy',
    inputDatePicker_dateRangeOverflowMessageDetail: (p) => 'הזן תאריך שחל ב-' + p['max'] + ' או לפניו.',
    inputDatePicker_dateRangeUnderflowMessageDetail: (p) => 'הזן תאריך שחל ב-' + p['min'] + ' או אחריו.',
    inputDatePicker_dateRestrictionMessageDetail: () => 'הזן תאריך זמין.',
    inputDatePicker_instruction: () => 'הקש על Tab כדי לגשת ללוח השנה.',
    inputDatePicker_selectDate: () => 'בחר תאריך',
    inputMonthMask_dateRangeOverflowMessageDetail: (p) => 'הזן חודש ושנה בתאריך ' + p['max'] + ' או לפניו.',
    inputMonthMask_dateRangeUnderflowMessageDetail: (p) => 'הזן חודש ושנה ב-' + p['min'] + ' או אחריו.',
    inputNumber_converterParseError: () => 'הזן מספר.',
    inputNumber_decrease: () => 'הקטן',
    inputNumber_increase: () => 'הגדל',
    inputPassword_hidden: () => 'סיסמה נסתרת',
    inputPassword_hide: () => 'הסתר סיסמה',
    inputPassword_show: () => 'הצג סיסמה',
    inputSensitiveText_hidden: () => 'טקסט רגיש נסתר',
    inputTimeMask_timeRangeOverflowMessageDetail: (p) => 'הזן שעה שמתחילה ב-' + p['max'] + ' או לפניה.',
    inputTimeMask_timeRangeUnderflowMessageDetail: (p) => 'הזן שעה שמתחילה ב-' + p['min'] + ' או אחריה.',
    inputTimeMask_time_cleared: () => 'שעת ניקוי',
    labelNone: () => 'ללא',
    list_msgFetchCompleted: () => 'נטען',
    list_msgFetchingData: () => 'טוען',
    list_suggestion: () => 'הצעה',
    menu_stateLoading: () => 'טוען',
    messageToast_allMessagesClosed: () => 'כל ההודעות המוקפצות סגורות.',
    message_close: () => 'סגור',
    message_confirmation: () => 'הצלחה',
    message_error: () => 'שגיאה',
    message_info: () => 'מידע',
    message_navigationFromMessagesRegion: () => 'מזין אזור הודעות. לחץ על F6 כדי לנווט חזרה לרכיב שהיה קודם במוקד.',
    message_navigationToMessagesRegion: () => 'אזור ההודעות מכיל הודעות חדשות. לחץ על F6 כדי לנווט לאזור ההודעות האחרון.',
    message_warning: () => 'אזהרה',
    noData_message: () => 'אין פריטים להצגה',
    overflowItemLabel: () => 'כרטיסיות נוספות',
    plural_separator: () => ', ',
    popup_close_skipLink: () => 'הקש פעמיים כדי לסגור את החלון הפתוח.',
    popup_liveRegion_F6_initial_focus_firstFocusable: () => 'מזין חלון צץ. לחץ על F6 כדי לנווט בין החלון הצץ והפקד המשויך.',
    popup_liveRegion_F6_initial_focus_none: () => 'חלון צץ נפתח. לחץ על F6 כדי לנווט בין החלון הצץ והפקד המשויך.',
    progressIndeterminate: () => 'בתהליך',
    radio_helpForRadio: () => 'עזרה לכפתור רדיו',
    radio_requiredMessageDetail: () => 'בחר ערך',
    selectMultiple_apply: () => 'החל',
    selectMultiple_back: () => 'חזרה',
    selectMultiple_countPlus: (p) => p['COUNT'] + '+',
    selectMultiple_highlightSelectedTagsInstructionText: () => 'השתמש במקשי החצים שמאלה וימינה כדי להדגיש את הערכים שנבחרו.',
    selectMultiple_removeSelectedTagInstructionText: () => 'הקש על Backspace או Delete כדי להסיר את הערך שנבחר.',
    selectMultiple_selectedValues: () => 'ערכים נבחרים',
    selectMultiple_showSelectedValues: () => 'הצג ערכים בלבד.',
    selectMultiple_valuesSelected: (p) => p['VALUE_COUNT'] + ' ערכים נבחרו.',
    select_addToList: () => 'הוספה לרשימה',
    select_addToListAvailable: () => 'הוספה לרשימה זמינה',
    select_moreSearchOptions: () => 'אפשרויות חיפוש נוספות',
    select_moreSearchOptionsAvailable: () => 'אפשרויות חיפוש נוספות זמינות',
    select_noMatchesFound: () => 'לא נמצאו התאמות.',
    select_oneMatchFound: () => 'נמצאה התאמה אחת',
    select_requiredMessageDetail: () => 'בחר ערך',
    select_sizeMatchesFound: (p) => p['TOTAL_SIZE'] + ' התאמות נמצאו',
    select_sizeOrMoreMatchesFound: (p) => p['TOTAL_SIZE'] + ' או יותר התאמות נמצאו',
    selectorAll_deselectAll: () => 'בטל את כל הבחירות',
    selectorAll_selectAll: () => 'בחר הכול',
    selector_selected: () => 'תיבת תיוג מסומנת',
    selector_unselected: () => 'תיבת תיוג לא מסומנת',
    sheet_close: () => 'סגור',
    tabBarNavigationList_removeCueText: () => 'ניתן להסרה',
    tabBar_tabContainsErrors: () => 'מכיל שגיאות',
    tabbar_accAnnounceRemovedKey: (p) => p['itemKey'] + ' הוסר',
    tabbar_labelContextMenu: () => 'כולל תפריט הקשר',
    tabbar_labelMoveLeft: () => 'העבר שמאלה',
    tabbar_labelMoveRight: () => 'העבר ימינה',
    tabbar_labelRemove: () => 'הסר',
    timeComponent_tooltipZoomIn: () => 'התקרבות',
    timeComponent_tooltipZoomOut: () => 'התרחקות',
    timeConverter_parseError: () => 'הזן שעת השלמה.',
    train_current: () => 'שוטף',
    train_message_type: () => 'סוג הודעה',
    train_not_visited: () => 'לא נרשמו ביקורים',
    train_status: (p) => 'שלב ' + p['currentStep'] + ' מתוך ' + p['numberOfSteps'],
    train_visited: () => 'נרשמו ביקורים',
    userAssistance_learnMore: () => 'פרטים נוספים',
    userAssistance_required: () => 'דרוש',
    vis_clearSelection: () => 'נקה בחירה',
    vis_drillable: () => 'ניתן לגלילה',
    vis_labelAndValue: (p) => p['LABEL'] + ': ' + p['VALUE'],
    vis_labelCountWithTotal: (p) => p['itemCount'] + ' מתוך ' + p['totalCount'],
    vis_marqueeSelectTooltip: () => 'בחירת ריבוע',
    vis_marqueeZoomTooltip: () => 'זום בריבוע',
    vis_noData: () => 'אין נתונים להצגה',
    vis_panTooltip: () => 'הזזה',
    vis_stateLoaded: () => 'נטען',
    vis_stateLoading: () => 'טוען',
    vis_stateSelected: () => 'נבחר',
    vis_stateUnselected: () => 'לא נבחר',
    warn: () => 'אזהרה'
};

export { bundle as default };
