// @oracle/oraclejet-webdriver/src/elements/oj-menu/OjMenu.ts
import { By, Key } from 'selenium-webdriver';
import { OjMenuBase } from './OjMenuBase';

/**
 * The component WebElement for [oj-menu](../../jsdocs/oj.ojMenu.html).
 * Do not instantiate this class directly, instead, use
 * [ojMenu](../functions/elements.ojMenu-1.html).
 */
export class OjMenu extends OjMenuBase {
  /**
   * Open the menu at the given launcher Id.  The launcher is the
   * DOM element which opens the menu when clicked.
   * @param launcher The Id of the element which launches the menu
   */
  async open(launcher: string) {
    await this.whenReady();
    const l = await this.getDriver().findElement(By.id(launcher));
    await l.click();
    await this.whenBusyContextReady();
  }
  /**
   * @deprecated Since 12.0.0. Use doMenuAction instead
   * Select a menu item from the list.  Note that the menu must
   * first be opened with {@link #open} before an item can be
   * selected.
   * @param id The Id of the &ltl;oj-option> to select
   */
  async changeSelection(id: string) {
    await this.whenReady();
    const item = await this.findElement(By.id(id));
    const a = await item.findElement(By.tagName('a')); // get the item anchor which should be active
    const aId = await a.getAttribute('id');
    let activeAId = await this.getAttribute('aria-activedescendant');
    while (activeAId !== aId) {
      await this.sendKeys(Key.DOWN);
      activeAId = await this.getAttribute('aria-activedescendant');
    }
    await item.click();
    await this.whenBusyContextReady();
  }
  /**
   * Fire the ojMenuAction event on the &lt;oj-menu>.
   * > Note that your application must be listening on the <b>oj-menu-action</b>
   * > event in the UI, not the deprecated <i>oj-action</i>.
   * @param selectedValue Value of the &ltl;oj-option> to select
   *
   */
  async doMenuAction(selectedValue: string) {
    await this.whenReady();
    return this.getDriver().executeScript<void>(
      (el: any, selectedValue: string) => {
        const evt = new CustomEvent('ojMenuAction', {
          detail: {
            selectedValue
          }
        });
        el.dispatchEvent(evt);
        el.close();
      },
      this,
      selectedValue
    );
  }
}
