// @oracle/oraclejet-testing/UNSAFE_Driver/TestDriver.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var waitFor = require('./waitFor.cjs');

/**
 * A base class for implementing the TestDriver interface. Platform-specific TestDrivers should
 * extend this class to inherit behavior for waiting on conditions.
 */
class TestDriverBase {
    /**
     * Wait for a condition to pass.
     * @param callback The function to run. Return truthy to stop the wait; anything else will retry
     * the callback. The truthy return value of the callback will be returned to the caller.
     * @param options Options for wait time/interval
     * @returns A Promise resolving to the returned value from the callback
     */
    async waitFor(callback, options) {
        return waitFor.waitFor_internal(callback, options);
    }
}
const ModifierKeys = {
    ALT: { key: 'ALT' },
    CONTROL: { key: 'CONTROL' },
    CONTROL_COMMAND: { key: 'CONTROL_COMMAND' },
    SHIFT: { key: 'SHIFT' }
};
const Keys = {
    ...ModifierKeys,
    BACKSPACE: { key: 'BACKSPACE' },
    TAB: { key: 'TAB' },
    ENTER: { key: 'ENTER' },
    ESCAPE: { key: 'ESCAPE' },
    PAGE_UP: { key: 'PAGE_UP' },
    PAGE_DOWN: { key: 'PAGE_DOWN' },
    END: { key: 'END' },
    HOME: { key: 'HOME' },
    ARROW_LEFT: { key: 'ARROW_LEFT' },
    ARROW_UP: { key: 'ARROW_UP' },
    ARROW_RIGHT: { key: 'ARROW_RIGHT' },
    ARROW_DOWN: { key: 'ARROW_DOWN' },
    DELETE: { key: 'DELETE' },
    F1: { key: 'F1' },
    F2: { key: 'F2' },
    F3: { key: 'F3' },
    F4: { key: 'F4' },
    F5: { key: 'F5' },
    F6: { key: 'F6' },
    F7: { key: 'F7' },
    F8: { key: 'F8' },
    F9: { key: 'F9' },
    F10: { key: 'F10' },
    F11: { key: 'F11' },
    F12: { key: 'F12' }
};
let _driver;
/**
 * Set the TestDriver instance to be used for testing. This method should be called by
 * tests during setup, before calling any test adapter methods.
 *
 * @param driver The TestDriver instance
 */
function setTestDriver(driver) {
    _driver = driver;
}
/**
 * Get the configured TestDriver instance to be used for testing. This method should
 * only be called by test adapters to interact with the browser environment.
 *
 * @returns The TestDriver instance
 */
function getTestDriver() {
    if (!_driver) {
        throw Error('Driver has not been set. Call setTestDriver()');
    }
    return _driver;
}
/**
 * Unwrap the given TestElement into its driver-specific element
 * @param element
 */
function unwrapElement(element) {
    return getTestDriver().unwrapElement(element);
}

exports.Keys = Keys;
exports.TestDriverBase = TestDriverBase;
exports.getTestDriver = getTestDriver;
exports.setTestDriver = setTestDriver;
exports.unwrapElement = unwrapElement;
