// @oracle/oraclejet-testing/UNSAFE_Driver/timeouts.mjs
const _timeouts = {
    /**
     * The window of time to wait for an element to exist in the DOM.
     * Set to 0 to wait indefinitely.
     * Defaults to 1000
     */
    elementExistsTimeout: 1000,
    /**
     * The window of time to wait for a remote script to complete execution.
     * Set to 0 to wait indefinitely.
     * Defaults to 30000
     */
    scriptTimeout: 30 * 1000,
    /**
     * The window of time with which the waitFor() function will retry the callback.
     * Set to 0 to wait indefinitely.
     * Defaults to 30000
     */
    waitForTimeout: 30 * 1000,
    /**
     * The delay between polling calls that waitFor() will wait before retrying the callback.
     * Defaults to 100
     */
    pollingInterval: 100
};
/**
 * Set the default timeouts for the TestDriver in milliseconds.
 * Every property is optional, and can be partially set.
 * @param timeouts
 */
function setTimeouts(timeouts) {
    Object.assign(_timeouts, timeouts);
}
/**
 * Get the default timeouts set for the TestDriver
 * @returns
 */
function getTimeouts() {
    return { ..._timeouts };
}

export { getTimeouts, setTimeouts };
