// @oracle/oraclejet-testing/UNSAFE_Driver/waitFor.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var timeouts = require('./timeouts.cjs');

/**
 * Wait for a condition to pass.
 * @param callback The function to run. Return truthy to stop the wait; anything else will retry
 * the callback. The truthy return value of the callback will be returned to the caller.
 * @returns A Promise resolving to the returned value from the callback
 */
async function waitFor(callback) {
    const { pollingInterval, waitForTimeout } = timeouts.getTimeouts();
    return waitFor_internal(callback, { timeout: waitForTimeout, interval: pollingInterval });
}
/**
 * Wait for a condition to pass.
 * @param callback The function to run. Return truthy to stop the wait; anything else will retry
 * the callback. The truthy return value of the callback will be returned to the caller.
 * @param options Options for wait time/interval
 * @returns A Promise resolving to the returned value from the callback
 */
async function waitFor_internal(callback, options) {
    const { waitForTimeout, pollingInterval } = timeouts.getTimeouts();
    const timeout = options?.timeout ?? waitForTimeout;
    const interval = options?.interval ?? pollingInterval;
    const start = performance.now();
    async function testCallback() {
        try {
            const result = await callback();
            if (result) {
                return result;
            }
            return retry();
        }
        catch {
            return retry();
        }
    }
    function retry() {
        const now = performance.now();
        const elapsed = Math.round(now - start);
        if (timeout && elapsed > timeout) {
            return Promise.reject(new WaitForTimeoutError(elapsed));
        }
        return new Promise((accept, reject) => setTimeout(() => testCallback().then(accept).catch(reject), interval));
    }
    return testCallback();
}
class WaitForTimeoutError extends Error {
    constructor(elapsed) {
        super(`timed out after ${elapsed}ms`);
        this.name = 'WaitForTimeoutError';
    }
}

exports.waitFor = waitFor;
exports.waitFor_internal = waitFor_internal;
