// @oracle/oraclejet-preact/cjs/UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });



var assistiveStyles = 'UserAssistanceStyles_assistiveStyles__14qa8xb3';
var dividerStyle = 'UserAssistanceStyles_dividerStyle__14qa8xb4';
var helpContentStyles = 'UserAssistanceStyles_helpContentStyles__14qa8xb7';
var helpIconBase = 'UserAssistanceStyles_helpIconBase__14qa8xb5';
var helpIconLabelEdgeStart = 'UserAssistanceStyles_helpIconLabelEdgeStart__14qa8xb6';
var iconUserAssistanceIcon = 'UserAssistanceStyles_iconUserAssistanceIcon__14qa8xba';
var iconUserAssistancePopupFocus = 'UserAssistanceStyles_iconUserAssistancePopupFocus__14qa8xbb';
var requiredIconBase = 'UserAssistanceStyles_requiredIconBase__14qa8xb8';
var requiredIconLabelEdgeStart = 'UserAssistanceStyles_requiredIconLabelEdgeStart__14qa8xb9';
var styles = {helpTextStyles:'UserAssistanceStyles_helpTextStyles__14qa8xb0',helpSourceStyles:'UserAssistanceStyles_helpSourceStyles__14qa8xb1',inlineContainerBaseStyles:'UserAssistanceStyles_inlineContainerBaseStyles__14qa8xb2'};
var variants = {align:{start:'UserAssistanceStyles_variants_align_start__14qa8xbc',end:'UserAssistanceStyles_variants_align_end__14qa8xbd'},visible:{hide:'UserAssistanceStyles_variants_visible_hide__14qa8xbe',show:''},container:{reflow:'UserAssistanceStyles_variants_container_reflow__14qa8xbf',efficient:'UserAssistanceStyles_variants_container_efficient__14qa8xbg'}};

exports.assistiveStyles = assistiveStyles;
exports.dividerStyle = dividerStyle;
exports.helpContentStyles = helpContentStyles;
exports.helpIconBase = helpIconBase;
exports.helpIconLabelEdgeStart = helpIconLabelEdgeStart;
exports.iconUserAssistanceIcon = iconUserAssistanceIcon;
exports.iconUserAssistancePopupFocus = iconUserAssistancePopupFocus;
exports.requiredIconBase = requiredIconBase;
exports.requiredIconLabelEdgeStart = requiredIconLabelEdgeStart;
exports.styles = styles;
exports.variants = variants;
