// @oracle/oraclejet-webdriver/elements/oj-selector/OjSelectorBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjSelectorBase = void 0;
const __1 = require("../");
/**
 * This is the base class for oj-selector WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjSelector.ts.
 */
class OjSelectorBase extends __1.OjWebElement {
    /**
     * Gets the value of <code>rowKey</code> property.
     * Specifies the row key of each selector. If the selectionMode property is 'all', rowKey is ignored.
     * @return The value of <code>rowKey</code> property.
     *
     */
    getRowKey() {
        return this.getProperty('rowKey');
    }
    /**
     * Sets the value of <code>indeterminate</code> property.
     * Visual only state to indicate partial selection
     * @param indeterminate The value to set for <code>indeterminate</code>
     *
     */
    changeIndeterminate(indeterminate) {
        return this.setProperty('indeterminate', indeterminate);
    }
    /**
     * Gets the value of <code>indeterminate</code> property.
     * Visual only state to indicate partial selection
     * @return The value of <code>indeterminate</code> property.
     *
     */
    getIndeterminate() {
        return this.getProperty('indeterminate');
    }
    /**
     * Sets the value of <code>selectedKeys</code> property.
     * Specifies the selectedKeys, should be hooked into the collection component.
     * @param selectedKeys The value to set for <code>selectedKeys</code>
     *
     */
    changeSelectedKeys(selectedKeys) {
        return this.setProperty('selectedKeys', selectedKeys);
    }
    /**
     * Gets the value of <code>selectedKeys</code> property.
     * Specifies the selectedKeys, should be hooked into the collection component.
     * @return The value of <code>selectedKeys</code> property.
     *
     */
    getSelectedKeys() {
        return this.getProperty('selectedKeys');
    }
    /**
     * Gets the value of <code>selectionMode</code> property.
     * Specifies the selection mode.
     * @return The value of <code>selectionMode</code> property.
     *
     */
    getSelectionMode() {
        return this.getProperty('selectionMode');
    }
}
exports.OjSelectorBase = OjSelectorBase;
