// @oracle/oraclejet-webdriver/src/elements/oj-chart/OjChart.ts
import { OjChartBase } from './OjChartBase';

/**
 * The component WebElement for [oj-chart](../../jsdocs/oj.ojChart.html).
 * Do not instantiate this class directly, instead, use
 * [ojChart](../functions/elements.ojChart-1.html).
 */

/**
 * The component WebElement for [oj-chart](../../jsdocs/oj.ojChart.html).
 * Do not instantiate this class directly, instead, use
 * [ojChart](../functions/elements.ojChart-1.html).
 *  <h4>Migration</h4>
 *  oj-chart is being broken down into multiple core pack chart components.
 *  When you migrate a page from using [oj-chart](../../jsdocs/oj.ojChart.html) to new core pack chart components [oj-c-line-chart](../../jsdocs/oj-c.LineChart.html), [oj-c-area-chart](../../jsdocs/oj-c.AreaChart.html), etc you will also need to migrate your WebDriver tests from OjChart to, LineChartWebElement, AreaChartWebElement, etc. To do that you will need to revise type and locator instances as follows
 *  </br><b>Type:</b> from OjChart to [LineChartWebElement](../../cpwdtsdoc/classes/LineChartWebElement.html)
 *  </br><b>Locator:</b> from [ojChart](../functions/elements.ojChart-1.html) to [findLineChart](../../cpwdtsdoc/functions/findLineChart.html)
 *  </br>For more details on migrating from legacy to core pack please see component's [migration section](../../jsdocs/oj.ojChart.html#migration-section)
 */

export class OjChart extends OjChartBase {
  /**
   * Dispatches group drill event.
   * @param {Array<string>} groupIds Id of the group axis label. It will be an array of outermost to innermost group data in hierarchical axis.
   */
  async doGroupDrill(groupId: Array<string>) {
    let drilling = await this.getDrilling();

    if (drilling === 'on' || drilling === 'groupsOnly') {
      await this.getDriver().executeScript(
        (chart: any, groupId: Array<string>) => chart.getAutomation().dispatchGroupDrill(groupId),
        this,
        groupId
      );
    }
  }

  /**
   * Dispatches series drill event.
   * @param {string} seriesId The series id of the item.
   */

  async doSeriesDrill(seriesId: string) {
    let drilling = await this.getDrilling();
    if (drilling === 'on' || drilling === 'seriesOnly') {
      await this.getDriver().executeScript(
        (chart: any, seriesId: string) => chart.getAutomation().dispatchSeriesDrill(seriesId),
        this,
        seriesId
      );
    }
  }

  /**
   * Dispatches item drill event.
   * @param {string} seriesId The series id of the item.
   * @param {Array<string>} groupId The group ids of the item.
   * @param {number}itemIndex
   */
  async doItemDrill(seriesId: string, groupId: Array<string>, itemIndex: number) {
    let drilling = await this.getDrilling();
    if (drilling === 'on' || drilling === 'itemsOnly') {
      await this.getDriver().executeScript(
        (chart: any, seriesId: string, groupId: Array<string>, itemIndex: number) =>
          chart.getAutomation().dispatchItemDrill(seriesId, groupId, itemIndex),
        this,
        seriesId,
        groupId,
        itemIndex
      );
    }
  }

  /**
   * Dispatches multi series drill in pie chart.
   */
  async doMultiSeriesDrill() {
    let multiSeriesDrill = await this.getMultiSeriesDrilling();
    if (multiSeriesDrill === 'on') {
      await this.getDriver().executeScript(
        (chart: any) => chart.getAutomation().dispatchMultiSeriesDrill(),
        this
      );
    }
  }
}
