// @oracle/oraclejet-webdriver/src/elements/oj-message/OjMessageBase.ts
import { OjWebElement } from '../';

/**
 * This is the base class for oj-message WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjMessage.ts.
 */
export class OjMessageBase extends OjWebElement {
  /**
   * Gets the value of <code>displayOptions</code> property.
   * Specifies the display options for contents of the message.
   * @return The value of <code>displayOptions</code> property.
   *
   */
  public getDisplayOptions(): Promise<DisplayOptions> {
    return this.getProperty<DisplayOptions>('displayOptions');
  }

  /**
   * Gets the value of <code>message</code> property.
   * Structured data specifying essential information to display a message on the UI.
   * @return The value of <code>message</code> property.
   *
   */
  public getMessage(): Promise<Message> {
    return this.getProperty<Message>('message');
  }

  /**
   * Gets the value of <code>translations</code> property.
   * A collection of translated resources from the translation bundle, or null if this component has no resources.
   * @return The value of <code>translations</code> property.
   *
   */
  public getTranslations(): Promise<Translations> {
    return this.getProperty<Translations>('translations');
  }
}
export interface DisplayOptions {
  /**
   * Specifies the display option for message category text in this message.
   */
  category: string;
}
export interface Message {
  /**
   * Specifies the duration in milliseconds that the message will be shown before it closes automatically. See the Help documentation for more information.
   */
  autoTimeout: number;
  /**
   * Specifies message category text which is shown next to the message icon. See the Help documentation for more information.
   */
  category: string;
  /**
   * Specifies the UI affordance provided to end users to be able to close the message. See the Help documentation for more information.
   */
  closeAffordance: string;
  /**
   * Specifies detail text for the message.
   */
  detail: string;
  /**
   * Specifies the URL for the custom image to be used as an icon representing the message. See the Help documentation for more information.
   */
  icon: string;
  /**
   * Specifies the severity of message.
   */
  severity: string;
  /**
   * Specifies the sound to be played when a message is opened. This is needed for accessibility. See the Help documentation for more information.
   */
  sound: string;
  /**
   * Specifies summary text for the message.
   */
  summary: string;
  /**
   * Specifies a timestamp for the message to be displayed in the message header. See the Help documentation for more information.
   */
  timestamp: string;
}
export interface Translations {
  /**
   * Category text to be displayed for the messages of different severity levels.
   */
  categories: TranslationsCategories;
  /**
   * Label for the message close button.
   */
  labelCloseIcon: string;
}
export interface TranslationsCategories {
  /**
   * Category text for the 'confirmation' severity level.
   */
  confirmation: string;
  /**
   * Category text for the 'error' severity level.
   */
  error: string;
  /**
   * Category text for the 'info' severity level.
   */
  info: string;
  /**
   * Category text when there is no severity level.
   */
  none: string;
  /**
   * Category text for the 'warning' severity level.
   */
  warning: string;
}
