// @oracle/oraclejet-preact/cjs/resources/nls/pl/bundle.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

// This file is auto-generated by @oracle/oraclejet-icu-l10n from src/resources/icu/bundle.json
// and should not be edited by hand. Update the JSON file and rerun the build to regenerate this content.
const bundle = {
    accAnnounceMovedKey: (p) => 'Przeniesiono: ' + p['itemKey'] + ' ' + p['position'] + ' ' + p['referenceKey'] + '.',
    accButtonProgressComplete: (p) => 'Czynność "' + p['label'] + '" ukończona',
    accButtonProgressStatus: (p) => 'Czynność "' + p['label'] + '" w toku',
    accColorPalette: () => 'Paleta kolorów',
    accColorSwatch: () => 'Paleta kolorów',
    accCommandKey: () => 'Polecenie',
    accContextMenuLabel: () => 'czynności',
    accControlKey: () => 'Zamknij',
    accPositionAfter: () => 'po',
    accPositionBefore: () => 'przed',
    accReorderInstructions: (p) => 'Aby zmienić kolejność, należy nacisnąć i przytrzymać przyciski ' +
        p['keyPressed'] +
        ' i Shift, a następnie użyć przycisków strzałek.',
    calendarDateConverter_parseError: (p) => 'Należy wprowadzić pełną datę, używając formatu podobnego do następującego: ' +
        p['dateExample'] +
        '.',
    calendarMonthConverter_parseError: (p) => 'Należy wprowadzić miesiąc i rok, używając formatu podobnego do następującego: ' +
        p['dateExample'] +
        '.',
    chart_labelGroup: () => 'Grupa',
    chart_labelSeries: () => 'Seria',
    chart_labelValue: () => 'Wartość',
    chart_labelWidthWeight: () => 'Z',
    chart_labelX: () => 'X',
    chart_labelY: () => 'Y',
    checkboxSet_requiredMessageDetail: () => 'Proszę wybrać wartość.',
    checkbox_requiredMessageDetail: () => 'Proszę wybrać wartość.',
    close: () => 'Zamknij',
    collection_apply: () => 'Zastosuj',
    collection_cancel: () => 'Anuluj',
    collection_loaded: () => 'Załadowano',
    collection_loading: () => 'Ładowanie',
    collection_noData: () => 'Brak danych do wyświetlenia.',
    collection_pixels: () => 'piks.',
    collection_resizeColumnDialog: () => 'Zmienianie rozmiaru kolumny',
    collection_resizeColumnMenu: () => 'Zmień rozmiar kolumny...',
    collection_resizeMessageDetail: (p) => 'Proszę podać liczbę z przedziału od ' + p['MIN'] + ' do ' + p['MAX'] + '.',
    collection_selectAllRows: () => 'Wybór wszystkich wierszy',
    collection_selectRow: (p) => 'Wybór wiersza "' + p['ROW_NAME'] + '"',
    collection_sortAscendingSubMenu: () => 'Rosnąco',
    collection_sortColumnMenu: () => 'Sortuj kolumnę',
    collection_sortDescendingSubMenu: () => 'Malejąco',
    collection_sortDisabled: () => 'Niesortowalne',
    collection_sortEnabled: () => 'Sortowalne',
    collection_width: () => 'Szerokość',
    colorPicker_label: () => 'Selektor kolorów',
    confirmation: () => 'Potwierdzenie',
    datePicker_currentDate: () => 'Data bieżąca',
    datePicker_goToToday: () => 'Przejdź do dnia dzisiejszego',
    datePicker_label: () => 'Selektor daty',
    datePicker_month: () => 'Miesiąc',
    datePicker_months: () => 'Miesiące',
    datePicker_next: () => 'Następne',
    datePicker_nextDecade: () => 'Następna dekada',
    datePicker_nextMonth: () => 'Następny miesiąc',
    datePicker_prev: () => 'Poprzednie',
    datePicker_prevDecade: () => 'Poprzednia dekada',
    datePicker_prevMonth: () => 'Poprzedni miesiąc',
    datePicker_restricted: () => 'Ograniczone',
    datePicker_selected: () => 'Wybrane',
    datePicker_today: () => 'Dzisiaj',
    datePicker_weekNum: (p) => 'Tydzień ' + p['NUM'],
    datePicker_weekNumColLabel: () => 'Numer tygodnia',
    datePicker_year: () => 'Rok',
    datePicker_years: () => 'Lata',
    error: () => 'Błąd',
    expandableList_expandCollapseInstructionText: () => 'Do rozwijania i zwijania proszę używać klawiszy ze strzałkami',
    expandableList_groupCollapse: () => 'Zwinięte',
    expandableList_groupExpand: () => 'Rozwinięte',
    filePicker_addFiles: () => 'Dodaj pliki',
    filePicker_dropzonePrimaryText: () => 'Przeciąganie i upuszczanie',
    filePicker_dropzoneSecondaryText: () => 'Proszę wybrać plik lub go tutaj przeciągnąć',
    filePicker_dropzoneSecondaryTextMultiple: () => 'Proszę wybrać pliki lub je tutaj przeciągnąć',
    filePicker_multipleFileTypeUploadError: (p) => 'Nie można wysłać plików typu ' + p['fileTypes'],
    filePicker_singleFileUploadError: () => 'Proszę wysyłać po jednym pliku',
    filePicker_singleTypeUploadError: (p) => 'Nie można wysłać plików typu ' + p['fileType'],
    filePicker_unknownFileTypeUploadError: () => 'nieznany',
    formControl_clear: () => 'Wyczyść',
    formControl_day: () => 'Dzień',
    formControl_empty_segment: () => 'Pusty',
    formControl_helpAvailable: () => 'Pomoc dostępna',
    formControl_hide: () => 'Ukryj',
    formControl_hour: () => 'Godzina',
    formControl_limitReached: (p) => 'Osiągnięto limit wynoszący ' +
        ({
            one: new Intl.NumberFormat('pl').format(p['CHARACTER_LIMIT']) + ' znak',
            few: new Intl.NumberFormat('pl').format(p['CHARACTER_LIMIT']) + ' znaki',
            many: new Intl.NumberFormat('pl').format(p['CHARACTER_LIMIT']) + ' znaków'
        }[new Intl.PluralRules('pl').select(p['CHARACTER_LIMIT'])] ||
            new Intl.NumberFormat('pl').format(p['CHARACTER_LIMIT']) + ' znaku') +
        '.',
    formControl_loading: () => 'Ładowanie',
    formControl_maxLength: (p) => 'Można wprowadzić maksymalnie następującą liczbę znaków: ' + p['MAX_LENGTH'] + '.',
    formControl_maxLengthExceeded: (p) => 'Przekroczono maksymalną długość ' + p['MAX_LENGTH'] + '.',
    formControl_maxLengthRemaining: (p) => 'Pozostało znaków: ' + p['CHARACTER_COUNT'] + '.',
    formControl_millisecond: () => 'Milisekunda',
    formControl_minute: () => 'Minuta',
    formControl_month: () => 'Miesiąc',
    formControl_readOnly: () => 'Tylko do odczytu',
    formControl_requiredMessageDetail: () => 'Proszę wprowadzić wartość.',
    formControl_second: () => 'Sekunda',
    formControl_selectionExactMessageDetail: (p) => 'Należy wybrać ' + p['n'] + ' wartości.',
    formControl_selectionOverflowMessageDetail: (p) => 'Osiągnięto limit wyboru wynoszący ' + p['max'] + '.',
    formControl_selectionRangeMessageDetail: (p) => 'Należy wybrać od ' + p['min'] + ' do ' + p['max'] + ' wartości.',
    formControl_selectionUnderflowMessageDetail: (p) => 'Należy wybrać co najmniej ' + p['min'] + ' wartości.',
    formControl_show: () => 'Pokaż',
    formControl_year: () => 'Rok',
    gantt_labelDate: () => 'Data',
    gantt_labelEnd: () => 'Koniec',
    gantt_labelLabel: () => 'Etykieta',
    gantt_labelRow: () => 'Wiersz',
    gantt_labelStart: () => 'Początek',
    indexer_disabledLabel: (p) => p['SECTION'] + '. Nie znaleziono zgodnej sekcji.',
    indexer_keyboardInstructionText: () => 'Aby wybrać wartość, proszę nacisnąć klawisz Enter.',
    indexer_othersLabel: () => '#',
    indexer_separatorLabel: (p) => 'Między ' + p['FROM_SECTION'] + ' a ' + p['TO_SECTION'] + '.',
    indexer_touchInstructionText: () => 'Aby przełączyć się do trybu gestów, proszę 2x stuknąć i przytrzymać, a następnie - aby dostosować wartość - przeciągnąć w górę lub w dół.',
    info: () => 'Informacja',
    inputColor_selectColor: () => 'Proszę wybrać kolor',
    inputDateMask_date_cleared: () => 'Data wyczyszczenia',
    inputDateMask_dayPlaceholder: () => 'dd',
    inputDateMask_empty_segment: () => 'Pusty',
    inputDateMask_monthPlaceholder: () => 'mm',
    inputDateMask_yearPlaceholder: () => 'rrrr',
    inputDatePicker_dateRangeOverflowMessageDetail: (p) => 'Należy wprowadzić datę ' + p['max'] + ' lub wcześniejszą.',
    inputDatePicker_dateRangeUnderflowMessageDetail: (p) => 'Należy wprowadzić datę ' + p['min'] + ' lub późniejszą.',
    inputDatePicker_dateRestrictionMessageDetail: () => 'Należy wprowadzić dostępną datę.',
    inputDatePicker_instruction: () => 'Aby uzyskać dostęp do kalendarza, należy nacisnąć klawisz Tab.',
    inputDatePicker_selectDate: () => 'Proszę wybrać datę',
    inputMonthMask_dateRangeOverflowMessageDetail: (p) => 'Należy wprowadzić miesiąc i rok przypadający na ' + p['max'] + ' lub wcześniejszy.',
    inputMonthMask_dateRangeUnderflowMessageDetail: (p) => 'Należy wprowadzić miesiąc i rok przypadający na ' + p['min'] + ' lub późniejszy.',
    inputNumber_converterParseError: () => 'Proszę wprowadzić liczbę.',
    inputNumber_decrease: () => 'Zmniejsz',
    inputNumber_increase: () => 'Zwiększ',
    inputPassword_hidden: () => 'Hasło ukryte',
    inputPassword_hide: () => 'Ukryj hasło',
    inputPassword_show: () => 'Pokaż hasło',
    inputSensitiveText_hidden: () => 'Tekst danych wrażliwych ukryty',
    inputTimeMask_timeRangeOverflowMessageDetail: (p) => 'Należy wprowadzić godzinę ' + p['max'] + ' lub wcześniejszą.',
    inputTimeMask_timeRangeUnderflowMessageDetail: (p) => 'Należy wprowadzić godzinę ' + p['min'] + ' lub późniejszą.',
    inputTimeMask_time_cleared: () => 'Czas wyczyszczenia',
    labelNone: () => 'Brak',
    list_msgFetchCompleted: () => 'Załadowano',
    list_msgFetchingData: () => 'Ładowanie',
    list_suggestion: () => 'Sugestia',
    menu_stateLoading: () => 'Ładowanie',
    messageToast_allMessagesClosed: () => 'Wszystkie wyskakujące powiadomienia są zamknięte.',
    message_close: () => 'Zamknij',
    message_confirmation: () => 'Powodzenie',
    message_error: () => 'Błąd',
    message_info: () => 'Informacja',
    message_navigationFromMessagesRegion: () => 'Wchodzenie do obszaru komunikatów. Aby się cofnąć do elementu, na którym znajdował się fokus, proszę nacisnąć klawisz F6.',
    message_navigationToMessagesRegion: () => 'W obszarze komunikatów są nowe komunikaty. Aby przejść do nich, proszę nacisnąć klawisz F6.',
    message_warning: () => 'Ostrzeżenie',
    noData_message: () => 'Brak elementów do wyświetlenia',
    overflowItemLabel: () => 'Więcej kart',
    plural_separator: () => ', ',
    popup_close_skipLink: () => 'Aby zamknąć okno wyskakujące, proszę dwa razy stuknąć.',
    popup_liveRegion_F6_initial_focus_firstFocusable: () => 'Wchodzenie do okna wyskakującego. Do przechodzenia między oknem wyskakującym a powiązanym formantem proszę używać klawisza F6.',
    popup_liveRegion_F6_initial_focus_none: () => 'Okno wyskakujące otwarte. Do przechodzenia między oknem wyskakującym a powiązanym formantem proszę używać klawisza F6.',
    progressIndeterminate: () => 'W toku',
    radio_helpForRadio: () => 'Pomoc dot. przycisku radiowego',
    radio_requiredMessageDetail: () => 'Proszę wybrać wartość.',
    selectMultiple_apply: () => 'Zastosuj',
    selectMultiple_back: () => 'Wstecz',
    selectMultiple_countPlus: (p) => p['COUNT'] + '+',
    selectMultiple_highlightSelectedTagsInstructionText: () => 'Aby wyróżnić wybrane wartości, należy użyć klawiszy ze strzałką w lewo lub w prawo.',
    selectMultiple_removeSelectedTagInstructionText: () => 'Aby usunąć wybraną wartość, proszę nacisnąć klawisz Backspace lub Delete',
    selectMultiple_selectedValues: () => 'Wybrane wartości',
    selectMultiple_showSelectedValues: () => 'Pokaż tylko wybrane wartości.',
    selectMultiple_valuesSelected: (p) => p['VALUE_COUNT'] + ' wybrane wartości.',
    select_addToList: () => 'Dodaj do listy',
    select_addToListAvailable: () => 'Dodaj do listy (dostępne)',
    select_moreSearchOptions: () => 'Więcej opcji wyszukiwania',
    select_moreSearchOptionsAvailable: () => 'Więcej opcji wyszukiwania (dostępne)',
    select_noMatchesFound: () => 'Nie znaleziono zgodnych.',
    select_oneMatchFound: () => 'Znaleziono zgodnych: 1',
    select_requiredMessageDetail: () => 'Proszę wybrać wartość.',
    select_sizeMatchesFound: (p) => 'Znaleziono zgodnych: ' + p['TOTAL_SIZE'],
    select_sizeOrMoreMatchesFound: (p) => 'Znaleziono zgodnych: ' + p['TOTAL_SIZE'] + ' lub więcej',
    selectorAll_deselectAll: () => 'Cofnij wybór wszystkich',
    selectorAll_selectAll: () => 'Wybierz wszystkie',
    selector_selected: () => 'Pole wyboru zaznaczone',
    selector_unselected: () => 'Pole wyboru niezaznaczone',
    sheet_close: () => 'Zamknij',
    tabBarNavigationList_removeCueText: () => 'Możliwe do usunięcia',
    tabBar_tabContainsErrors: () => 'Zawiera błędy',
    tabbar_accAnnounceRemovedKey: (p) => 'Usunięto "' + p['itemKey'] + '"',
    tabbar_labelContextMenu: () => 'zawiera menu kontekstowe',
    tabbar_labelMoveLeft: () => 'Przenieś w lewo',
    tabbar_labelMoveRight: () => 'Przenieś w prawo',
    tabbar_labelRemove: () => 'Usuń',
    timeComponent_tooltipZoomIn: () => 'Powiększ',
    timeComponent_tooltipZoomOut: () => 'Pomniejsz',
    timeConverter_parseError: () => 'Należy wprowadzić godzinę w pełnym formacie.',
    train_current: () => 'Bieżące',
    train_message_type: () => 'Typ wiadomości',
    train_not_visited: () => 'Nieodwiedzone',
    train_status: (p) => 'Etap ' + p['currentStep'] + ' z ' + p['numberOfSteps'],
    train_visited: () => 'Odwiedzone',
    userAssistance_learnMore: () => 'Dalsze informacje',
    userAssistance_required: () => 'Wymagane',
    vis_clearSelection: () => 'Wyczyść wybór',
    vis_drillable: () => 'Możliwe do drążenia',
    vis_labelAndValue: (p) => p['LABEL'] + ': ' + p['VALUE'],
    vis_labelCountWithTotal: (p) => p['itemCount'] + ' z ' + p['totalCount'],
    vis_marqueeSelectTooltip: () => 'Zaznaczanie obszaru',
    vis_marqueeZoomTooltip: () => 'Powiększanie obszaru',
    vis_noData: () => 'Brak danych do wyświetlenia',
    vis_panTooltip: () => 'Panoramowanie',
    vis_stateLoaded: () => 'Załadowano',
    vis_stateLoading: () => 'Ładowanie',
    vis_stateSelected: () => 'Wybrane',
    vis_stateUnselected: () => 'Niewybrane',
    warn: () => 'Ostrzeżenie'
};

exports["default"] = bundle;
