// @oracle/oraclejet-webdriver/elements/oj-list-view/OjListView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjListView = void 0;
const OjListViewBase_1 = require("./OjListViewBase");
const __1 = require("../../");
const selenium_webdriver_1 = require("selenium-webdriver");
const oj_module_proxy_1 = require("../../lib/oj-module-proxy");
/**
 * The component WebElement for [oj-list-view](../../jsdocs/oj.ojListView.html).
 * Do not instantiate this class directly, instead, use
 * [ojListView](../functions/elements.ojListView-1.html).
 */
class OjListView extends OjListViewBase_1.OjListViewBase {
    /**
     * Sets the value of "selected" property.
     * Specifies the current selected items in the listview. See the Help documentation for more information.
     * @param selected The value to set for "selected"
     * @override
     * @typeparam K Type of keys
     */
    async changeSelected(selected) {
        await this.whenReady();
        return (0, oj_module_proxy_1.executeWithModules)(this.getDriver(), ['KeySet'], [this, selected], ({ KeySetImpl }, ele, selected) => {
            ele.selected = new KeySetImpl(selected);
        });
    }
    /**
     * Gets the value of "selected" property.
     * Retrieves the current selected items in the listview. See the Help documentation for more information.
     * @override
     * @typeparam K Type of keys
     * @return The value of "selected" property.
     */
    async getSelected() {
        await this.whenReady();
        const selected = await this.getDriver().executeScript((ele) => {
            return Array.from(ele.selected.values());
        }, this);
        return selected;
    }
    /**
     * Retrieve a SlotProxy which represents a single listview item.
     * @param key The key within the Collection's dataset associated with the item.
     */
    async findItem(itemLocator) {
        // explicitly override scrollToKey to ensure changeScrollPosition regardless
        // of DataProvider or existing scrollToKey value
        await this.setProperty('scrollToKey', 'always');
        // Call changeScrollPosition to scroll the row into view and ensure it's
        // rendere in DOM
        // TODO: Base class should mark position properties as optional
        // @ts-ignore
        await this.changeScrollPosition({
            key: itemLocator.key
        });
        await this.whenBusyContextReady();
        const display = await this.getProperty('display');
        let el = await this.getDriver().executeScript((ele, key) => ele.getNodeBySubId({ subId: 'oj-listview-item', key }), this, itemLocator.key);
        if (display === 'list') {
            if (el) {
                el = el.findElement(selenium_webdriver_1.By.css('[role="gridcell"]'));
            }
            else {
                throw new Error('The key specified is invalid or does not exist');
            }
        }
        return (0, __1.slotProxy)(el, this);
    }
    /**
     * Move an item specified by key to the specified position.
     * @param key key of the item to be reordered
     * @param position the position where the item should be move to, the value can be one of:
     * <ul>
     * <li>key - the key of the item to move in front of. If the key is null, the item will be move to the end of the list</li>
     * <li>index - the index of the item relative to the root node to move in front of</li>
     * <li>groupKey with itemIndex - the key of the group to move the item inside. The itemIndex indicates the index of the item relative to the group in which to move in front of. If not specified, then it will be moved to the end of the group</li>
     * <li>groupIndex with itemIndex - an array of indexes to locate the group to move the item inside. The itemIndex indicates the index of the item relative to the group in which to move in front of. If not specified, then it will be moved to the end of the group</li>
     * </ul>
     * @override
     * @typeparam K Type of keys
     * @typeparam number Type of index
     */
    async doReorder(key, position) {
        await this.whenReady();
        await (0, oj_module_proxy_1.executeWithModules)(this.getDriver(), ['CustomElementUtils'], [this, key, position], ({ getElementBridge }, element, key, position) => {
            const bridge = getElementBridge(element);
            const widget = bridge._WIDGET_INSTANCE;
            return widget.listview._doReorderHelper(null, key, position);
        });
    }
}
exports.OjListView = OjListView;
