// @oracle/oraclejet-preact/cjs/UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.lg.tokens.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('../../../utils-a66430fc.js');
var themeContract_css = require('../../../themeContract.css-a53fd740.js');
require('../../../colorUtils-01f3caa2.js');
require('../../../_curry1-e8f0d7ea.js');

const scaleSchemeLg = {
    borderRadius: themeContract_css.baseVars.borderRadius.md,
    fontSize: themeContract_css.typographyVars.body.sm.fontSize,
    sizes: {
        xs: {
            textToEdgePadding: utils.xUnits(4),
            // 24px x 24px is the minimum size per WCAG Success Criterion 2.5.8 Target Size (Minimum)
            height: utils.max(`calc(${utils.xUnits(6)} + ${utils.densityUnits(3)})`, '24px')
        },
        sm: {
            textToEdgePadding: utils.xUnits(4),
            // 24px x 24px is the minimum size per WCAG Success Criterion 2.5.8 Target Size (Minimum)
            height: utils.max(`calc(${utils.xUnits(9)} + ${utils.densityUnits(3)})`, '24px')
        },
        md: {
            textToEdgePadding: utils.xUnits(4),
            height: `calc(${utils.xUnits(11)} + ${utils.densityUnits(2)})`
        },
        lg: {
            textToEdgePadding: utils.xUnits(6),
            height: `calc(${utils.xUnits(13)} + ${utils.densityUnits(2)})`
        }
    }
};

exports.scaleSchemeLg = scaleSchemeLg;
