// @oracle/oraclejet-webdriver/elements/oj-label/OjLabelBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OjLabelBase = void 0;
const __1 = require("../");
/**
 * This is the base class for oj-label WebElement, and is generated from the
 * component's metadata. Do not modify these contents since they'll be replaced
 * during the next generation.
 * Put overrides into the WebElements's subclass, OjLabel.ts.
 */
class OjLabelBase extends __1.OjWebElement {
    /**
     * Gets the value of <code>for</code> property.
     * Specifies the form element associated with this label. See the Help documentation for more information.
     * @return The value of <code>for</code> property.
     *
     */
    getFor() {
        return this.getProperty('for');
    }
    /**
     * Gets the value of <code>help</code> property.
     * Help information associated with this label.
     * @return The value of <code>help</code> property.
     *
     */
    getHelp() {
        return this.getProperty('help');
    }
    /**
     * Gets the value of <code>labelId</code> property.
     * Specifies the id to set on the internal label element, if required. See the Help documentation for more information.
     * @return The value of <code>labelId</code> property.
     *
     */
    getLabelId() {
        return this.getProperty('labelId');
    }
    /**
     * Gets the value of <code>showRequired</code> property.
     * Specifies whether the label should render an icon indicating that the associated form field requires a value. See the Help documentation for more information.
     * @return The value of <code>showRequired</code> property.
     *
     */
    getShowRequired() {
        return this.getProperty('showRequired');
    }
    /**
     * Gets the value of <code>translations</code> property.
     * A collection of translated resources from the translation bundle, or null if this component has no resources.
     * @return The value of <code>translations</code> property.
     *
     */
    getTranslations() {
        return this.getProperty('translations');
    }
}
exports.OjLabelBase = OjLabelBase;
